<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Workshop 11: Simulations in R</title>
    <meta charset="utf-8" />
    <meta name="author" content="Québec Centre for Biodiversity Science" />
    <script src="assets/header-attrs-2.11/header-attrs.js"></script>
    <link href="assets/remark-css-0.0.1/default.css" rel="stylesheet" />
    <script src="assets/htmlwidgets-1.5.3/htmlwidgets.js"></script>
    <script src="assets/jquery-1.12.4/jquery.min.js"></script>
    <link href="assets/leaflet-1.3.1/leaflet.css" rel="stylesheet" />
    <script src="assets/leaflet-1.3.1/leaflet.js"></script>
    <link href="assets/leafletfix-1.0.0/leafletfix.css" rel="stylesheet" />
    <script src="assets/proj4-2.6.2/proj4.min.js"></script>
    <script src="assets/Proj4Leaflet-1.0.1/proj4leaflet.js"></script>
    <link href="assets/rstudio_leaflet-1.3.1/rstudio_leaflet.css" rel="stylesheet" />
    <script src="assets/leaflet-binding-2.0.4.1/leaflet.js"></script>
    <script src="assets/leaflet-providers-1.9.0/leaflet-providers_1.9.0.js"></script>
    <script src="assets/leaflet-providers-plugin-2.0.4.1/leaflet-providers-plugin.js"></script>
    <link href="assets/HomeButton-0.0.1/home-button.css" rel="stylesheet" />
    <script src="assets/HomeButton-0.0.1/home-button.js"></script>
    <script src="assets/HomeButton-0.0.1/easy-button-src.min.js"></script>
    <script src="assets/clipboard-0.0.1/setClipboardText.js"></script>
    <link href="assets/mapviewCSS-0.0.1/mapview-popup.css" rel="stylesheet" />
    <link href="assets/mapviewCSS-0.0.1/mapview.css" rel="stylesheet" />
    <link rel="stylesheet" href="assets/font-awesome-4.7.0/css/font-awesome.min.css">
    <link rel="stylesheet" href="assets/academicons-1.8.1/css/academicons.min.css">
    <link rel="stylesheet" href="qcbsR.css" type="text/css" />
    <link rel="stylesheet" href="qcbsR-fonts.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Workshop 11: Simulations in R
## QCBSx R Workshop Series <br> x = independent
### Québec Centre for Biodiversity Science

---




## Outline

1. Why are simulations relevant? 
2. What you need to know before doing simulations? 
3. Let's explore what underlie some statistical processes
3. Simulating data for power analysis (in development)
4. Simulating data for models (lm, glm, lmm, in development)

Please check out the [Simulation in R Cheat Sheet](https://docs.google.com/presentation/d/11bNgLUCyvlL8Ndq_nF-N6T71dZwgETbhbSHuc40VqvU/edit#slide=id.p)

???

This workshop was designed to give the tools for attendees to perform simulations. 
- Simulations to get data in order to perform a statistical model, 
- but also to be able to simulate a situation and verify how a parameter change the outcome of a model. 

---
## Prerequisites
&lt;!-- I will be using base plot for this workshop: this is because it is MUCH EASIER to get quick BEAUTIFUL plots and show stuff --&gt;
For this workshop it is useful to have a solid understanding of 

1. Linear models and generalized linear models (GLMs)
2. Programming in R 

See the [QCBS R workshops](https://r.qcbs.ca/) if you want to revise these topics. Select the [R workshops](http://r.qcbs.ca/workshops/) needed.

- We are going to review some aspects of these workshops, but getting some experience from those subjects can clearly help you for this workshop. 
- We are going to be more explicitly show assumptions of models and use for loops in this workshop 

???

This workshop builds on the previous workshops. But don't worry, we are going to refresh some of the concepts here. 

---
## Disclaimer

- I am not a statistician and do not pretend to fully understand all the implications of what will be presented. 
- In addition, this is a work in progress. Any input you have is more than welcome. 

???

Might not apply to you

---
## Learning objectives

1. Develop intuition and skills to perform simulations
2. Explore useful functions when designing simulations 
3. Articulate what are the assumptions underlying a simulation (for the models tested or the scope of the simulation)
4. Know how and when to use important statistical distributions
5. Simulate different data structure to perform LMs, GLMs, and (G)LMMs. 

???

Hopefully, this is what people will get out of this workshop. 


#################################
---
class: inverse, center, middle

# Why are simulations relevant?

???

In this section, we will explore 
- What are simulations 
- What is the idea behind simulations
- What statistics has to do with simulations.
- Understand, with all of these, wy simulations are relevant.

- See [Beginner’s Guide to Simulating Data](http://www.page-gould.com/simulation/Guide-to-Simulation-for-R.html)

---
## Defining simulations

&lt;br&gt;
&lt;br&gt;

.center[
[...] the term 'simulation' describes a wealth of varied and useful techniques, all connected with the mimicking of the rules of a model of some kind.
]

.right[— Byron J. T. Morgan 1984]

???
- and models are lke maps: they are an abstraction of the real worl. It is not the real world. No same person thinks a map is "right". The only map that is right is the 1:1 real map. But this map woulb not be interesting to get a sense of how the world is. 

---
## Simulations are games 

- Simulations are games that we make up the rules
- It equips us with a very powerful tool to create 'alternative worlds'. 
- It makes us more responsible about the conscious choices we need to make in order to get an answer (and we sometimes need to make a lot of them). 
- The challenge in this game is to figure out what the parameters are for the processes that make up the natural world. 


.alert[Are these two distribution showing the same random process? ] 
&lt;img src="workshopXX-en_files/figure-html/normal_compare_theoretical_simulated-1.png" width="576" style="display: block; margin: auto;" /&gt;

???
- Like Étienne Low-Décarie said it "it eliminates the need for dice when playing Dungeons and Dragons"
```
library(TeachingDemos)
dice(rolls=3, ndice=4, sides=6, plot.it=T,
     load=rep(1, 6))
```

Are these two distribution showing the same random process?
- Yes! They are both normally distributed 
  * left: standard normal curve and 
  * right: density plot of 6 pts randomly drawn from a normal distribution

- "responsible about the conscious choices [...]": 
  * What I mean here is that when we design a simulation, we have to build in the assumptions of the models or the process we are trying to recreate. 
  * So we need to *decide* (or plan), what is going to make up the process. 
  * In the simulation, the parameters and variable are chosen by you, but they need to be plausible (from the literature, or maybe a collegue could help you on this with data that has already been collected).
- In a nutshell, you need to be *statistically aware* of what you are modeling and what you want to know. 

---
## Find patterns in random processes

- Simulations are useful to test the properties of randomly generated data 
- Since we design the simulation, we know parameters of the processes.
???
You probably learned how to use statistical models on certain types of data. 
- In order to answer your research question, you need to collect data. 
- We then have an a posteriori understanding of the parameters that make up the phenomenon we are interested in. 
- Simulations help you, 
  * bypass the requirement of data, AND
  * Let you SET the exact parameters you think are relevant for the pattern you want to characterize (perhaps based on the literature on a subject.)
  * BUT! Do not simulate data to generate FAKE data. 
--

- It is then possible to test various methods or statistical analysis to 
    1. see if they work and verify their assumptions, 
    2. do [power analysis](https://cran.r-project.org/web/packages/pwr/vignettes/pwr-vignette.html) and dry run [an experimental design](https://low-decarie.github.io/Introduction-to-simulation-for-experimental-design-using-R/#/),
    3. learn how data is generated,
    4. prepare your analysis script before data collection (or preregistration),
    5. understand and "see" what a model or equation predicts,
    6. [give you teaching material](https://serc.carleton.edu/sp/library/datasim//examples.html), 
    7. explore ideas, theories, hypothesis, the existence of a phenomenon or a slow process,
    8. etc. 
    
???
- Forest management: could be long to get data: simulations can help to take faster decision
- epidemics, road congestion
- See book Morgan, B.J.T., 1984. Elements of simulation, Boca Raton, FL: CRC Press.

---
## Find patterns in the natural world

- We might start with a research question or simply want to understand a process.
- The idea here is that perhaps we are interested in understanding how a certain process was generated (how `\(X\)` influences `\(Y\)`, which could be written as `\(X \xrightarrow{Affects} Y\)` or `\(Y \sim X\)`). 
- This is probably the most important points for simulations in biology as we are trying to understand **natural processes** (which are themselves manifestations of **random processes**)
- Therefore, being able to do simulations can be a nice addition to your research toolkit. 

???

Everything that we measure (a phenotype, the number of eggs a bird lays, the survival of individuals in a population, etc.), is caused by factors that scientists, try to make sense. 
- the question here is how these measures came to be? 
  * What process governs the value that we measured? 
  * In other words, 
     1. how is the data that I've measured distributed? and 
     2. What causes my data to be distributed that way?
- See also `power.prop.test()`

---
## Probability distributions are crucial for simulations

- When performing simulations, one has to keep in mind the
  1. **type of distribution** underlying the data of interest
  2. **parameters** of the distribution itself (mean, standard deviation, rate, degrees of freedom, etc.)
  3. **statistical model** making the relationship between the response variable and the explanatory variable. 
  4. ideas or things we *want to show or learn* with the simulation
???
- That might not apply to all sorts of simulations, but will at least cover the ones that relate to statistical models.
- It is still very useful to know the different types of distributions that random variables can take. 
- Note that a "parameter" is a quantity that is CONSTANT. A variable can take multiple values. BUT a parameter could become a variable if this is the thing we want to see how it affects our models. 
- Greek letters are often used to denote these parameters 

---
## Tips when performing simulation

.alert[Description section]. Add a 'Description' section to all your simulation scripts to introduce what the script is about. 

- You can add information how to use important arguments or a step-by-step description on how to use the script.
- You can add some references that you used to build your scrip. 

.small[
```r
# Description  ------------------------------------------------------------
#### ### ### ## #### ### ### ## #### ### ### ## 
# TITLE OF THE SCRIPT
# Created by YOUR_NAME
# 
# Why: 
# Requires:
# NOTES: 
# Reference : 
#### ### ### ## #### ### ### ## #### ### ### ## 

# Code here ... 

```
]

???
- Before the meat of the presentation, I want to give you this tip. 
- Here we present a 'header' that you could use at the beginning of EACH of you script
- Believe us, it makes it much easier when you go back to an old script if you have at least this information in the header. 


---
## Tips when performing simulation

.alert[Comments]. Be extra generous when commenting your code to describe as precisely as possible for the information that is not variable. 


.small[

```r
# Description  ------------------------------------------------------------
## This is the description section as previously presented 

# Libraries ---------------------------------------------------------------
## Here load the libraries used in the script 
library(ggplot2)

# Functions ---------------------------------------------------------------
## Add and describe the functions used in the script 
## Add more information
function.name = function(var){tmp=2+var; return(tmp)}

# Plots -------------------------------------------------------------------
## Plotting the data simulated 
## Add more information
plot(function.name(1:10)+rnorm(10))
```
]

.small[
(We used fewer comments to lighten the presentation, but please add comments.)
]

???
- Also don't forget to add sections in your script. 
  - On macOS press `Alt + Shift + K` to get the shortcut list of RStudio. 
  - See the `Cmd + Shift + R` to automatically add a section that will be visible in the document outline. 

- Time to run a simulation is important: knowing that some functions are faster in R can make a difference for a slow simulation with a fast one.

#################################
---
class: inverse, center, middle

# When could a simulation be useful? 

???
You might be wondering: 
- "OK, I somewhat understand how learning to perform simulations in R might be useful to me. 
- But WHEN should I use simulations in my work?"

---
## ALL THE TIME!!!! 

- Simulations can help you at all stages of your research from planning to publishing and beyond! 
- Whether you want to: 
1. Learn about science in general
  - learn how a mechanism work (i.e., genetic drift, natural selection), 

2. Design a particular study 
  - test a statistical method or a research question you have, 
  - play with data before your field work, 
  - simulate random point on a map to sample it, 

3. Publish an article 
  - respond to reviewers for a paper you want to publish,
  
4. etc. 


???
- The surprising answer is that it can be **used at all stages of research**. 
- Here is a short list of elements for when you could be using it

---
## Quick example: Genetic drift

- Here we have 4 plots showing genetic drift with the same simulation. 
- The only difference is the number of individuals in each population
- Your can deduce many things from the allele frequency change.

&lt;img src="workshopXX-en_files/figure-html/r_tips2-1.png" width="936" style="display: block; margin: auto;" /&gt;

???
- Here is an example of simulation that was made in R to show genetic drift without natural selection. 
- The same script, or function, can be used to generate patterns like these. 
- Each plot contains 5 populations that started with an allele which had a frequency of 50%. 
- You can see now see with the simulation that 
  * No *particular* allele will be fixated (unbiased)
  * the number of individuals seems an important factor in the fixation of alleles. The fluctuations in allele fq are largers in smaller pop. 
  * some genetic variation can be lost
  * makes populations different, even if the pop are identical at the beginning. 
  * Fixation of an allele without natural selection 

---
## Quick (?) example: Natural selection 

- How would you model natural selection?
--

- What is the type of selection? 
- What is being selected? What is that type of data (continuous [range?], discrete [counts?])? 


&lt;img src="workshopXX-en_files/figure-html/fake_fitness_functions-1.png" width="576" style="display: block; margin: auto;" /&gt;
--

- But what all of these mean? 
- What do they assume? 
- What is the data generated to make these lines and curves? 


???
- In order to build a natural selection simulator in R, we need a couple of tools. 

#################################
---
class: inverse, center, middle

# Let's build some simulation intuitions 

???
- In this section, I want to explore a graph and try to extract all of its juice
- NOTE: The TACIT idea here is to find the underlying DISTRIBUTIONS or HOW THE RANDOM VARIABLE WAS GENERATED.

---
## What do you see here? 

- Can you give *as much information* as you can within and between each graphs.


.pull-left[
&lt;img src="workshopXX-en_files/figure-html/normalX_Y1-1.png" width="360" style="display: block; margin: auto;" /&gt;
]

.pull-right[
&lt;img src="workshopXX-en_files/figure-html/normalX_Y2-1.png" width="360" style="display: block; margin: auto;" /&gt;
]

???
Here are 2 plots with an independent variable X and a response variable y. 
--

Find correlation and covariance
- The correlation of the first graph is 0.92 and the covariance is 45.74.
- The correlation of the second graph is 0.97 and the covariance is 2.04.
???
What can be seen? 
- Points, continuous x and y
- mean (dashed lines), sd (the shades)
- Correlation, covariance 
- Linear model
- Assumptions of linear model 
  - Linearity
  - Independence? Don't know
  - Normal residuals (not shown)
  - Equality of variance (see residuals)

---
## What do you see here? 

- Can you give *as much information* as you can within and between each graphs.

.pull-left[
&lt;img src="workshopXX-en_files/figure-html/normalX_Y3-1.png" width="360" style="display: block; margin: auto;" /&gt;
]

.pull-right[
&lt;img src="workshopXX-en_files/figure-html/normalX_Y4-1.png" width="360" style="display: block; margin: auto;" /&gt;
]

???
What can be seen? 
- mean (dashed lines), sd (the shades)

---
## What do you see here? 

- Can you give *as much information* as you can within and between each graphs.

.pull-left[
&lt;img src="workshopXX-en_files/figure-html/normalX_Y5-1.png" width="360" style="display: block; margin: auto;" /&gt;
]

.pull-right[
&lt;img src="workshopXX-en_files/figure-html/normalX_Y6-1.png" width="360" style="display: block; margin: auto;" /&gt;
]

???
What can be seen? 
- mean (dashed lines), sd (the shades)


---
## What do you see here? 

- Can you give *as much information* as you can within and between each graphs.

.pull-left[
&lt;img src="workshopXX-en_files/figure-html/normalX_Y7-1.png" width="360" style="display: block; margin: auto;" /&gt;
]

.pull-right[
&lt;img src="workshopXX-en_files/figure-html/normalX_Y8-1.png" width="360" style="display: block; margin: auto;" /&gt;
]

???
What can be seen? 
- Correlation, covariance 
- Linear model
- Assumptions of linear model 
  - Linearity
  - Independence? Don't know
  - Normal residuals (not shown)
  - Equality of variance (see residuals)

---
## What do you see here? 

- Can you give *as much information* as you can within and between each graphs.

.pull-left[
&lt;img src="workshopXX-en_files/figure-html/normalX_Y9-1.png" width="360" style="display: block; margin: auto;" /&gt;
]

.pull-right[
&lt;img src="workshopXX-en_files/figure-html/normalX_Y10-1.png" width="360" style="display: block; margin: auto;" /&gt;
]

???
What can be seen? 
- Distribution of x and y to better understand what could have PRODUCED these variables. 


#################################
---
class: inverse, center, middle

# What you need to know before doing simulations? 

???
We'll explore the functions that you can use in R to help you develop simulations and play with random number generators. 

---
## Functions useful in simulations (RNG)

- When performing simulations, you will have to play with (pseudo-)randomly generated numbers from a random number generator (RNG). 
- This is a challenge if we want to replicate the analysis we are performing. 
- `R` has the function `set.seed()` that help us to play with the RNG.

The example below uses a RNG to extract numerical value between 1 and 10



```r
runif(n = 1, min = 1, max = 10) # Gives a random number between 1 and 10
# [1] 3.588198
runif(n = 1, min = 1, max = 10) # RNG wasn't reset, different answer (see above)
# [1] 8.094746
runif(n = 1, min = 1, max = 10) # Different again... 
# [1] 4.680792

set.seed(42); runif(n = 1, min = 1, max = 10) # This sets the RNG 
# [1] 9.233254
set.seed(42); runif(n = 1, min = 1, max = 10) # The exact same number 
# [1] 9.233254
```

???
- To go deeper: The actual pseudorandom number generator (PRNG) which is default in R is called the Mersenne Twister (see `??set.seed`) https://en.wikipedia.org/wiki/Mersenne_Twister
- `set.seed(seed, kind = "Mersenne-Twister")`

---
## Functions useful in simulations (sample)

- The function `sample()` randomly picks a value from a vector (i.e., random sampling).

The example below uses a RNG to extract numerical value between 1 and 10



```r
set.seed(12) # Set the RNG 
v.1.10 = 1:10 # Make a vector from 1 to 10 
# Randomly pick 1 (size) value from the vector (x), without replacement 
sample(x = v.1.10, size = 1, replace = FALSE) 
# [1] 2
```

- The values don't have to be numerical: they could be characters or factors


```r
set.seed(3) # Set the RNG 
# Randomly pick 5 (size) letters from the vector (x), without replacement 
sample(x = LETTERS, size = 5, replace = FALSE) 
# [1] "E" "L" "G" "D" "H"
sample(x = as.factor(month.abb), size = 5, replace = FALSE) 
# [1] Nov Aug Apr Jul Dec
# Levels: Apr Aug Dec Feb Jan Jul Jun Mar May Nov Oct Sep
```

???
- Very versatile 

---
## Functions useful in simulations (sample)

&lt;!-- SEE http://faculty.washington.edu/kenrice/sisg/sisg-lie11-05.pdf --&gt;

- The function `sample()` can actually be used in order to do permutations 
- Let's say we have a data frame with 2 columns 

.verysmall[
.pull-left[



```r
set.seed(123)
n = 40; col = viridis::viridis(n = n)
x = 1:n ; y = 2+.5*x + rnorm(n, sd=7)
df.xy = data.frame(x,y, col )
```
]

.pull-right[

```r
set.seed(321)
df.xy$x.s = sample(df.xy$x)
df.xy$y.s = sample(df.xy$y)
# We break up the link of X and Y 
```
]


```r
par(mfrow=c(1,3), mar=c(4,4,1,1), cex = 1.2)
plot(y~x,  col=col, data=df.xy, pch=19);abline(lm(y~x,  data=df.xy)) # Original 
plot(y~x.s,col=col, data=df.xy, pch=19);abline(lm(y~x.s,data=df.xy)) # Permutated 
plot(y.s~x.s,col=col, data=df.xy, pch=19);abline(lm(y~x.s,data=df.xy)) # Permutated 
```

&lt;img src="workshopXX-en_files/figure-html/permutations_plot-1.png" width="648" style="display: block; margin: auto;" /&gt;
]

???
- for permutations 
  * 1st plot: not permutated
  * 2nd plot: x permutated
  * 3rd plot: x and y permutated
  
&lt;!-- --- --&gt;
&lt;!-- ## Functions useful in simulations (sample) --&gt;
&lt;!-- - Keep in mind that permutations are valid if the null hypothesis tested is that there is *no association* between the variables studied.
http://faculty.washington.edu/kenrice/sisg/sisg-lie11-05.pdf
--&gt;

---
## Functions useful in simulations (sample)

- You can also simulate random dates in a time interval. 


```r
set.seed(42)
# GEt a sequence of dates
datae.seq = seq(from = as.Date('2010/01/01'), 
                to = as.Date('2022/01/01'), 
                by = "day")
# Look at beginning and end of sequence 
head(datae.seq, 4); tail(datae.seq, 4)
# [1] "2010-01-01" "2010-01-02" "2010-01-03" "2010-01-04"
# [1] "2021-12-29" "2021-12-30" "2021-12-31" "2022-01-01"
# Get only 5 elements of the generated sequence 
sample(datae.seq, 5)
# [1] "2017-02-21" "2021-02-20" "2016-06-26" "2013-01-02" "2013-06-05"
```
???
- for dates and time as well. 

---
## Functions useful in simulations (sample)

- You can also generate a die and randomly toss it. 
- You can change the probability of landing on one face more than the other. 

.small[

```r
set.seed(50)
p_dice = c(1,1,1,1,1,5) # Here we have twice the change of landing on 6
                        # Same as writing p_dice/sum(p_dice) or the prob.
nb.tosses = 100
die_results &lt;- sample(x = 1:6, # or seq(from = 1, to=6, by=1)
                      size = nb.tosses,
                      replace = T, prob = p_dice) 
barplot(table(die_results)) # table(die_results)/nb.tosses
```

&lt;img src="workshopXX-en_files/figure-html/unnamed-chunk-2-1.png" width="576" style="display: block; margin: auto;" /&gt;
]

???
- simulate a die and change the probability of landing on a particular face 


---
## Functions useful in simulations (rep)

- You will not only see numerical values in simulations. Characters (or factors) can be generated in R easily. 
- The `rep()` function can help you with this


```r
(let4first = LETTERS[1:4])
# [1] "A" "B" "C" "D"
rep(let4first, times = 2) # Repeat the WHOLE sequence twice 
# [1] "A" "B" "C" "D" "A" "B" "C" "D"
rep(let4first, each = 2) # Repeat each element twice 
# [1] "A" "A" "B" "B" "C" "C" "D" "D"
# Set the number of repeat for each element in the vector 
rep(let4first, times = c(1,2,3,6))
#  [1] "A" "B" "B" "C" "C" "C" "D" "D" "D" "D" "D" "D"
# Complete replication: replicate each element twice and do this three times 
rep(let4first, each = 2, times = 3)
#  [1] "A" "A" "B" "B" "C" "C" "D" "D" "A" "A" "B" "B" "C" "C" "D" "D" "A" "A" "B"
# [20] "B" "C" "C" "D" "D"
rep(let4first, length.out = 6) # Repeat the vector until you hit the maximum length
# [1] "A" "B" "C" "D" "A" "B"
```

???
- It is useful to make a vector with multiple copies of some values
- rather than doing it by hand, you can use `rep()`


---
## Functions useful in simulations (gl)

- There are functions in `R` that can help you generate factor levels (with the `gl` function). 


```r
number.of.levels = 2
number.of.replicates = 8
labels.for.the.factors = c("Control", "Treat")
## First control, then treatment:
gl(n = number.of.levels, k = number.of.replicates, labels = labels.for.the.factors)
#  [1] Control Control Control Control Control Control Control Control Treat  
# [10] Treat   Treat   Treat   Treat   Treat   Treat   Treat  
# Levels: Control Treat
## 20 alternating 1s and 2s
gl(n = 2, k = 1, length = 20)
#  [1] 1 2 1 2 1 2 1 2 1 2 1 2 1 2 1 2 1 2 1 2
# Levels: 1 2
## alternating pairs of 1s and 2s
gl(n = 2, k = 2, length = 20)
#  [1] 1 1 2 2 1 1 2 2 1 1 2 2 1 1 2 2 1 1 2 2
# Levels: 1 2
```

???
- rather than using rep and then factors, you can use the Generate Factor Levels `gl()` function

---
## Functions useful in simulations

- In some cases, you want to replicate data, for example when you want to generate multiple populations at once with the same parameters
- `replicate()` can be used instead of a for loop and make simulations faster. 


```r
set.seed(1234)
data.replicated = replicate(n = 2,expr = data.frame(gr = rep(LETTERS[1:3],each = 2),
                                                    y = rnorm(6)), 
                            simplify = FALSE)
```
.pull-left[

```r
data.replicated[[1]]
#   gr          y
# 1  A -1.2070657
# 2  A  0.2774292
# 3  B  1.0844412
# 4  B -2.3456977
# 5  C  0.4291247
# 6  C  0.5060559
```
]

.pull-right[

```r
data.replicated[[2]]
#   gr          y
# 1  A -0.5747400
# 2  A -0.5466319
# 3  B -0.5644520
# 4  B -0.8900378
# 5  C -0.4771927
# 6  C -0.9983864
```
]

???
- Not entirely sure of that one. 
- In this case, it generates 2 simulated (different) dataframes which can be used in a simulation to get some replication all at once. I think this comes handy with the `apply`, `do.call`, etc. 

---
## Challenge 1 ![:cube]()

Select randomly 4 values out of a vector of numbers ranging from 1 to 10 : 1. without replacement and 2. with replacement. 

Use `sample()` to perform this task. 


```r
sample()
```


---
## Challenge 1 - Solution ![:cube]()

Select randomly 4 values out of a vector of numbers ranging from 1 to 10 : 1. without replacement and 2. with replacement.


```r
set.seed(12345) # Sets the random number generator to a fix value
vec.1.10 = 1:10 # Make the vector to choose from 
sample(x = vec.1.10, size = 4, replace = FALSE) # Sample 4 nb without replacement
# [1] 3 8 2 5
sample(x = vec.1.10, size = 4, replace = TRUE) # Sample 4 nb with replacement
# [1] 8 2 6 6
```

As you can see in the last example, there are 2 "6"s, since each time a random number was picked, all numbers could be randomly chosen. 

1000 draws with replacement from 0 or 1, with equal probability for each. 

```r
set.seed(123); table(sample(x = 0:1, size = 1000, replace = T,prob = c(.5,.5)))
# 
#   0   1 
# 493 507
```

???
- I added a bonus here *1000 draws with replacement from 0 or 1*


---
## Challenge 2 ![:cube]()

Create a data frame with variable 
- x ranging from 1 to 10, 
- `\(y = 2+ 3 * x\)` and 
- a grouping factor (gr) with one group with the 5 smallest values and another group with the 5 largest values. 

---
## Challenge 2 - Solution ![:cube]()

```r
x = 1:10
y = 2 + 3 * x
gr = rep(letters[1:2],each = 5)
linear.df = data.frame(x,y,gr)
plot(y~x, col = as.factor(gr), data = linear.df, pch = 19, cex = 1.5)
```

&lt;img src="workshopXX-en_files/figure-html/linear_no_error-1.png" width="576" style="display: block; margin: auto;" /&gt;


???

#################################
---
class: inverse, center, middle

# Let's explore what underlies some statistical processes

???
- This is where the statistical background is important. 
- we will revise some concepts to make sure we can continue on programming simulations

---
## Probability and odds 

- When describing the "chance" or "likelihood" of an outcome, you can use a probability or odds
- They are related: odds of an event is `\(odds = p/(1-p)\)`  or `\(odds = \frac{\text{probability of success}}{\text{probability of failure}}\)`
- Why bother using probability or odds? 
--

- probability ranges between 0 and 1, but odds range from 0 to `\(\infty\)`. 

.pull-left[
&lt;img src="workshopXX-en_files/figure-html/unnamed-chunk-4-1.png" width="338.4" style="display: block; margin: auto;" /&gt;
]

.pull.right[
&lt;img src="workshopXX-en_files/figure-html/unnamed-chunk-5-1.png" width="338.4" style="display: block; margin: auto;" /&gt;
]

???
- See how log(1/6) = -1.791759 and log(6/1) = 1.791759
- Seriously, check this post: https://towardsdatascience.com/probability-vs-odds-f47fbc6789f4. It is amazing! 
- And check this video for odds: https://www.youtube.com/watch?v=ARfXDSkQf1Y&amp;ab_channel=StatQuestwithJoshStarmer


---
## Probability and odds 

- Log (natural log, or ln) odds can range from `\(-\infty\)` to `\(\infty\)` and makes the odds symmetric. It is calculated as `\(ln (odds) = ln(p/(1-p))\)` (i.e., the logit function).
- This will become handy when building a logistic model as this is the 'logit' transformation. 

&lt;img src="workshopXX-en_files/figure-html/unnamed-chunk-6-1.png" width="338.4" style="display: block; margin: auto;" /&gt;


&lt;!-- .alert[Careful!] This is not the same as saying `\(success/(success + failure)\)`. This last calculation is the probability --&gt;
???
- And this video for odds ratio: [Odds Ratios and Log(Odds Ratios), Clearly Explained!!!](https://www.youtube.com/watch?v=8nm0G-1uJzA&amp;ab_channel=StatQuestwithJoshStarmer)
  - The odds ratio and the log odds ratio is a bit like the R-squared: it indicates a relationship between 2 things (it's an effect size, or how one is a good predictor of the other (if the value is large))

- Also, the Geometric mean (which is the mean of logs) is BETTER when using log-based data (qPCR, where the data doubles each round) and is less sensitive to outliers (which the arithmetic mean is actually VERY sensitive)


```r
raw.num = c(1,2,8) # numbers NOT on the log scale (simulate a DOUBLING process like qPCR)
mean(raw.num) # Arithmetic average (which would be WRONG)
# [1] 3.666667
mean.log = mean(log(raw.num)/log(2)) # Taking the mean of the log (this is the geometric mean)
mean.log # Mean, on the log scale, for the raw numbers transformed: log(raw.num)
# [1] 1.333333
2^mean.log
# [1] 2.519842
# gm = exp(mean(log(raw.num))) # calculating the geometric mean 
gm = 2^(mean(log(raw.num)/log(2))) # Here we calculate the geometric mean with a base 2
# Now compare the geometric mean of the raw numbers with the mean of the log, transformed back into the raw number scale 
gm
# [1] 2.519842
2^mean.log
# [1] 2.519842
```


---
## Probability and odds

- Example: a horse runs 100 races and wins 20 of them.  What is the *odds* of winning?
- `\(odds = p/(1-p)\)`  or `\(odds = \frac{\text{probability of success}}{\text{probability of failure}}\)`


```r
totl.nb.events = 100; successes = 20
odds.favor = (successes)/(totl.nb.events-successes) # Odds in favor of the event (here 1 in 4)
odds.against = (totl.nb.events-successes)/(successes) # Odds in against of the event (here 4 to 1)
```

&lt;!-- In the calculation above, I'm NOT using the probability of successes as the division by 100 (to make it a probability) would cancel out and make it harder to get a quick intuition from the toy example --&gt;

The horse wins 1 race to 4 fails (odds of winning = 0.25).
So, for 5 races it will win 1 and loose 4. 

&lt;img src="workshopXX-en_files/figure-html/unnamed-chunk-8-1.png" width="338.4" style="display: block; margin: auto;" /&gt;


Conversely, the horse  odds of failing is 4:1 (read 4 *to* 1). 
The horse is 4 times more *likely* to fail than to succeed in a race.

???
- If the odds = 1: 
  - it means that we are as likely to win, as we are likely to lose 

- Another example: 
  - for a 6-sided die, 
    * the odds of rolling a 6 is 1:5 (one TO 5)
    * the odds of rolling a 5 OR a 6 is 2:4 (2 TO 4)

- in French, Odds is the "cote" (https://fr.wikipedia.org/wiki/Cote_(probabilités))
- Also, be careful. Probability and likelihood IS NOT THE SAME

---
## Probability and odds 

- Example: a horse runs 100 races and wins 20 of them. What is the *probability* of winning?
- `\(p = odds/(1+odds)\)`


```r
totl.nb.events = 100; successes = 20
probability.favor = (successes)/(totl.nb.events) # probability of the event (here 20%)
probability.favor # fractions(probability.favor)
# [1] 0.2
```
So the horse *wins* 0.2 or 20% of the times. 
--

- You can see that `\(odds = p/(1-p) = (.2)/(1-.2) = .2/.8 = 1/4\)`
--

- So the probability of a particular outcome (say getting a *head* for a coin) for a particular event (tossing a coin) is the number of a particular outcome divided by the total number of results from the event. 

---
## Probability and odds 


```r
log_odds = seq(from = -5, to = 5, by = .25)
odds = exp(log_odds)
inv.logit &lt;- function(x) {exp(x)/(1 + exp(x))}
p = inv.logit(log_odds)
p2 = odds/(1 + odds)
# Probability of failure (1-p)
q = 1-p
# store log_odds and y in data frame for use with ggplot
d = data.frame(log_odds, odds, p, p2, q) 
head(d, 4)
#   log_odds        odds           p          p2         q
# 1    -5.00 0.006737947 0.006692851 0.006692851 0.9933071
# 2    -4.75 0.008651695 0.008577485 0.008577485 0.9914225
# 3    -4.50 0.011108997 0.010986943 0.010986943 0.9890131
# 4    -4.25 0.014264234 0.014063627 0.014063627 0.9859364
```

&lt;!-- Reference 
https://www.montana.edu/rotella/documents/502/Prob_odds_log-odds.pdf 
https://stats.oarc.ucla.edu/other/mult-pkg/faq/general/faq-how-do-i-interpret-odds-ratios-in-logistic-regression/
https://www.statisticshowto.com/probability-and-statistics/probability-main-index/odds-ratio/
--&gt;
&lt;!-- Good example on odds ration 
https://sphweb.bumc.bu.edu/otlt/mph-modules/bs/bs704_confidence_intervals/BS704_Confidence_Intervals10.html --&gt;

---
## Probability and odds 

.tiny[

```r
par(mfrow = c(1,3), mar =c(4,4,.5,.5), cex = 1.2)
plot(d$odds~d$log_odds, type = "l", ylab = "Odds", xlab = "Ln odds", lwd = 3) ; abline(v = 0, lty = 3)
plot(d$p~d$odds, type = "l", ylab = "p", xlab = "Odds", lwd = 3) ; abline(h = .5, v = 0, lty = 3)
plot(d$p~d$log_odds, type = "l", ylab = "p", xlab = "Ln odds", lwd = 3) ; abline(h = .5, v = 0, lty = 3)
```

&lt;img src="workshopXX-en_files/figure-html/odds_prob_plot-1.png" width="864" style="display: block; margin: auto;" /&gt;
]

---
## Probability and odds 

- What is the difference between a probability and a p-value?
--

- A p-value consists of 3 parts. They are the :
  1. *Probability* that random chance generated the data (or would give the result we got) **PLUS**
  2. *Probability* of observing something else that is equally rare **PLUS**
  3. *Probability* of observing something else that is rarer (or more extreme).
--

- **NOTE:** the probability of an event occurring *is not the same* as the p-value, since the p-value is the sum of 3 parts.
  - We add part 2. and 3. because it gives us information about the *specialness* of this rare event. 
  - In other words, we want to know if the event we measured is statistically special or, as you might have heard it, statistically significant!

???
- See [StatQuest: P Values, clearly explained](https://www.youtube.com/watch?v=5Z9OIYA8He8&amp;ab_channel=StatQuestwithJoshStarmer), [How to calculate p-values ](https://www.youtube.com/watch?v=JQc3yx0-Q9E&amp;t=0s&amp;ab_channel=StatQuestwithJoshStarmer) and [p-values: What they are and how to interpret them](https://www.youtube.com/watch?v=vemZtEM63GY&amp;ab_channel=StatQuestwithJoshStarmer)

---
## Why statistical distributions matter? 

- To help us in performing simulations, we need to get some data (random variables, X) which numerical values are outcomes of random processes. 
- The idea here is to find the distribution underlying a random process which generate a particular phenomenon in nature and use it to output values to be used in your simulation.
- After generating the data, you can analyze it and see how it can be manipulated. 

---
## What is a random variable? 

- Imagine a number is presented to you like so:




```
# [1] 13.9
```
--

- From which distribution is this variable coming from?
--

- Would be helpful to have more information 

???

This is an exercice to let the attendees feel what are random variables and why staitstical distributions are interesting. 

---
## What is a random variable? 

- Now imagine you are able to get more numbers from the same process:


```
# [1] 13.9 14.5 18.1 15.1 15.3
```
--

- This is more informative. We can take the mean and the standard deviation of these numbers: mean = 15.4, sd = 1.6. 
--

- Again, from which distribution is this variable coming from?
--

- We could plot the result to get a feel of the distribution: 

.pull-right2[

&lt;img src="workshopXX-en_files/figure-html/feel_random_var_hist_5-1.png" width="432" style="display: block; margin: auto;" /&gt;

]

--

- still hard to definitly tell 


---
## What is a random variable? 

- What if you get even more numbers (say 100) from the same random process:

.tiny[

```
#   [1] 13.9 14.5 18.1 15.1 15.3 18.4 15.9 12.5 13.6 14.1 17.4 15.7 15.8 15.2 13.9
#  [16] 18.6 16.0 11.1 16.4 14.1 12.9 14.6 12.9 13.5 13.7 11.6 16.7 15.3 12.7 17.5
#  [31] 15.9 14.4 16.8 16.8 16.6 16.4 16.1 14.9 14.4 14.2 13.6 14.6 12.5 19.3 17.4
#  [46] 12.8 14.2 14.1 16.6 14.8 15.5 14.9 14.9 17.7 14.5 18.0 11.9 16.2 15.2 15.4
#  [61] 15.8 14.0 14.3 13.0 12.9 15.6 15.9 15.1 16.8 19.1 14.0 10.4 17.0 13.6 13.6
#  [76] 17.1 14.4 12.6 15.4 14.7 15.0 15.8 14.3 16.3 14.6 15.7 17.2 15.9 14.3 17.3
#  [91] 17.0 16.1 15.5 13.7 17.7 13.8 19.4 18.1 14.5 12.9
```
]

--

- The mean = 15.4, sd = 1.6. 
- Check the graph

.pull-right2[

&lt;img src="workshopXX-en_files/figure-html/feel_random_var_hist_100-1.png" width="432" style="display: block; margin: auto;" /&gt;
]

--

- That looks normal! 

---
## Functions for various distributions

- In `R`, there is a convention on how to call different distributions. 

| Type | Definition | Comment |
| ---- | ---------- | ------- |
| p	   | *probability* (*cumulative* distribution function or CDF; gives you the probability for a given quantile) | gives the probability that a random variable takes in a certain *range of values* for that random variable |
| q	   | *quantile* (inverse of CDF; gives you the x value for a given probability) | gives the position on the x axis where the area under the curve is a certain probability |
| d	   | *density* density function (probability density function or PDF) | gives probability that a random variables have for a specific value if the random variable  |
| r	   | *random* random variable coming from a certain distribution | We use this to generate random numbers from a specific distribution |

- In the next slide, you'll see some built in functions to use these distributions 

---
## Functions for various distributions

.verysmall[
| Distribution | Probability | Quantile | Density | Random | PDF or PMF |
| ------------ | ----------- | -------- | ------- | ------ | ---------- |
| Beta	       | `pbeta`	 | `qbeta`	| `dbeta`	| `rbeta` | |
| Binomial	   | `pbinom`	 | `qbinom`	| `dbinom`	| `rbinom` | `\(f(x) = {n \choose x} p^x(1-p)^{n-x}\)` |
| Cauchy       | `pcauchy` | `qcauchy` | `dcauchy`	| `rcauchy` ||
| Chi-Square	 | `pchisq`	 | `qchisq`	 | `dchisq`	| `rchisq` | |
| Exponential	 | `pexp`    | `qexp`	 | `dexp`	| `rexp` | `\(f(x) = \lambda e^{-\lambda x}\)` (when `\(x \geq 0\)`, otherwise, 0) |
| F	           | `pf`	     | `qf`	| `df`	| `rf` | |
| Gamma	       | `pgamma`	 | `qgamma`	| `dgamma`	| `rgamma` | |
| Geometric	   | `pgeom`	 | `qgeom`	| `dgeom`	| `rgeom` | `\(f(x)=p(1-p)^{x-1}\)` |
| Hypergeometric | `phyper`	| `qhyper`	| `dhyper`	| `rhyper` ||
| Logistic	     | `plogis`	| `qlogis`	| `dlogis`	| `rlogis` | |
| Log Normal	   | `plnorm`	| `qlnorm`	| `dlnorm`	| `rlnorm` | |
| Negative Binomial	| `pnbinom`	| `qnbinom`	| `dnbinom`	| `rnbinom` | |
| Normal	     | `pnorm`	| `qnorm`	| `dnorm`	| `rnorm` | `\(f(x)={\frac{1}{\sigma\sqrt{2\pi}}}e^{-{\frac{1}{2}}\Bigl(\frac {x-\mu}{\sigma}\Bigr)^2}\)` |
| Poisson	     | `ppois`	| `qpois`	| `dpois`	| `rpois` | `\(f(x) = e^{-\lambda}\frac{\lambda^x}{x!}\)` |
| Student t	   | `pt`	| `qt`	| `dt`	| `rt` ||
| Studentized Range | `ptukey`	| `qtukey`	| `dtukey`	| `rtukey` | |
| Uniform	     | `punif`	| `qunif`	| `dunif`	| `runif` | `\(f(x) = \frac{1}{b-a}\)` for `\(x \in [a,b]\)`, `\(0\)` otherwise |
| Weibull	     | `pweibull`	| `qweibull`	| `dweibull`	| `rweibull` ||
| Wilcoxon Rank Sum Statistic	   | `pwilcox`	| `qwilcox`	| `dwilcox`	| `rwilcox` ||
| Wilcoxon Signed Rank Statistic | `psignrank`	| `qsignrank`	| `dsignrank`	| `rsignrank` ||
]

---
## Functions for various distributions

.verysmall[
| Distribution | Probability | Quantile | Density | Random | PDF or PMF |
| ------------ | ----------- | -------- | ------- | ------ | ---------- |
| Beta	       | `pbeta`	 | `qbeta`	| `dbeta`	| `rbeta` ||
| &lt;span style="background-color: #FFFF00"&gt;**Binomial**&lt;/span&gt;	   | `pbinom`	 | `qbinom`	| `dbinom`	| `rbinom` | `\(f(x) = {n \choose x} p^x(1-p)^{n-x}\)` |
| Cauchy       | `pcauchy` | `qcauchy` | `dcauchy`	| `rcauchy` ||
| &lt;span style="background-color: #FFFF00"&gt;**Chi-Square**&lt;/span&gt;	 | `pchisq`	 | `qchisq`	 | `dchisq`	| `rchisq` ||
| &lt;span style="background-color: #FFFF00"&gt;**Exponential**&lt;/span&gt;	 | `pexp`    | `qexp`	 | `dexp`	| `rexp` | `\(f(x) = \lambda e^{-\lambda x}\)` (when `\(x \geq 0\)`, otherwise, 0) |
| &lt;span style="background-color: #FFFF00"&gt;**F**&lt;/span&gt;	           | `pf`	     | `qf`	| `df`	| `rf` ||
| Gamma	       | `pgamma`	 | `qgamma`	| `dgamma`	| `rgamma` ||
| Geometric	   | `pgeom`	 | `qgeom`	| `dgeom`	| `rgeom` | `\(f(x)=p(1-p)^{x-1}\)` |
| Hypergeometric | `phyper`	| `qhyper`	| `dhyper`	| `rhyper` ||
| &lt;span style="background-color: #FFFF00"&gt;**Logistic**&lt;/span&gt;	     | `plogis`	| `qlogis`	| `dlogis`	| `rlogis` ||
| &lt;span style="background-color: #FFFF00"&gt;**Log Normal**&lt;/span&gt;	   | `plnorm`	| `qlnorm`	| `dlnorm`	| `rlnorm` ||
| Negative Binomial	| `pnbinom`	| `qnbinom`	| `dnbinom`	| `rnbinom` ||
| &lt;span style="background-color: #FFFF00"&gt;**Normal**&lt;/span&gt;	     | `pnorm`	| `qnorm`	| `dnorm`	| `rnorm` | `\(f(x)={\frac{1}{\sigma\sqrt{2\pi}}}e^{-{\frac{1}{2}}\Bigl(\frac {x-\mu}{\sigma}\Bigr)^2}\)` |
| &lt;span style="background-color: #FFFF00"&gt;**Poisson**&lt;/span&gt;	     | `ppois`	| `qpois`	| `dpois`	| `rpois` | `\(f(x) = e^{-\lambda}\frac{\lambda^x}{x!}\)` |
| &lt;span style="background-color: #FFFF00"&gt;**Student t**&lt;/span&gt;	   | `pt`	| `qt`	| `dt`	| `rt` ||
| Studentized Range | `ptukey`	| `qtukey`	| `dtukey`	| `rtukey` ||
| &lt;span style="background-color: #FFFF00"&gt;**Uniform**&lt;/span&gt;	     | `punif`	| `qunif`	| `dunif`	| `runif` | `\(f(x) = \frac{1}{b-a}\)` for `\(x \in [a,b]\)`, `\(0\)` otherwise |
| Weibull	     | `pweibull`	| `qweibull`	| `dweibull`	| `rweibull` ||
| Wilcoxon Rank Sum Statistic	   | `pwilcox`	| `qwilcox`	| `dwilcox`	| `rwilcox` ||
| Wilcoxon Signed Rank Statistic | `psignrank`	| `qsignrank`	| `dsignrank`	| `rsignrank` ||
]

&lt;!-- 
PDF for continuous 
PMF for discrete 
# See also how to convert them from one another
https://en.wikipedia.org/wiki/Relationships_among_probability_distributions
--&gt;

---
## Common statistical distributions
&lt;!-- ![:scale 90%](images/distributions_all.png) --&gt;

&lt;img src="workshopXX-en_files/figure-html/Statistical_dist-1.png" width="750" height="600" style="display: block; margin: auto;" /&gt;

---
## Selecting the right distribution

- See "Figure 6A.15 from Probabilistic approaches to risk by Aswath Damodaran"

&lt;!-- ![:scale 100%](images/probabilistic.png) --&gt;
![:scale 100%](images/Prob.distrib.choices.png)


---
## Summary of important distribution 

.small[
- Distributions of random variables comes in 2 flavors: 
1. Discrete (only *countable* number of *distinct* values, e.g., 0, 5, 19,...)
2. Continuous (infinite number of possible values, e.g., 0.1,0.62113,...)
]



.tiny[
| Distribution | Definition | Type of data | Example |
| ------------ | ---------- | ------------ | ------------ |
| Binomial	   | probability of *n* repeated *yes/no* experiments with probability *p* of *success* | discrete; yes/no; 0/1 | Number of heads in a row after 3 tosses |
| Poisson	     | probability of a number of cases happening in a defined range of time or space given a constant rate | discrete; | ticks of a radiation counter, DNA mutation, colonization of animals on remote islands (something per unit time or space for example) |
| Logistic	   | probability of an individual to be in one class or another | continuous; | alive or not, yes or no, diseased or not, etc. |
| Normal	     | probability function which informs how values are distributed, centered on a mean, being symmetric, and having a specific variance | continuous; | height of individuals, beak length, shoe size, time required to run a kilometer, the data shows a 'bell-shaped' curve |
| Uniform	     | probability of all outcomes are equal | continuous; | angle between 0-360 | 
| Exponential	 | | continuous;| Used in maximum likelihood, Generating genealogies, continuous rate |
&lt;!-- | Student t	   |            | continuous; | | --&gt;
&lt;!-- | Geometric	   |            | continuous; | Gene conversion (recombination) | --&gt;
&lt;!-- | Beta	       |            |              | --&gt;
&lt;!-- | Chi-Square	 |            |              | --&gt;
&lt;!-- | F	           |            |              | --&gt;

- see [Probability Cheat Sheet](https://web.cs.elte.hu/~mesti/valszam/kepletek) for more details
]

---
## The same distribution? 

.alert[Careful!]

- Some distribution are similar but don't show the same process 

&lt;img src="workshopXX-en_files/figure-html/similar_plot_from_different_distributions-1.png" width="504" style="display: block; margin: auto;" /&gt;


---
## Other distributions with packages

- Other packages have other distributions. See the [CRAN Task View: Probability Distributions](https://cran.r-project.org/web/views/Distributions.html) if you need a funky distribution that doesn't come in `base R`.
- Some distributions can be recreated in R even if they haven't their own functino name (e.g., Bernoulli). 
- Bernoulli distribution can be simulated with a binomial distribution with `size = 1`.

.tiny[
&lt;img src="workshopXX-en_files/figure-html/equivalence_between_distributions_Bern_binom-1.png" width="504" style="display: block; margin: auto;" /&gt;
]





---
## Poisson distribution

- Poisson distribution assume that 
  - the probability of some event is small over a short period of time (or area), but that there are a lot of events (as there would be a lot of time passing, or a lot of space) in which the process could happen 
  - The number of event reported is independent from one time point to the other. 
  - In summary, it models *rare* events that happen at a certain *rate*. In other words, there are many opportunities to succeed (n is large), but the probability of success (for each trial) is small

`$$f(x) = e^{-\lambda}\frac{\lambda^x}{x!}$$` where `\(\mu = \sigma = \lambda\)`


---
## Poisson distribution vs binomial

- Difference with Binomial :
  - Number of events in Poisson can be infinite (so n tends towards infinity and the probability of each event approaches 0)
  - Number of trials is finite ( `\(n\)` ) in Binomial distribution 

&lt;img src="workshopXX-en_files/figure-html/equivalence_between_distributions_Binom_poisson-1.png" width="432" style="display: block; margin: auto;" /&gt;

```
# [1] 0.6766764
# [1] 0.6766764
```


---
## Poisson distribution

- Examples (something per unit time or space for example): 
  - Nb of deaths due to a disease (e.g., SARS-CoV 2) for a period of time 
  - Ticks of a radiation counter (Geiger counter which detects and measures ionizing radiation)
  - DNA mutation
  - Colonization of animals on remote islands
  - Mutation models
  - Recombination models 
  - Bacterial colony growing on an agar plate 

&lt;!-- Examples taken from
https://ani.stat.fsu.edu/~debdeep/p4_s14.pdf
--&gt;

&lt;img src="workshopXX-en_files/figure-html/petri_dish_bacteria_colonny-1.png" width="360" style="display: block; margin: auto;" /&gt;

&lt;!-- Useful resource --&gt;
&lt;!-- Poisson definition see https://ani.stat.fsu.edu/~debdeep/p4_s14.pdf --&gt;
&lt;!-- Poisson example see https://www.pnas.org/content/115/37/9270 --&gt;

---
## Binomial distribution

- Binomial distribution assume that 
- Each *independent* trial is either a success (1, T, Yes) or failure (0, F, No)
* Note that a 'success' or 'failure' is a *contrast* between 2 elements. 
* These could be 'heads/tail', 'girl/boy', 'diseased/not', etc.  
- There is a *fixed* number of trials and the probability on each trial is constant.

&lt;!-- , 'cancer cell death/alive' --&gt;


$$f(x) = {n \choose x} p^x(1-p)^{n-x} $$
.tiny[
where 
- `\(p\)` is the probability of success in a single trial (therefore, `\((1-p)\)` is the probability of failure, sometimes seen as `\(q\)`)
- `\(n\)` is the number of trials 
- `\(x\)` is the number of trials that are successes
- `\({n \choose x}\)` is the number of combinations (which is equal to `\(\frac{n!}{r!(n-r)!}\)`) where the order doesn't matter. 
]

---
## Binomial distribution

- Examples 
  - Number of heads in a row after 3 tosses
  - Genetic drift 
  - Wright-Fisher model 

&lt;!-- Examples taken from 
http://www.biology.arizona.edu/biomath/tutorials/polynomial/applications/BinomialEx3.html
https://www.zoology.ubc.ca/~bio300b/binomialnotes.html
https://sphweb.bumc.bu.edu/otlt/mph-modules/bs/bs704_probability/bs704_probability7.html
--&gt;

---
## Binomial distribution

- Examples 
  - probability of having a certain number of offspring given the genotypes of the parents 
  - for example, if a male is AA and female Aa, what is the probability that 6 of their 7 offspring has Aa
  - Find the probability of getting Aa from AA x Aa [1/2 AA and 1/2 Aa], 
  - then out of n=7 offspring from that couple, 6 are of genotype Aa, `\(P(6|7) = {7 \choose 6} (1/2)^6(1-1/2)^{7-6} = 7/128\)`)
  - Probability of a certain number of your children being daughters 
  - if you have 5 children, these are independent trials, so n = 5
  - "success" is the probability that the child is a daughter (p = 1/2)
  - `\(P(3|5) = {5 \choose 3} (1/2)^3(1-1/2)^{5-3} = 5/16\)`


---
## Normal distribution (Gaussian)

- This distribution is the most important to know (see Central limit theorem). 
- It has interesting properties (mean: center, sd: spread, bell-shaped, symmetry, etc.)
- Can approximate other distribution (might require data transformation)
- The normal distribution with mean = 0 and sd = 1 is called the **Standard Normal**

The general form of the probability density function for a normal distribution is 

`$$f(x)={\frac{1}{\sigma\sqrt{2\pi}}}e^{-{\frac{1}{2}}\Bigl(\frac {x-\mu}{\sigma}\Bigr)^2}$$`

The important part here is the mean `\(\mu\)` and the standard deviation `\(\sigma\)`. Usually summarized as `\(N(\mu, \sigma)\)`

&lt;!-- Material inspiration
https://ani.stat.fsu.edu/~debdeep/p4_s14.pdf
--&gt;

---
## Normal distribution (Gaussian)

- Examples: 
  - Distribution of a lot of continuous phenotypes 
  - Arm length, or human height in a population
  - velocity of a collection of molecules in a gas of liquid
  - the error from measurements (length of beak).


---
## Normal distribution (Gaussian)

- The `iris` dataset contains values that are 'drawn' from a normal distribution. 
- Here are the histogram for sepal length (mm).

&lt;img src="workshopXX-en_files/figure-html/Iris_normal-1.png" width="792" style="display: block; margin: auto;" /&gt;


---
## Normal distribution (dissection)

.pull-left[

```r
curve(expr = dnorm(x = x, mean=0,sd=1), 
      from = -5, to = 5, ylim = c(0,1),
      col="black", ylab = "Density",
      main = "Density normal") 
abline(h=seq(0,1, by = .1), 
       lty = 3, lwd = .3)
```

&lt;img src="workshopXX-en_files/figure-html/Density_normal-1.png" width="360" style="display: block; margin: auto;" /&gt;
]

--

.pull-right[
For the density distribution, the *area* under the curve between two points is the *probability*. For example, the 95% probability (red region), and 2 tails 2.5% (blue region).

&lt;img src="workshopXX-en_files/figure-html/Density_normal_dissection_all_info-1.png" width="360" style="display: block; margin: auto;" /&gt;
]



---
## Quantile value for a normal distribution 

If X is a random variable distributed with a *Standard normal distribution*, what is the probability of finding X less or equal to `\(1.645\)` or `\(X\sim N(0,1)\)` where `\(P(X\leq 1.645)\)`? 
.pull-left[

```r
pnorm(1.645)
# [1] 0.9500151

qnorm(p = 0.05,lower.tail = F)
# [1] 1.644854
qnorm(p = 0.025,lower.tail = F)
# [1] 1.959964
```
]

--

.pull-right[
&lt;img src="workshopXX-en_files/figure-html/normal_shade_95_5-1.png" width="360" style="display: block; margin: auto;" /&gt;
]


---
## Normal distribution 



.tiny[

```r
par(mfrow=c(2,3), mar = c(4,4,1,1), cex = 1.1)
# The function is not shown, but can be found in the script (Markdown)
draw.normal(where = "both",  prob = 0.05/2)
draw.normal(where = "both",  prob = 0.2/2 )
draw.normal(where = "both",  prob = 0.5/2 )
draw.normal(where = "both",  prob = 0.95/2)
draw.normal(where = "left",  prob = 0.05  )
draw.normal(where = "right", prob = 0.05  )
```

&lt;img src="workshopXX-en_files/figure-html/normal_dist_area-1.png" width="648" style="display: block; margin: auto;" /&gt;
]

---
## Normal distribution 

Find the probability of finding data relative to a standard deviation number 

.small[

```r
sd = 1
probability.left.side = (pnorm(q = c(sd*1,sd*2,sd*3),lower.tail = F)*100)
probability.right.side = (pnorm(q = c(sd*1,sd*2,sd*3),lower.tail = T)*100)
percent.data.under.curve = probability.right.side - probability.left.side
p.from.mean = round(percent.data.under.curve,2)
```
]

- So from the mean of the standard normal distribution:
  * if you are at `\(\mu \pm 1 \sigma\)` , you have 68.27% of the data. 
  * At `\(\mu \pm 2 \sigma\)` it's 95.45% and at `\(\mu \pm 3 \sigma\)` it's 99.73%

&lt;img src="workshopXX-en_files/figure-html/normal_dist_area2-1.png" width="792" style="display: block; margin: auto;" /&gt;

---
## Normal distribution 

Find the "x" value (quantile) based on the probability (area under the curve between 2 values) of a standard normal distribution.


```r
qnorm(p = c(.75, .95,.975, .995), mean = 0, sd = 1, lower.tail = F)
# [1] -0.6744898 -1.6448536 -1.9599640 -2.5758293
```
- for a probability of `\(50\%\)` between the "x" value is -0.67, 0.67. 
- for a probability of `\(90\%\)` between the "x" value is -1.64, 1.64. 
- for a probability of `\(95\%\)` between the "x" value is -1.96, 1.96. 
- for a probability of `\(99\%\)` between the "x" value is -2.58, 2.58. 

&lt;img src="workshopXX-en_files/figure-html/normal_dist_area3-1.png" width="1008" style="display: block; margin: auto;" /&gt;


---
## Normal distribution

.pull-left[

- Mean of population is = 1

&lt;img src="workshopXX-en_files/figure-html/unnamed-chunk-12-1.png" width="432" style="display: block; margin: auto;" /&gt;

]

.pull-right[

- Mean of population is = 2 
&lt;img src="workshopXX-en_files/figure-html/unnamed-chunk-13-1.png" width="432" style="display: block; margin: auto;" /&gt;

]

???

- It is NOT a problem if the y value in the probability density function (PDF) is GREATER than one. The important point here is that the AREA UNDER THE CURVE must sum to 1. 
- see https://stackoverflow.com/questions/42661973/r-density-plot-y-axis-larger-than-1 

---
## Normal distribution

.pull-left[

- Now, if we have distributions that have a standard deviation of 5 and a sample size of 10.

&lt;img src="workshopXX-en_files/figure-html/unnamed-chunk-14-1.png" width="432" style="display: block; margin: auto;" /&gt;
]

--

- Let's see what happens if we increase the sample size from 10 to 100.

.pull-right[
&lt;img src="workshopXX-en_files/figure-html/unnamed-chunk-15-1.png" width="432" style="display: block; margin: auto;" /&gt;

]

---
## Convert a Normal dist. to Standard normal 

- If your data is normally distributed (i.e., `\(X \sim N(\mu, \sigma)\)`), you can convert the data to be `\(X \sim N(0, 1)\)` with a scaling
$$Z = \frac{(X-\mu)}{\sigma} $$ 

&lt;img src="workshopXX-en_files/figure-html/Standard_normal_transformation-1.png" width="720" style="display: block; margin: auto;" /&gt;




---
## Challenge 3 ![:cube]()

Let's say you want to simulate the length of different beak sizes for 10 birds. 

Use `rnorm()` to generate 10 random normal numbers. 


```r
rnorm()
```


---
## Challenge 3 - Solution ![:cube]()

Let's say you want to simulate the length of different beak sizes for 10 birds. 

Use `rnorm()` to generate 10 random normal numbers. 


```r
set.seed(1234)
n &lt;-10
rnorm(n)
#  [1] -1.2070657  0.2774292  1.0844412 -2.3456977  0.4291247  0.5060559
#  [7] -0.5747400 -0.5466319 -0.5644520 -0.8900378
```

- Why are there negative values? Because the mean = 0, so some values spill over the mean on both sides, and in the negative numbers. 

---
## Multivariate normal distribution

- There is a way to generate multiple normal distributions at once. 



```r
set.seed(24601) # setting this so the random results will be repeatable 
library(MASS)
covmat &lt;- matrix(c(1.0,   0.2,   0.6, 
                   0.2,   1.0,  -0.5, 
                   0.6,  -0.5,   1.0), nrow=3) # variance covariance matrix of the data
data &lt;- mvrnorm(n = 300, mu = c(0,0,0), Sigma=covmat) # generate random data that match that variance covariance matrix
plot(data)
```

&lt;img src="workshopXX-en_files/figure-html/unnamed-chunk-16-1.png" width="576" style="display: block; margin: auto;" /&gt;


---
## Central limit theorem (CLT)

- The Central limit theorem (CLT) stipulates that:
  * The more you add (sum) identically distributed random variables, the probability of the new variable will tend to converge to a normal distribution (or is approximated by a normal distribution). 

- We will use simulations to test this assertion! 

???
See 
- [Demonstrating The Central Limit Theorem in R](https://www.r-bloggers.com/2020/08/demonstrating-the-central-limit-theorem-in-r/)
- [Dice Rolls - Theoretical and Empirical](https://rpubs.com/careybaldwin/346995)

---
## Central limit theorem (CLT)

- The CLT could be shown by adding more and more random variables distributed as an exponential with rate = 1. 

.tiny[

```r
n = 1000 # Number of points 
par(mfrow=c(2,2)) # set window 
# Generate multiple addition of random variables 
for(i in c(2,50,1000,5000)){
  clt = replicate(i, rexp(n, rate = 1), simplify = FALSE)
  hist(apply(do.call(cbind,clt),1,sum), main = paste("Histogram of",i,"variables"), xlab = "x") # Draw the histogram 
}
```

&lt;img src="workshopXX-en_files/figure-html/unnamed-chunk-17-1.png" width="576" style="display: block; margin: auto;" /&gt;
]

---
## Central limit theorem (CLT)

- Same thing as above, but for a random variable coming from a uniform *discrete* distribution!

.pull-right2[
&lt;img src="workshopXX-en_files/figure-html/unnamed-chunk-18-1.png" width="216" style="display: block; margin: auto;" /&gt;
]

.tiny[

```r
par(mfrow=c(2,2)) # set window 
# Generate multiple addition of random variables 
for(i in c(2,50,1000,5000)){
  clt = replicate(i, runifdisc(n),simplify = FALSE)
  # Draw the histogram 
  hist(apply(do.call(cbind,clt),1,sum), main = paste("Histogram of",i,"variables"),xlab = "x")
}
```

&lt;img src="workshopXX-en_files/figure-html/CLT_uniform_example_dice-1.png" width="576" style="display: block; margin: auto;" /&gt;
]

---
## Central limit theorem (CLT)

- You are asked to throw 20 dice. If the sum of all 20 dice is `\(\geq 100\)`, you get a candy, if not, you don't get one. What is the probability of having the candy? 

.tiny[

```r
n = 1e6 # Number of points 
par(mfrow=c(1,2), cex = 1) # set window 
i = 20 # flip 20 dice
# Generate multiple addition of random variables 
clt = replicate(i, runifdisc(n,1,6), simplify = FALSE)
sum.rdm.var = apply(do.call(cbind, clt),1,sum) # mean
mean.rdmv = mean(sum.rdm.var); sd.rdmv = sd(sum.rdm.var) #sd
hist(sum.rdm.var, main = paste("Histogram of",i,"variables"), xlab = "x")
# If you dived by the number of dice that were rolled, you can see that it actually tends towards the mean or rolling 1 die a lot of time: 3.5! 
# hist(sum.rdm.var/i, main = paste("Histogram of",i,"variables"), xlab = "x") underlying distribution 
curve(expr = dnorm(x,mean = mean.rdmv,sd = sd.rdmv),
      from = mean.rdmv-5*sd.rdmv, to = mean.rdmv+5*sd.rdmv,ylab = "Density")
```

&lt;img src="workshopXX-en_files/figure-html/unnamed-chunk-19-1.png" width="576" style="display: block; margin: auto;" /&gt;
]

--

- Using the CLT, the probability of getting a sum `\(\geq 100\)` is approx. 4.29e-05.

.tiny[
- This example was taken from [this video](https://www.youtube.com/watch?v=vWTKVMf5HQI&amp;ab_channel=Lastatistiqueexpliqu%C3%A9e%C3%A0monchat)
]

???
- See [Demonstrating The Central Limit Theorem in R](https://medium.com/@orrymr/demonstrating-the-central-limit-theorem-in-r-a5fef67cbbb)
- In other words, the distribution shown here is the sampling distribution of the mean. Its mean is the same as the original mean of the population, its standard deviation is called the standard error of the mean (SEM) and it is normally distributed! SEM is the variability in the collection of sample means. 
- See what happens when the sample size increases: `\(SEM = \sigma/\sqrt{n}\)`. This means that as `\(n \to \infty\)`, the standard error gets smaller. 
- Therefore, the *sampling distribution* of the *sample mean* (so the mean taken out of a sample, multiple times with various samples), approximate a normal distribution, even if the underlying process is not from a normal distribution. `\(\bar X \sim N(\mu,\frac{\sigma}{\sqrt {n}})\)` or the mean of x is approximately normally distributed as with mean mu and variance 
- See also [Using R to simulate the Central Limit Theorem](https://consultglp.com/wp-content/uploads/2016/10/using-r-to-simulate-the-central-limit-theorem.pdf)

&lt;img src="workshopXX-en_files/figure-html/unnamed-chunk-20-1.png" width="432" style="display: block; margin: auto;" /&gt;


#################################
---
class: inverse, center, middle

# Simulating data for power analysis

---
## Power analysis (proportions)

- The [vignette](https://cran.r-project.org/web/packages/pwr/vignettes/pwr-vignette.html) for the `pwr` package explains how to perform power analysis. 
- Example for tossing a coin, which we believe has a 75% chance of landing on 'heads':

.pull-left[
.tiny[

```r
p1 = .75 # Proportion to test 
p2 = .50 # proportion of the null hypothesis 
alpha = 0.05 # Type 1 error 
power = 0.80 # power or 1-beta = power. Beta is the type II error 
coin.prop.power = power.prop.test(p1 = p1, 
                                  p2 = p2, 
                                  sig.level = alpha,
                                  power = power)
n = ceiling(coin.prop.power$n)

coin.pip = rbinom(n, size = 1, prob = p1)
# coin.pip = rbinom(1,size = n,prob = p1) # similar to above 
# prop.test(x = coin.pip, n = n,alternative = "greater")
p.table = table(coin.pip)[c(2,1)]
ptest = prop.test(p.table, alternative = "greater")
ptest
# 
# 	1-sample proportions test with continuity correction
# 
# data:  p.table, null probability 0.5
# X-squared = 10.776, df = 1, p-value = 0.0005142
# alternative hypothesis: true p is greater than 0.5
# 95 percent confidence interval:
#  0.6101819 1.0000000
# sample estimates:
#         p 
# 0.7241379
```
]
]

.pull-right[

```r
curve(dchisq(x, df = ptest$parameter), 
      xlim = c(0, ceiling(max(ptest$statistic))))
abline(v = ptest$statistic, lty = 3)
```

&lt;img src="workshopXX-en_files/figure-html/unnamed-chunk-22-1.png" width="504" style="display: block; margin: auto;" /&gt;
]

???
- For the moment, please refer to [this link](https://www.r-bloggers.com/2020/05/power-analysis-by-data-simulation-in-r-part-ii/).

---
## Power analysis (linear model)

.pull-left[

```r
library(pwr)
r2 = seq(0,0.9,by =.1)
f2 &lt;- function(r2) {
  r2/(1-r2)}
```
]

.pull-right[

```r
plot(f2(r2)~r2)
curve(expr = (x/(1-x)),add = T)
```

&lt;img src="workshopXX-en_files/figure-html/unnamed-chunk-24-1.png" width="504" style="display: block; margin: auto;" /&gt;
]


```r
nb.coef.in.model = 2
pwr.lm = pwr.f2.test(u = nb.coef.in.model, f2 = f2(.3), sig.level = 0.001, power = 0.8)

# sample size (n = v + u + 1)
n = ceiling(pwr.lm$v + nb.coef.in.model + 1)
n
# [1] 53
```


???
- See also the package `pwr2ppl`

#################################
---
class: inverse, center, middle

# Simulating data for models

---
## Linear model (lm) refresher

`$$Y = \beta_{0} + \beta_{1} x_{1} + \cdots + \beta_{p} x_{p} + \epsilon$$`

- `\(Y\)` is the response variable
- `\(\beta_0\)` is the intercept
- `\(\beta_1\)` is the coefficient of variation for the first explanatory variable ($x_1$)
- `\(\beta_p\)` is the coefficient of variation for the `\(p^{th}\)` explanatory variable for the `\(p^{th}\)` `\(x_p\)` explanatory variable
- `\(\epsilon\)` is the residual of the model. Note that `\(\epsilon \sim N(\mu=0,sd = \sigma)\)`

---
## Linear model (lm) refresher

- The goal is to find the **best estimation** of the parameters ( `\(\beta\)` s), while minimizing the residuals, and assess the goodness of fit of the model
- We use the least-squares in order to fit the line to the data.
- In order to interpret your results correctly when using a linear model, be sure you are not crossing the *LINE*: 
  * L: Linearity, 
  * I: Independence, 
  * N: normally distributed residuals, 
  * E: Equality of variance 

???
Reference for the section 
- [Linear model](https://aosmith.rbind.io/2018/01/09/simulate-simulate-part1/) and [Linear mixed model](https://aosmith.rbind.io/2018/04/23/simulate-simulate-part-2/)
- [Poisson model](https://aosmith.rbind.io/2018/07/18/simulate-poisson-edition/)
- [binomial generalized linear mixed model](https://aosmith.rbind.io/2020/08/20/simulate-binomial-glmm/)


---
## Generalized linear model (GLM) refresher

| Type | Equation |
| ---- | -------- |
| Linear | `\(Y = \beta_{0} + \beta_{1} x_{1}  + \epsilon\)` |
| Poisson | `\(Y \sim Poisson(\mu)\)` with `\(\text {ln} \mu=\beta_0+\beta_1x\)` or `\(\mu=e^{\beta_0+\beta_1x}\)` (no separate error term as `\(\lambda\)` determines both the mean and variance) |
| Logistic | `\(Y \sim Binomial(p)\)` with `\(\text{log} \Bigl(\frac{p}{1-p}\Bigr) = \beta_0 +\beta_1x\)` or `\(p=\frac{e^{\beta_0+\beta_1x}}{1+e^{\beta_0+\beta_1x}}\)` where `\(\text{log} \Bigl(\frac{p}{1-p}\Bigr)\)` is the log odds or log likelihood. The Y values are determined by a Bernoulli distribution (binomial of size = 1) |

&lt;!-- 
In GLM, we don'T model the individual Y values, but the mean
https://www.theanalysisfactor.com/generalized-linear-models-no-error-term/
In other words "there's no common error distribution independent of predictor values"
see https://stats.stackexchange.com/questions/124818/logistic-regression-error-term-and-its-distribution
--&gt;
&lt;!-- For poisson (there is no 'error' term like linear regression, since there is inherent variability)?? --&gt;
&lt;!-- logistic regression there is an error (or randomness hidden in the Bernoulli (or binomial(size =1)) with probability p)  --&gt;
&lt;!-- Simulate logistic regression https://stats.stackexchange.com/questions/46523/how-to-simulate-artificial-data-for-logistic-regression/46525 --&gt;

---
## GLM: Linear model

.pull-left[
.small[

```r
set.seed(12345678)
n = 100
x.lm = rnorm(n = n, mean = 10, sd = 1)
beta0 = 2.5
beta1 = .8
err = rnorm(n = n, mean = 0, sd = 1)

# Linear combination 
y.lm = beta0 + beta1*x.lm + err

# Make a dataframe of the data 
df.lm = data.frame(x = x.lm, y = y.lm)
# Colour 
b.5 = scales::alpha("black",alpha = .5)
```
]
]

.pull-right[
&lt;img src="workshopXX-en_files/figure-html/unnamed-chunk-27-1.png" width="288" style="display: block; margin: auto;" /&gt;
]

.small[

```r
summary(lm.out)
# 
# Call:
# lm(formula = y ~ x, data = df.lm)
# 
# Residuals:
#      Min       1Q   Median       3Q      Max 
# -2.33332 -0.55887  0.06904  0.48977  2.33641 
# 
# Coefficients:
#             Estimate Std. Error t value Pr(&gt;|t|)    
# (Intercept)  2.21244    0.98278   2.251   0.0266 *  
# x            0.84330    0.09863   8.550 1.67e-13 ***
# ---
# Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
# 
# Residual standard error: 0.9218 on 98 degrees of freedom
# Multiple R-squared:  0.4273,	Adjusted R-squared:  0.4214 
# F-statistic: 73.11 on 1 and 98 DF,  p-value: 1.675e-13
```
]

---
## GLM: Linear model

- You can simulate the response (fitted line), from a model that you created. 


```r
sim.lm = simulate(lm.out,nsim = 2000,seed = 12)
par(mfrow = c(1,3), cex = 1.2)
hist(rowMeans(sim.lm))
hist(fitted(lm.out))
mean(rowMeans(sim.lm))
# [1] 10.57977
hist(x.lm)
```

&lt;img src="workshopXX-en_files/figure-html/unnamed-chunk-29-1.png" width="576" style="display: block; margin: auto;" /&gt;

```r
mean(x.lm)
# [1] 9.920725
# compare 
head(rowMeans(sim.lm))
#         1         2         3         4         5         6 
# 11.607784 11.973944  8.641016 10.445591 11.387259 10.006595
head(fitted(lm.out))
#         1         2         3         4         5         6 
# 11.613416 12.010193  8.639148 10.428930 11.397939 10.010954

mean(rowMeans(sim.lm))
# [1] 10.57977
mean(fitted(lm.out))
# [1] 10.57855
```


---
## GLM: Linear model

- Simulate 2 independent variables. 

&lt;img src="workshopXX-en_files/figure-html/unnamed-chunk-30-1.png" width="576" style="display: block; margin: auto;" /&gt;

```
# 
# Call:
# lm(formula = y ~ x1 + x2, data = df.lm)
# 
# Residuals:
#      Min       1Q   Median       3Q      Max 
# -1.87048 -0.66548  0.02784  0.49743  2.36528 
# 
# Coefficients:
#             Estimate Std. Error t value Pr(&gt;|t|)    
# (Intercept)  5.44825    0.94089   5.791 8.65e-08 ***
# x1           0.54194    0.09216   5.880 5.81e-08 ***
# x2           0.47958    0.01572  30.513  &lt; 2e-16 ***
# ---
# Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
# 
# Residual standard error: 0.8606 on 97 degrees of freedom
# Multiple R-squared:  0.9102,	Adjusted R-squared:  0.9084 
# F-statistic: 491.7 on 2 and 97 DF,  p-value: &lt; 2.2e-16
#              (Intercept)            x1            x2
# (Intercept)  0.885280976 -8.324939e-02 -3.279173e-03
# x1          -0.083249389  8.493999e-03 -6.417156e-05
# x2          -0.003279173 -6.417156e-05  2.470253e-04
```

---
## GLM: Linear model (categorical)

.small[

```r
set.seed(12345678)
n = 100
x.lm = rnorm(n = n, mean = 10, sd = 1)
catego.lm = c("Control", "Treatment")
# Note that gl is creating factors by default 
gr = gl(n = length(catego.lm), k = n/2, labels = catego.lm)
# Make a dummy matrix
dummy.mat = matrix(data = 0,
                   nrow = n,
                   ncol = length(catego.lm))
colnames(dummy.mat) &lt;- paste0(c(catego.lm))
dummy.mat[,1] &lt;- 1
dummy.mat[,2] &lt;- ifelse(gr=="Treatment",1,0)

# betaControl = .8
beta1 = .8
betaControl = .2
betaTreament = .7

lin.pred.x = dummy.mat %*% c(betaControl,betaTreament) + beta1 * x.lm
err = rnorm(n = n, mean = 0, sd = 1)
y.lm = lin.pred.x + err
df.lm = data.frame(x = x.lm, y = y.lm, gr = gr)
b.5 = scales::alpha("black",alpha = .5)
r.5 = scales::alpha("red",alpha = .5)
```
]

???
- see [Simulating from regression model with categorical variable with 4 categories?](https://stats.stackexchange.com/questions/471947/simulating-from-regression-model-with-categorical-variable-with-4-categories)

---
## GLM: Linear model (categorical)


```r
plot(y~x, data = df.lm, pch = 19, col = ifelse(gr=="Treatment",r.5,b.5))
lm.out = lm(y~x+gr, data = df.lm)
abline(a = lm.out$coefficients[1] + lm.out$coefficients[3], b = lm.out$coefficients[2], col = "red")
abline(a = lm.out$coefficients[1], b = lm.out$coefficients[2], col = "black")
boxplot(y~gr, data = df.lm, pch = 19, col = ifelse(gr=="Treatment",r.5,b.5))
```

&lt;img src="workshopXX-en_files/figure-html/sim_lm_cat_plot-1.png" width="576" style="display: block; margin: auto;" /&gt;

```r
summary(lm.out)
# 
# Call:
# lm(formula = y ~ x + gr, data = df.lm)
# 
# Residuals:
#      Min       1Q   Median       3Q      Max 
# -2.34958 -0.54678  0.08423  0.50000  2.32288 
# 
# Coefficients:
#             Estimate Std. Error t value Pr(&gt;|t|)    
# (Intercept) -0.11840    1.00510  -0.118 0.906470    
# x            0.84485    0.09956   8.486 2.47e-13 ***
# grTreatment  0.73082    0.18612   3.927 0.000161 ***
# ---
# Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
# 
# Residual standard error: 0.9264 on 97 degrees of freedom
# Multiple R-squared:  0.4577,	Adjusted R-squared:  0.4465 
# F-statistic: 40.93 on 2 and 97 DF,  p-value: 1.291e-13
```


---
## GLM: Linear model (interaction; discrete)

- Linear models can also take interactions in consideration. 
- `\(y=\beta_0+\beta_1x_1+\beta_2x_2+\beta_3x_1x_2+\epsilon\)`
- In this simple example `\(\beta_1\)` and `\(\beta_2\)` is the effect that the predictor `\(x_1\)` and `\(x_2\)` have, and `\(\beta_3\)` is the effect that both `\(x_1\)` and `\(x_2\)` predictor have (or that the effect of one variable (e.g., `\(x_1\)`) *depends* on the value of the other (e.g., `\(x_2\)`)). 

&lt;img src="workshopXX-en_files/figure-html/sim_lm_interaction_discrete-1.png" width="576" style="display: block; margin: auto;" /&gt;

```
# 
# Call:
# lm(formula = y ~ x1 * x2, data = df.lm)
# 
# Residuals:
#      Min       1Q   Median       3Q      Max 
# -2.88640 -0.74654 -0.03946  0.71986  2.28651 
# 
# Coefficients:
#             Estimate Std. Error t value Pr(&gt;|t|)    
# (Intercept)   2.5390     0.1476  17.206  &lt; 2e-16 ***
# x1            1.4714     0.2130   6.908 5.36e-10 ***
# x2            2.1023     0.1495  14.065  &lt; 2e-16 ***
# x1:x2        -3.0811     0.2140 -14.399  &lt; 2e-16 ***
# ---
# Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
# 
# Residual standard error: 1.062 on 96 degrees of freedom
# Multiple R-squared:  0.7517,	Adjusted R-squared:  0.7439 
# F-statistic: 96.87 on 3 and 96 DF,  p-value: &lt; 2.2e-16
# (Intercept)          x1          x2       x1:x2 
#    2.538957    1.471399    2.102262   -3.081078
```

---
## GLM: Linear model (interaction; discrete)


```r
# Binary x1 
plot(x = df.lm[df.lm$x1 == 0, ]$x2, y = df.lm[df.lm$x1 == 0, ]$y, 
     col = rgb(red = 0, green = 0, blue = 1, alpha = 0.25), pch = 19,
     xlab = "x2", ylab = "y")
abline(a = coef(lm.out)[1], b = coef(lm.out)[3], col = "blue", pch = 19, lwd = 2)
points(x = df.lm[df.lm$x1 == 1, ]$x2, y = df.lm[df.lm$x1 == 1, ]$y, 
       col = rgb(red = 1, green = 0, blue = 0, alpha = 0.25), pch = 19)
# If x1 = 1
# $\hat{y}=\beta_0+\beta_1x_1+\beta_2x_2+\beta_3x_1x_2$
# $\hat{y}=\beta_0+\beta_1 (1)+\beta_2x_2+\beta_3(1)x_2$
# $\hat{y}=(\beta_0+\beta_1)+x_2(\beta_2+\beta_3)$
abline(a = coef(lm.out)[1] + coef(lm.out)[2], b = coef(lm.out)[3] + coef(lm.out)[4], 
       col = "red", lwd = 2)
abline(h = 0, v = 0,lty = 3)
```

&lt;img src="workshopXX-en_files/figure-html/unnamed-chunk-31-1.png" width="576" style="display: block; margin: auto;" /&gt;


???
See [Tutorial 11: Interaction terms, by Johannes Karreth](http://www.jkarreth.net/files/RPOS517_Day11_Interact.html)


---
## GLM: Linear model (interaction; continuous)

- Here is the simulation of a linear model with an interaction term between two continuous variables. 
- The plot is to show that the relationship between `\(y\)` and `\(x_1\)` is modulated by `\(x_2\)` 
- If `\(\hat{y}=\beta_0+\beta_1x_1+\beta_2x_2+\beta_3x_1x_2\)`, but you want to show the graph for `\(x_1\)`, we can rearrange the equation to this: 
  * `\(\hat{y}=(\beta_0+\beta_2x_2)+(\beta_1+\beta_3x_2)x_1\)`
  * You can see from the equation above that in order to get an intercept and slope for `\(x_1\)`, you need to specify a `\(x_2\)` value. Therefore, if you want to show the effect of `\(x_1\)` on `\(y\)` you need to set `\(x_2\)` to a certain value. So for a *given value of* `\(x_2\)`, the intercept on the y-axis is the first term `\((\beta_0+\beta_2x_2)\)` while the slope of `\(x_1\)` is `\((\beta_1+\beta_3x_2)\)`.



???
See [Tutorial 11: Interaction terms, by Johannes Karreth](http://www.jkarreth.net/files/RPOS517_Day11_Interact.html)

[DECOMPOSING, PROBING, AND PLOTTING INTERACTIONS IN R](https://stats.oarc.ucla.edu/r/seminars/interactions-r/)
[HOW CAN I EXPLAIN A CONTINUOUS BY CONTINUOUS INTERACTION? | R FAQ](https://stats.oarc.ucla.edu/r/faq/how-can-i-explain-a-continuous-by-continuous-interaction/)

[how to interpret the interaction term in lm formula in R?](https://stats.stackexchange.com/questions/56784/how-to-interpret-the-interaction-term-in-lm-formula-in-r)


---
## GLM: Linear model (interaction; continuous)


```r
hist(df.lm$x1);hist(df.lm$x2);hist(df.lm$y)
```

&lt;img src="workshopXX-en_files/figure-html/unnamed-chunk-32-1.png" width="576" style="display: block; margin: auto;" /&gt;

```r

lm.out = lm(y~x1*x2, data = df.lm)
# summary(lm.out)
# coef(lm.out)

# To show the plot of an interaction between 2 continuous variables, we need to use the marginal effects or the conditional effect (effect of x1 conditional on the values of x2)
# The marginal effect of x_1 is $x_1 = \beta_1+\beta_3\times x_2$
# The marginal effect of x_2 is $x_2 = \beta_2+\beta_3\times x_1$

# MAke a new range of x2 values on which we will test the effect of x1 
x2r = range(x.lm2)
# x2.sim = seq(floor(x2r[1]),ceiling(x2r[2]), by = 1)
x2.sim = seq(x2r[1],x2r[2], by = .5)
# this is the effect of x1 at different values of x2 (which moderates the effect of x1)
eff.x1 &lt;- coef(lm.out)["x1"] + coef(lm.out)["x1:x2"] * x2.sim # this gets you the slopes  
eff.x1.int &lt;- coef(lm.out)["(Intercept)"] + coef(lm.out)["x2"] * x2.sim # this gets you the intercepts  
eff.dat &lt;- data.frame(x2.sim, eff.x1, eff.x1.int)
virPal &lt;- viridis::viridis(length(x2.sim),alpha = .8)
eff.dat$x2.col &lt;- virPal[as.numeric(cut(eff.dat$x2.sim,breaks = length(x2.sim)))]
# eff.dat
df.lm$x2.col &lt;- virPal[as.numeric(cut(df.lm$x2,breaks = length(x2.sim)))]
df.lm$line &lt;- c("black","red")[as.numeric(cut(df.lm$x2,breaks = length(x2.sim))) %% 2 + 1]
```


---
## GLM: Linear model (interaction; continuous)


```r
par(mfrow=c(1,1), mar =c(4,4,1,1))
plot(x = df.lm$x1, y = df.lm$y, bg = df.lm$x2.col, #col =df.lm$line,
     # xlim=range(x2.sim),
     pch = 21, xlab = "x1", ylab = "y")
apply(eff.dat, 1, function(x) abline(a = x[3], b = x[2], col = x[4], lwd  = 2))
# NULL
abline(h = 0, v = 0,lty = 3)
legend("topleft", title = "x2",legend = round(eff.dat$x2.sim,1), lty = 1, lwd = 3,
       col = eff.dat$x2.col, bg = scales::alpha("white",.5))
```

&lt;img src="workshopXX-en_files/figure-html/unnamed-chunk-33-1.png" width="576" style="display: block; margin: auto;" /&gt;

```r
# summary(lm.out)
```


---
## GLM: Logistic

- Recall that logistic is `\(Y \sim Binomial(p)\)` with `\(\text{log} \Bigl(\frac{p}{1-p}\Bigr) = \beta_0 +\beta_1x\)` or `\(p=\frac{e^{\beta_0+\beta_1x}}{1+e^{\beta_0+\beta_1x}}\)` where `\(\text{log} \Bigl(\frac{p}{1-p}\Bigr)\)` is the log odds or log likelihood. The Y values are determined by a Bernoulli distribution (binomial of size = 1) 
- We use maximum likelihood to fit a line to the data. It doesn't have 'residuals' and don't have an `\(R^2\)`
- Remember that is order to make the domain `\(]-\infty,\infty[\)` for the y values (instead of `\([0,1]\)`), we need to transform the probability to the `\(log(odds)\)`. 

---
## GLM: Logistic

- This simulation concerns a continuous variable. 

&lt;!-- Inspiration : https://stats.stackexchange.com/questions/46523/how-to-simulate-artificial-data-for-logistic-regression/46525
--&gt;

```r
set.seed(987654)
n = 1000
x1 = rnorm(n = n, mean = 6, sd = 1)
x2 = rnorm(n = n, mean = 0, sd = 1)
# Rescale the data
x1z = scale(x1)
x2z = scale(x2)
z = 0 + 2*x1z + 3*x2 # this is the equation that gives the LOG odds. Note $p = odds/(1+odds)$
pr = 1/(1+exp(-z)) # Transform to get the LOG(odds) # inverse-logit function; Note that 1/(1+exp(-x))== exp(x)/(1+exp(x)), same as pr2 = boot::inv.logit(z)
# pr2 = exp(z)/(1+exp(z)) #  z being the log odds. Using the exp (exponential with natural number), you get the odds. The use the $p = odds/(1+odds)$ formula to get teh probability
y = rbinom(n = n, size = 1, prob = pr) # Bernoulli response variable (which is a special case of the binomial with size =1 )

# Combine the data in a dataframe 
df = data.frame(y = y, x1 = x1, x2 = x2)
```

---
## GLM: Logistic


```r
#now feed it to glm:
glm.logist = glm( y~x1+x2, data=df, family="binomial")
plot(y~x1, data = df, col = scales::alpha("black",.5), pch = 19)
newdata &lt;- data.frame(x1=seq(min(x1), max(x1),len=n), x2 = seq(min(x2), max(x2),len=n))
newdata$y = predict(object = glm.logist, newdata = newdata, type = "response") 
lines(x = newdata$x1,
      y = newdata$y, col = "red",lwd = 2)
```

&lt;img src="workshopXX-en_files/figure-html/sim_logistic_glm-1.png" width="360" style="display: block; margin: auto;" /&gt;

- The coefficients are given in log(odds)


```
# (Intercept)          x1          x2 
#   -12.97938    12.99357    14.59993
```

- The z-values are the number of standard deviations the estimate is from a standard normal distribution (Wald's test, as this value is given by the ratio of the *estimate* value divided by the *standard error*). 
- For every one unit in X, the log(odds) increases by about 2.



---
## GLM: Logistic


```
#              Estimate Std. Error   z value   Pr(&gt;|z|)
# (Intercept) -8.901572  4.5911357 -1.938861 0.05251832
# x1           1.600191  0.7896567  2.026439 0.04271982
```

&lt;img src="workshopXX-en_files/figure-html/unnamed-chunk-35-1.png" width="576" style="display: block; margin: auto;" /&gt;

```
# [1] 0.0001361562
# [1] 0.8320451
```




---
## GLM: Logistic (categories)

- We can also simulate a model with categorical variables 



&lt;img src="workshopXX-en_files/figure-html/unnamed-chunk-37-1.png" width="432" style="display: block; margin: auto;" /&gt;

???
See https://stats.stackexchange.com/questions/49916/simulating-data-for-logistic-regression-with-a-categorical-variable

See https://blogs.uoregon.edu/rclub/2016/04/05/plotting-your-logistic-regression-models/



---
## GLM: Poisson

- Recall that Poisson is `\(Y \sim Poisson(\mu)\)` with `\(\text {ln} \mu=\beta_0+\beta_1x\)` or `\(\mu=e^{\beta_0+\beta_1x}\)` (no separate error term as `\(\lambda\)` determines both the mean and variance)


```r
set.seed(42)
n = 1000
x = rnorm(n = n, mean = 0, sd = 1)
# Rescale the data
xz = scale(x)
log.mu = 1 + 2*xz
y = rpois(n = n, lambda = exp(log.mu)) 

# Combine the data in a dataframe 
df = data.frame(y = y, x = x)
```

---
## GLM: Poisson

```r
#now feed it to glm:
glm.poisson = glm( y~x, data=df, family="poisson")
plot(y~x, data = df, col = scales::alpha("black",.5), pch = 19)
newdata &lt;- data.frame(x=seq(min(x), max(x),len=n))
newdata$y = predict(object = glm.poisson, newdata = newdata, type = "response") 
lines(x = newdata$x,
      y = newdata$y, col = "red",lwd = 2)
```

&lt;img src="workshopXX-en_files/figure-html/sim_poisson_glm_plot-1.png" width="360" style="display: block; margin: auto;" /&gt;



---
## Simulate categories (t-test, Anova)
&lt;!-- Although this is with only 2 groups, I'm showing this as a quick example --&gt;
.pull-left[

```r
set.seed(1234); n = 1000
y1 = rnorm(n, mean = 15, sd = 1)
y2 = rnorm(n, mean = 15.5, sd = 1)

sim.aov1 = data.frame(y = y1, gr = "A")
sim.aov2 = data.frame(y = y2, gr = "B")
df.aov = rbind(sim.aov1, sim.aov2)
df.aov$gr = factor(df.aov$gr)

# t.test(y~gr, data = df.aov) or
aov.out = aov(y~gr, data = df.aov)
#summary(aov.out)
tk.test = TukeyHSD(aov.out)
round(tk.test$gr,2)
#     diff  lwr  upr p adj
# B-A 0.54 0.45 0.63     0
```
]

.pull-right[

```r
plot(y~gr, data = df.aov)
```

&lt;img src="workshopXX-en_files/figure-html/Sim_t_test_anova_plot-1.png" width="360" style="display: block; margin: auto;" /&gt;

]

---
## Linear mixed model (LMMs) refresher

- A simple linear model is actually the 'simplest' mixed model (although the convention is that we don't called it mixed)
- What is the random effect in a linear model? 

`$$Y = \beta_{0} + \beta_{1} x_{1} + \cdots + \beta_{p} x_{p} + \epsilon$$`

---
## Linear mixed model (LMMs) refresher

- A simple linear model is actually the 'simplest' mixed model (although the convention is that we don't called it mixed)
- What is the random effect in a linear model? 

`$$Y = \beta_{0} + \beta_{1} x_{1} + \cdots + \beta_{p} x_{p} + \epsilon$$`

- The residual of the model ( `\(\epsilon\)` ) is actually the 'random effect' since it is drawn from a distribution which can change for each data point `\(\epsilon \sim N(\mu=0,sd = \sigma)\)`. 

---
## Linear mixed model (LMMs) refresher

- to simplify the linear model equation we are going to write it this way:
`$$\mathbf{Y} = \mathbf{X}\beta + \epsilon$$`
- This only means that `\(\mathbf{Y}\)` is a vector (or matrix) made up of the linear combination of fixed effects `\(\mathbf{X}\beta\)` and a random part `\(\epsilon\)`. 

- In this type of notation, LMMs is (random intercept):
`$$\mathbf{Y} = \mathbf{X}\beta + \mathbf{Z\upsilon} + \epsilon$$`

- With all the previous elements being the same but the random part which is composed of `\(\mathbf{Z \upsilon} + \epsilon\)`, for all categories `\(\mathbf{\upsilon}\)`. `\(\mathbf{X}\)` is the *design matrix* and `\(\mathbf{Z}\)`  is the *block matrix*.
- Just keep in mind that, as before, `\(\epsilon \sim N(\mu=0,sd = \sigma)\)` and `\(\mathbf{\upsilon} \sim N(\mu=0,sd = \mathbf{D})\)`, where `\(\mathbf{D}\)` is a variance-covariance matrix. Also, `\(\mathbf{\upsilon}\)` and `\(\epsilon\)` are independent. 

&lt;!-- 
See 
https://towardsdatascience.com/linear-mixed-model-from-scratch-f29b2e45f0a4 
https://www.bristol.ac.uk/cmm/learning/videos/random-slopes.html
--&gt;

---
## Linear mixed model (LMMs) refresher

- This is just another way to write the model (there is a part that specify the random slopes)

`$$y_{ij} = \beta_{0} + \beta_{1} x_{1ij} + \upsilon_{1j} x_{1ij} + \upsilon_{0ij} + \epsilon_{0ij}$$`
- `\(\epsilon_{0ij} \sim N(\mu=0,sd = \sigma_{e0})\)` and `\(\mathbf{\upsilon_{0j}} \text{ and } \mathbf{\upsilon_{1j}} \sim N(\mu=0,sd = \Omega_{\upsilon})\)` and `\(\Omega_{\upsilon} = \left[\begin{align*} \sigma_{\upsilon0}^2 &amp; \sigma_{\upsilon10} \\ \sigma_{\upsilon01} &amp; \sigma_{\upsilon1}^2 \end{align*}\right]\)`


---
## Linear mixed model (LMMs) refresher

- Just so that everything is extra clear
`$$\mathbf{Y} = \mathbf{X}\beta + \mathbf{Z \upsilon} + \epsilon$$`

One implementation of that could be ( `\(\beta_1\)` and `\(\beta_2\)` would represent different treatments):

`$$\left[\begin{array}{l} y_{11} \\ y_{21} \\ y_{12} \\ y_{22} \end{array} \right] = \left[ \begin{array}{ll} 1 &amp; 0 \\ 0 &amp; 1 \\ 1 &amp; 0 \\ 0 &amp; 1 \end{array}\right]\left[\begin{array}{l} \beta_{1} \\ \beta_{2} \end{array}\right]+\left[\begin{array}{ll} 1 &amp; 0 \\ 1 &amp; 0 \\ 0 &amp; 1 \\ 0 &amp; 1 \end{array}\right]\left[\begin{array}{l} u_{1} \\ u_{2} \end{array}\right]+\left[\begin{array}{l} \epsilon_{11} \\ \epsilon_{21} \\ \epsilon_{12} \\ \epsilon_{22} \end{array}\right]$$`

---
## LMMs simulation

- (G)LMMs are a neat extention of the LM models that can take into account (see [QCBS workshop ](http://r.qcbs.ca/Workshops/workshop07/workshop07-en/workshop07-en.html#1) on the topic)

For now, please see [this link](https://debruine.github.io/tutorials/sim-lmer.html). 


```r
library(lmerTest, warn.conflicts = F)
n = 20; sd.n = 2
# Generate dataframe 
x = 1:n
values = rnorm(n = n,mean = 0,sd = sd.n)
gr = rep(c("short","tall"), each = n/2)
sim.df = data.frame(x,values,gr)

plot(density(sim.df[sim.df$gr%in%"short","values"]), col = "black", ylim = c(0,1), main = "Density")
lines(density(sim.df[sim.df$gr%in%"tall","values"]), col = "red")
legend("toprigh",legend = c("Short","Tall"),col = c("black","red"), lty = 1)
```

&lt;img src="workshopXX-en_files/figure-html/lmmmmmmmmms-1.png" width="576" style="display: block; margin: auto;" /&gt;

---
## Time series simulation

- We suggest that for the moment, you consult the [Time Series Analysis CRAN Task View](https://cran.r-project.org/web/views/TimeSeries.html) which contains packages that can simualte time series for you to analyze. 
- You can also see [this guide to better understand Time series](https://www.analyticsvidhya.com/blog/2015/12/complete-tutorial-time-series-modeling/)
- There is a neat tutorial by [Daniel Pinedo called Time series analysis in R](https://rpubs.com/odenipinedo/time-series-analysis-in-R)


```r
??arima.sim
```


---
## Spatial simulation

- The examples here were taken from [this post on Stack Overflow](https://stackoverflow.com/questions/68761166/create-random-points-inside-a-shape-in-r)
- See also the help for `st_sample()`. In the *Examples* section, there are some neat simulations of points in polygons. 

.tiny[


```r
library(sf); library(ggplot2)
polygon = list(matrix(c(2, 2, 3, 3, 2.5, 4, 
                        3, 4, 2.5, 5, 1, 4, 
                        0, 5, 1, 3, 2, 2), ncol=2, byrow=T)) 
polygon = sf::st_polygon(polygon) # Create an sf polygon
points = sf::st_sample(polygon, size=50) # Sample 50 random points within the polygon

# Plot using the ggplot geom_sf function.
ggplot() + 
  geom_sf(aes(), data=polygon) + 
  geom_sf(aes(), col = alpha("black",.4), data=points) + theme_classic()
```

&lt;img src="workshopXX-en_files/figure-html/unnamed-chunk-39-1.png" width="576" style="display: block; margin: auto;" /&gt;
]

---
## Spatial simulation

- [QGIS](https://qgis.org/en/site/) was used to draw the Jardin Botanique de Montréal. Buildings were added to remove them from the Jardin Botanique polygon. 
- Here, 5 random points were generated and added to the polygon

<div id="htmlwidget-d6077d8d0cc4b31dae18" style="width:432px;height:432px;" class="leaflet html-widget"></div>
<script type="application/json" data-for="htmlwidget-d6077d8d0cc4b31dae18">{"x":{"options":{"minZoom":1,"maxZoom":52,"crs":{"crsClass":"L.CRS.EPSG3857","code":null,"proj4def":null,"projectedBounds":null,"options":{}},"preferCanvas":false,"bounceAtZoomLimits":false,"maxBounds":[[[-90,-370]],[[90,370]]]},"calls":[{"method":"addProviderTiles","args":["CartoDB.Positron","CartoDB.Positron","CartoDB.Positron",{"errorTileUrl":"","noWrap":false,"detectRetina":false,"pane":"tilePane"}]},{"method":"addProviderTiles","args":["CartoDB.DarkMatter","CartoDB.DarkMatter","CartoDB.DarkMatter",{"errorTileUrl":"","noWrap":false,"detectRetina":false,"pane":"tilePane"}]},{"method":"addProviderTiles","args":["OpenStreetMap","OpenStreetMap","OpenStreetMap",{"errorTileUrl":"","noWrap":false,"detectRetina":false,"pane":"tilePane"}]},{"method":"addProviderTiles","args":["Esri.WorldImagery","Esri.WorldImagery","Esri.WorldImagery",{"errorTileUrl":"","noWrap":false,"detectRetina":false,"pane":"tilePane"}]},{"method":"addProviderTiles","args":["OpenTopoMap","OpenTopoMap","OpenTopoMap",{"errorTileUrl":"","noWrap":false,"detectRetina":false,"pane":"tilePane"}]},{"method":"createMapPane","args":["polygon",420]},{"method":"addPolygons","args":[[[[{"lng":[-73.5548356516709,-73.5547468608664,-73.5546059390972,-73.5544983138797,-73.554463335684,-73.5545090764015,-73.5545817234233,-73.5546059390972,-73.5546436079234,-73.5547095283691,-73.5547579597169,-73.55477679413,-73.5547579597169,-73.5547041471082,-73.554647643869,-73.5546691689125,-73.5547041471082,-73.5550485478042,-73.5554306173264,-73.5556539396527,-73.5565727899471,-73.5579382848942,-73.5594114050588,-73.5620266978442,-73.5643729275859,-73.5661635421421,-73.5662254266422,-73.5662657860987,-73.5673447289042,-73.5689106758189,-73.5704833493097,-73.5721824824311,-73.5734107552259,-73.5733650145084,-73.5731793610083,-73.5731094046169,-73.5703461271574,-73.5672344130563,-73.5639922033789,-73.5605387792122,-73.5563521582512,-73.5548356516709],"lat":[45.5548173679962,45.5548588170997,45.5554475794742,45.5564423383664,45.557307557042,45.5581120053505,45.5592955784404,45.56061476443,45.5612882390868,45.5627919893757,45.5640583485491,45.5643927115851,45.5647638051317,45.5650463620259,45.565221075661,45.5652606333898,45.5653077258877,45.5654150966353,45.5655526062404,45.5656712786427,45.5659721968247,45.5664143882237,45.566896132719,45.5677503596227,45.5685254627602,45.5691145528272,45.5691220871125,45.5690994842533,45.5678699691887,45.5661177106725,45.564365868429,45.5624241794468,45.5610560560784,45.5609204193132,45.560713195845,45.5606585640761,45.5597684355302,45.5587577222118,45.5577229703115,45.5566161367635,45.5552836686487,45.5548173679962]},{"lng":[-73.5716860611154,-73.5718528802025,-73.5726519974425,-73.5724824877249,-73.5716860611154],"lat":[45.5616301538998,45.561375837295,45.5616358053669,45.5618957722358,45.5616301538998]},{"lng":[-73.5727192632035,-73.572347956203,-73.5725605160076,-73.5726143286164,-73.5726224005077,-73.5725093940293,-73.5723022154856,-73.5719793398331,-73.5724179125944,-73.5727488601382,-73.572544372225,-73.5726493068121,-73.5727031194208,-73.5728161258992,-73.5727461695078,-73.5728484134644,-73.572802672747,-73.5729049167036,-73.5727192632035],"lat":[45.5615623362511,45.5614474228261,45.5611422417696,45.5611591963162,45.5611422417696,45.561106448821,45.5614304683664,45.5613212061704,45.5606449061256,45.560737214968,45.5610574689597,45.5610819588957,45.5609934183074,45.5610216759571,45.5611309387357,45.5611554286396,45.5612270144516,45.5612590396537,45.5615623362511]},{"lng":[-73.5728511040949,-73.5728968448123,-73.5728591759862,-73.5729237511167,-73.5731255483995,-73.5730394482255,-73.5730098512907,-73.5729614199428,-73.5728511040949],"lat":[45.561217595271,45.5611177518599,45.5611121003407,45.5610028375255,45.5610650043258,45.5611874538826,45.5611818023704,45.5612533881487,45.561217595271]},{"lng":[-73.5551628995978,-73.554961102315,-73.5551978777935,-73.5551790433805,-73.5552301653588,-73.5552463091414,-73.555391603185,-73.555588019207,-73.5551628995978],"lat":[45.5616645336488,45.561472383549,45.5613537022767,45.5613329801241,45.5613009549641,45.5613216771286,45.5612519752724,45.5614384746396,45.5616645336488]},{"lng":[-73.5692712202976,-73.5690667323843,-73.5689026039276,-73.5688824241993,-73.5687640364601,-73.5687411661014,-73.56867120971,-73.5686389221448,-73.568597217373,-73.5685689657534,-73.5685797282751,-73.5686254689926,-73.5686913894383,-73.5687505833079,-73.5688541725798,-73.5689012586124,-73.5689281649168,-73.5689577618516,-73.5691891560693,-73.5691703216562,-73.5692712202976],"lat":[45.5643818801377,45.5645052643284,45.5643771708121,45.5643903569228,45.5642933447498,45.5643074727466,45.5642547282072,45.5642782748827,45.5642236465804,45.5641680763563,45.5641068548594,45.5640588194843,45.5641125060773,45.564083308112,45.5641501808488,45.5641266341196,45.5641360528124,45.5641181572946,45.564297112216,45.5643074727466,45.5643818801377]},{"lng":[-73.5579813349813,-73.5580136225465,-73.557776847068,-73.5576530780679,-73.5576772937418,-73.5573517274589,-73.5573167492632,-73.5569562047845,-73.5569669673063,-73.5568889390236,-73.5570073267629,-73.5572306490892,-73.5572656272849,-73.5570396143281,-73.5570638300021,-73.5573517274589,-73.5572817710675,-73.5576880562636,-73.5577068906766,-73.557819897155,-73.5578737097638,-73.5577741564376,-73.5578145158941,-73.5577311063506,-73.5576988187853,-73.5575562153721,-73.5576207905026,-73.5575346903286,-73.5574055400676,-73.5575938841982,-73.5575831216765,-73.5576584593287,-73.5574270651111,-73.5574566620459,-73.5573624899806,-73.557346346198,-73.5568781765018,-73.5564234599579,-73.5568970109149,-73.556872795241,-73.5566736885886,-73.556571444632,-73.5566709979581,-73.5566279478711,-73.556547228958,-73.5563346691534,-73.5563857911318,-73.5562297345664,-73.5562943096969,-73.5559956497183,-73.5559310745878,-73.5560252466531,-73.5559902684574,-73.5558611181964,-73.555834211892,-73.5558691900877,-73.555858427566,-73.5559256933269,-73.5559310745878,-73.5560198653923,-73.5559821965661,-73.5562243533055,-73.556547228958,-73.5563104534795,-73.5561221093489,-73.5560252466531,-73.556178612588,-73.5560602248488,-73.5559472183704,-73.5558476650442,-73.5557938524355,-73.5555839832613,-73.5556647021745,-73.5562243533055,-73.5562943096969,-73.5565741352624,-73.5565364664363,-73.5564718913058,-73.5564073161753,-73.5563992442839,-73.5563481223056,-73.556331978523,-73.5562055188924,-73.5561247999793,-73.5563212160013,-73.5563158347403,-73.5560333185444,-73.5558288306312,-73.5557588742398,-73.556159778175,-73.5561490156532,-73.5561113468271,-73.5561005843054,-73.5559956497183,-73.555963362153,-73.5563239066317,-73.5563857911318,-73.5562835471751,-73.5562943096969,-73.5566790698495,-73.5567678606539,-73.5567382637191,-73.5568216732626,-73.5568620327192,-73.5569292984802,-73.556657544806,-73.5566952136321,-73.5567759325452,-73.5568512701975,-73.5567920763278,-73.55680552948,-73.5567678606539,-73.556848579567,-73.5577418688723,-73.5581104852423,-73.5581266290249,-73.5581616072206,-73.5585356048514,-73.5584172171122,-73.5579813349813],"lat":[45.5569717388549,45.5569208714089,45.556845512145,45.5568982636403,45.5569378272293,45.5570602857807,45.557030142162,45.5571714402346,45.5572223074538,45.5572694066898,45.5574144720885,45.5573259257208,45.5573767928001,45.5574521513515,45.5574841787053,45.5573636050433,45.5572863624051,45.5571186889957,45.5571337607833,45.5571111531004,45.5571921639221,45.5572336112741,45.5572637547835,45.557292014309,45.5572712906584,45.5573315776208,45.5573994003762,45.5574295437967,45.5575105541594,45.5577253252532,45.5577347450192,45.5578364783917,45.5579269079015,45.5579400955261,45.5579834262712,45.5579683547115,45.5581624007335,45.5576386633317,45.5574464994637,45.5574163560523,45.5575124381199,45.5573843286598,45.5573466493513,45.5573108539848,45.557354185215,45.5574464994637,45.5574841787053,45.5575595371127,45.5576198237658,45.5577366289723,45.5576688066239,45.5576273595929,45.557593448363,45.557651851024,45.557834594442,45.5578327104923,45.5578854610609,45.5578929968523,45.5579306757945,45.5579269079015,45.5578703894749,45.5577460487364,45.5581284898265,45.5582339903589,45.558299928091,45.5581831240555,45.5581096504249,45.5579796583816,45.5579570510391,45.5579664707662,45.5582245706765,45.5582189188664,45.5575529432561,45.5572873043891,45.5573306356375,45.5572232494389,45.5571780341361,45.5570894873958,45.5571139790613,45.5571045591912,45.5571177470089,45.5571045591912,45.5571478905805,45.5570555758413,45.5569651449293,45.5569425371785,45.5570706476458,45.5568351502383,45.5567654428163,45.5565789279378,45.5565582040242,45.556569507978,45.5565600880167,45.5566015358348,45.5565355961097,45.5563961804362,45.5564489323533,45.5564941482429,45.5565224081554,45.5563622684633,45.5564583523333,45.5564847282689,45.5565826959213,45.556577043946,45.556691967332,45.5567974705615,45.5568238463379,45.5568012385303,45.5569010896119,45.5569387692192,45.556953841055,45.5569651449292,45.5570386200564,45.5566712434601,45.5567918186078,45.5567748627433,45.5567823986837,45.5569010896119,45.5570913713705,45.5569717388549]}]]],null,"only.park",{"crs":{"crsClass":"L.CRS.EPSG3857","code":null,"proj4def":null,"projectedBounds":null,"options":{}},"pane":"polygon","stroke":true,"color":"#333333","weight":0.5,"opacity":0.9,"fill":true,"fillColor":"#00FF00","fillOpacity":0.6,"smoothFactor":1,"noClip":false},"<div class='scrollableContainer'><table class=mapview-popup id='popup'><tr class='coord'><td><\/td><th><b>Feature ID&emsp;<\/b><\/th><td>1&emsp;<\/td><\/tr><tr><td>1<\/td><th>geometry&emsp;<\/th><td>sfc_POLYGON&emsp;<\/td><\/tr><\/table><\/div>",{"maxWidth":800,"minWidth":50,"autoPan":true,"keepInView":false,"closeButton":true,"closeOnClick":true,"className":""},"1",{"interactive":false,"permanent":false,"direction":"auto","opacity":1,"offset":[0,0],"textsize":"10px","textOnly":false,"className":"","sticky":true},{"stroke":true,"weight":1,"opacity":0.9,"fillOpacity":0.84,"bringToFront":false,"sendToBack":false}]},{"method":"addScaleBar","args":[{"maxWidth":100,"metric":true,"imperial":true,"updateWhenIdle":true,"position":"bottomleft"}]},{"method":"addHomeButton","args":[-73.5734107552259,45.5548173679962,-73.554463335684,45.5691220871125,true,"only.park","Zoom to only.park","<strong> only.park <\/strong>","bottomright"]},{"method":"addLegend","args":[{"colors":["#00FF00"],"labels":["only.park"],"na_color":null,"na_label":"NA","opacity":1,"position":"topright","type":"factor","title":"","extra":null,"layerId":null,"className":"info legend","group":"only.park"}]},{"method":"createMapPane","args":["point",440]},{"method":"addCircleMarkers","args":[[45.5603224593996,45.5601521623187,45.5656216862751,45.5665696841017,45.5641270724154],[-73.5694219489232,-73.5595248442306,-73.5584802415173,-73.5671287388955,-73.56800549546],6,null,"rdm.pt",{"crs":{"crsClass":"L.CRS.EPSG3857","code":null,"proj4def":null,"projectedBounds":null,"options":{}},"pane":"point","stroke":true,"color":"#333333","weight":1,"opacity":0.9,"fill":true,"fillColor":"#6666ff","fillOpacity":0.6},null,null,null,{"maxWidth":800,"minWidth":50,"autoPan":true,"keepInView":false,"closeButton":true,"closeOnClick":true,"className":""},["1","2","3","4","5"],{"interactive":false,"permanent":false,"direction":"auto","opacity":1,"offset":[0,0],"textsize":"10px","textOnly":false,"className":"","sticky":true},null]},{"method":"addHomeButton","args":[-73.5694219489232,45.5601521623187,-73.5584802415173,45.5665696841017,true,"rdm.pt","Zoom to rdm.pt","<strong> rdm.pt <\/strong>","bottomright"]},{"method":"addLayersControl","args":[["CartoDB.Positron","CartoDB.DarkMatter","OpenStreetMap","Esri.WorldImagery","OpenTopoMap"],["only.park","rdm.pt"],{"collapsed":true,"autoZIndex":true,"position":"topleft"}]},{"method":"addHomeButton","args":[-73.5734107552259,45.5548173679962,-73.554463335684,45.5691220871125,true,null,"Zoom to full extent","<strong>Zoom full<\/strong>","bottomleft"]}],"limits":{"lat":[45.5548173679962,45.5691220871125],"lng":[-73.5734107552259,-73.554463335684]},"fitBounds":[45.5548173679962,-73.5734107552259,45.5691220871125,-73.554463335684,[]]},"evals":[],"jsHooks":{"render":[{"code":"function(el, x, data) {\n  return (\n      function(el, x, data) {\n      // get the leaflet map\n      var map = this; //HTMLWidgets.find('#' + el.id);\n      // we need a new div element because we have to handle\n      // the mouseover output separately\n      // debugger;\n      function addElement () {\n      // generate new div Element\n      var newDiv = $(document.createElement('div'));\n      // append at end of leaflet htmlwidget container\n      $(el).append(newDiv);\n      //provide ID and style\n      newDiv.addClass('lnlt');\n      newDiv.css({\n      'position': 'relative',\n      'bottomleft':  '0px',\n      'background-color': 'rgba(255, 255, 255, 0.7)',\n      'box-shadow': '0 0 2px #bbb',\n      'background-clip': 'padding-box',\n      'margin': '0',\n      'padding-left': '5px',\n      'color': '#333',\n      'font': '9px/1.5 \"Helvetica Neue\", Arial, Helvetica, sans-serif',\n      'z-index': '700',\n      });\n      return newDiv;\n      }\n\n\n      // check for already existing lnlt class to not duplicate\n      var lnlt = $(el).find('.lnlt');\n\n      if(!lnlt.length) {\n      lnlt = addElement();\n\n      // grab the special div we generated in the beginning\n      // and put the mousmove output there\n\n      map.on('mousemove', function (e) {\n      if (e.originalEvent.ctrlKey) {\n      if (document.querySelector('.lnlt') === null) lnlt = addElement();\n      lnlt.text(\n                           ' lon: ' + (e.latlng.lng).toFixed(5) +\n                           ' | lat: ' + (e.latlng.lat).toFixed(5) +\n                           ' | zoom: ' + map.getZoom() +\n                           ' | x: ' + L.CRS.EPSG3857.project(e.latlng).x.toFixed(0) +\n                           ' | y: ' + L.CRS.EPSG3857.project(e.latlng).y.toFixed(0) +\n                           ' | epsg: 3857 ' +\n                           ' | proj4: +proj=merc +a=6378137 +b=6378137 +lat_ts=0.0 +lon_0=0.0 +x_0=0.0 +y_0=0 +k=1.0 +units=m +nadgrids=@null +no_defs ');\n      } else {\n      if (document.querySelector('.lnlt') === null) lnlt = addElement();\n      lnlt.text(\n                      ' lon: ' + (e.latlng.lng).toFixed(5) +\n                      ' | lat: ' + (e.latlng.lat).toFixed(5) +\n                      ' | zoom: ' + map.getZoom() + ' ');\n      }\n      });\n\n      // remove the lnlt div when mouse leaves map\n      map.on('mouseout', function (e) {\n      var strip = document.querySelector('.lnlt');\n      if( strip !==null) strip.remove();\n      });\n\n      };\n\n      //$(el).keypress(67, function(e) {\n      map.on('preclick', function(e) {\n      if (e.originalEvent.ctrlKey) {\n      if (document.querySelector('.lnlt') === null) lnlt = addElement();\n      lnlt.text(\n                      ' lon: ' + (e.latlng.lng).toFixed(5) +\n                      ' | lat: ' + (e.latlng.lat).toFixed(5) +\n                      ' | zoom: ' + map.getZoom() + ' ');\n      var txt = document.querySelector('.lnlt').textContent;\n      console.log(txt);\n      //txt.innerText.focus();\n      //txt.select();\n      setClipboardText('\"' + txt + '\"');\n      }\n      });\n\n      }\n      ).call(this.getMap(), el, x, data);\n}","data":null},{"code":"function(el, x, data) {\n  return (function(el,x,data){\n           var map = this;\n\n           map.on('keypress', function(e) {\n               console.log(e.originalEvent.code);\n               var key = e.originalEvent.code;\n               if (key === 'KeyE') {\n                   var bb = this.getBounds();\n                   var txt = JSON.stringify(bb);\n                   console.log(txt);\n\n                   setClipboardText('\\'' + txt + '\\'');\n               }\n           })\n        }).call(this.getMap(), el, x, data);\n}","data":null}]}}</script>

---
## Spatial simulation

- 2 points were added to that polygon with known location. 
- But we'd like to add random points in the 100m buffer around these points 

<div id="htmlwidget-70a527e468a173799ac3" style="width:432px;height:432px;" class="leaflet html-widget"></div>
<script type="application/json" data-for="htmlwidget-70a527e468a173799ac3">{"x":{"options":{"minZoom":1,"maxZoom":52,"crs":{"crsClass":"L.CRS.EPSG3857","code":null,"proj4def":null,"projectedBounds":null,"options":{}},"preferCanvas":false,"bounceAtZoomLimits":false,"maxBounds":[[[-90,-370]],[[90,370]]]},"calls":[{"method":"addProviderTiles","args":["CartoDB.Positron","CartoDB.Positron","CartoDB.Positron",{"errorTileUrl":"","noWrap":false,"detectRetina":false,"pane":"tilePane"}]},{"method":"addProviderTiles","args":["CartoDB.DarkMatter","CartoDB.DarkMatter","CartoDB.DarkMatter",{"errorTileUrl":"","noWrap":false,"detectRetina":false,"pane":"tilePane"}]},{"method":"addProviderTiles","args":["OpenStreetMap","OpenStreetMap","OpenStreetMap",{"errorTileUrl":"","noWrap":false,"detectRetina":false,"pane":"tilePane"}]},{"method":"addProviderTiles","args":["Esri.WorldImagery","Esri.WorldImagery","Esri.WorldImagery",{"errorTileUrl":"","noWrap":false,"detectRetina":false,"pane":"tilePane"}]},{"method":"addProviderTiles","args":["OpenTopoMap","OpenTopoMap","OpenTopoMap",{"errorTileUrl":"","noWrap":false,"detectRetina":false,"pane":"tilePane"}]},{"method":"createMapPane","args":["polygon",420]},{"method":"addPolygons","args":[[[[{"lng":[-73.5548356516709,-73.5547468608664,-73.5546059390972,-73.5544983138797,-73.554463335684,-73.5545090764015,-73.5545817234233,-73.5546059390972,-73.5546436079234,-73.5547095283691,-73.5547579597169,-73.55477679413,-73.5547579597169,-73.5547041471082,-73.554647643869,-73.5546691689125,-73.5547041471082,-73.5550485478042,-73.5554306173264,-73.5556539396527,-73.5565727899471,-73.5579382848942,-73.5594114050588,-73.5620266978442,-73.5643729275859,-73.5661635421421,-73.5662254266422,-73.5662657860987,-73.5673447289042,-73.5689106758189,-73.5704833493097,-73.5721824824311,-73.5734107552259,-73.5733650145084,-73.5731793610083,-73.5731094046169,-73.5703461271574,-73.5672344130563,-73.5639922033789,-73.5605387792122,-73.5563521582512,-73.5548356516709],"lat":[45.5548173679962,45.5548588170997,45.5554475794742,45.5564423383664,45.557307557042,45.5581120053505,45.5592955784404,45.56061476443,45.5612882390868,45.5627919893757,45.5640583485491,45.5643927115851,45.5647638051317,45.5650463620259,45.565221075661,45.5652606333898,45.5653077258877,45.5654150966353,45.5655526062404,45.5656712786427,45.5659721968247,45.5664143882237,45.566896132719,45.5677503596227,45.5685254627602,45.5691145528272,45.5691220871125,45.5690994842533,45.5678699691887,45.5661177106725,45.564365868429,45.5624241794468,45.5610560560784,45.5609204193132,45.560713195845,45.5606585640761,45.5597684355302,45.5587577222118,45.5577229703115,45.5566161367635,45.5552836686487,45.5548173679962]},{"lng":[-73.5716860611154,-73.5718528802025,-73.5726519974425,-73.5724824877249,-73.5716860611154],"lat":[45.5616301538998,45.561375837295,45.5616358053669,45.5618957722358,45.5616301538998]},{"lng":[-73.5727192632035,-73.572347956203,-73.5725605160076,-73.5726143286164,-73.5726224005077,-73.5725093940293,-73.5723022154856,-73.5719793398331,-73.5724179125944,-73.5727488601382,-73.572544372225,-73.5726493068121,-73.5727031194208,-73.5728161258992,-73.5727461695078,-73.5728484134644,-73.572802672747,-73.5729049167036,-73.5727192632035],"lat":[45.5615623362511,45.5614474228261,45.5611422417696,45.5611591963162,45.5611422417696,45.561106448821,45.5614304683664,45.5613212061704,45.5606449061256,45.560737214968,45.5610574689597,45.5610819588957,45.5609934183074,45.5610216759571,45.5611309387357,45.5611554286396,45.5612270144516,45.5612590396537,45.5615623362511]},{"lng":[-73.5728511040949,-73.5728968448123,-73.5728591759862,-73.5729237511167,-73.5731255483995,-73.5730394482255,-73.5730098512907,-73.5729614199428,-73.5728511040949],"lat":[45.561217595271,45.5611177518599,45.5611121003407,45.5610028375255,45.5610650043258,45.5611874538826,45.5611818023704,45.5612533881487,45.561217595271]},{"lng":[-73.5551628995978,-73.554961102315,-73.5551978777935,-73.5551790433805,-73.5552301653588,-73.5552463091414,-73.555391603185,-73.555588019207,-73.5551628995978],"lat":[45.5616645336488,45.561472383549,45.5613537022767,45.5613329801241,45.5613009549641,45.5613216771286,45.5612519752724,45.5614384746396,45.5616645336488]},{"lng":[-73.5692712202976,-73.5690667323843,-73.5689026039276,-73.5688824241993,-73.5687640364601,-73.5687411661014,-73.56867120971,-73.5686389221448,-73.568597217373,-73.5685689657534,-73.5685797282751,-73.5686254689926,-73.5686913894383,-73.5687505833079,-73.5688541725798,-73.5689012586124,-73.5689281649168,-73.5689577618516,-73.5691891560693,-73.5691703216562,-73.5692712202976],"lat":[45.5643818801377,45.5645052643284,45.5643771708121,45.5643903569228,45.5642933447498,45.5643074727466,45.5642547282072,45.5642782748827,45.5642236465804,45.5641680763563,45.5641068548594,45.5640588194843,45.5641125060773,45.564083308112,45.5641501808488,45.5641266341196,45.5641360528124,45.5641181572946,45.564297112216,45.5643074727466,45.5643818801377]},{"lng":[-73.5579813349813,-73.5580136225465,-73.557776847068,-73.5576530780679,-73.5576772937418,-73.5573517274589,-73.5573167492632,-73.5569562047845,-73.5569669673063,-73.5568889390236,-73.5570073267629,-73.5572306490892,-73.5572656272849,-73.5570396143281,-73.5570638300021,-73.5573517274589,-73.5572817710675,-73.5576880562636,-73.5577068906766,-73.557819897155,-73.5578737097638,-73.5577741564376,-73.5578145158941,-73.5577311063506,-73.5576988187853,-73.5575562153721,-73.5576207905026,-73.5575346903286,-73.5574055400676,-73.5575938841982,-73.5575831216765,-73.5576584593287,-73.5574270651111,-73.5574566620459,-73.5573624899806,-73.557346346198,-73.5568781765018,-73.5564234599579,-73.5568970109149,-73.556872795241,-73.5566736885886,-73.556571444632,-73.5566709979581,-73.5566279478711,-73.556547228958,-73.5563346691534,-73.5563857911318,-73.5562297345664,-73.5562943096969,-73.5559956497183,-73.5559310745878,-73.5560252466531,-73.5559902684574,-73.5558611181964,-73.555834211892,-73.5558691900877,-73.555858427566,-73.5559256933269,-73.5559310745878,-73.5560198653923,-73.5559821965661,-73.5562243533055,-73.556547228958,-73.5563104534795,-73.5561221093489,-73.5560252466531,-73.556178612588,-73.5560602248488,-73.5559472183704,-73.5558476650442,-73.5557938524355,-73.5555839832613,-73.5556647021745,-73.5562243533055,-73.5562943096969,-73.5565741352624,-73.5565364664363,-73.5564718913058,-73.5564073161753,-73.5563992442839,-73.5563481223056,-73.556331978523,-73.5562055188924,-73.5561247999793,-73.5563212160013,-73.5563158347403,-73.5560333185444,-73.5558288306312,-73.5557588742398,-73.556159778175,-73.5561490156532,-73.5561113468271,-73.5561005843054,-73.5559956497183,-73.555963362153,-73.5563239066317,-73.5563857911318,-73.5562835471751,-73.5562943096969,-73.5566790698495,-73.5567678606539,-73.5567382637191,-73.5568216732626,-73.5568620327192,-73.5569292984802,-73.556657544806,-73.5566952136321,-73.5567759325452,-73.5568512701975,-73.5567920763278,-73.55680552948,-73.5567678606539,-73.556848579567,-73.5577418688723,-73.5581104852423,-73.5581266290249,-73.5581616072206,-73.5585356048514,-73.5584172171122,-73.5579813349813],"lat":[45.5569717388549,45.5569208714089,45.556845512145,45.5568982636403,45.5569378272293,45.5570602857807,45.557030142162,45.5571714402346,45.5572223074538,45.5572694066898,45.5574144720885,45.5573259257208,45.5573767928001,45.5574521513515,45.5574841787053,45.5573636050433,45.5572863624051,45.5571186889957,45.5571337607833,45.5571111531004,45.5571921639221,45.5572336112741,45.5572637547835,45.557292014309,45.5572712906584,45.5573315776208,45.5573994003762,45.5574295437967,45.5575105541594,45.5577253252532,45.5577347450192,45.5578364783917,45.5579269079015,45.5579400955261,45.5579834262712,45.5579683547115,45.5581624007335,45.5576386633317,45.5574464994637,45.5574163560523,45.5575124381199,45.5573843286598,45.5573466493513,45.5573108539848,45.557354185215,45.5574464994637,45.5574841787053,45.5575595371127,45.5576198237658,45.5577366289723,45.5576688066239,45.5576273595929,45.557593448363,45.557651851024,45.557834594442,45.5578327104923,45.5578854610609,45.5578929968523,45.5579306757945,45.5579269079015,45.5578703894749,45.5577460487364,45.5581284898265,45.5582339903589,45.558299928091,45.5581831240555,45.5581096504249,45.5579796583816,45.5579570510391,45.5579664707662,45.5582245706765,45.5582189188664,45.5575529432561,45.5572873043891,45.5573306356375,45.5572232494389,45.5571780341361,45.5570894873958,45.5571139790613,45.5571045591912,45.5571177470089,45.5571045591912,45.5571478905805,45.5570555758413,45.5569651449293,45.5569425371785,45.5570706476458,45.5568351502383,45.5567654428163,45.5565789279378,45.5565582040242,45.556569507978,45.5565600880167,45.5566015358348,45.5565355961097,45.5563961804362,45.5564489323533,45.5564941482429,45.5565224081554,45.5563622684633,45.5564583523333,45.5564847282689,45.5565826959213,45.556577043946,45.556691967332,45.5567974705615,45.5568238463379,45.5568012385303,45.5569010896119,45.5569387692192,45.556953841055,45.5569651449292,45.5570386200564,45.5566712434601,45.5567918186078,45.5567748627433,45.5567823986837,45.5569010896119,45.5570913713705,45.5569717388549]}]]],null,"only.park",{"crs":{"crsClass":"L.CRS.EPSG3857","code":null,"proj4def":null,"projectedBounds":null,"options":{}},"pane":"polygon","stroke":true,"color":"#333333","weight":0.5,"opacity":0.9,"fill":true,"fillColor":"#00FF00","fillOpacity":0.6,"smoothFactor":1,"noClip":false},"<div class='scrollableContainer'><table class=mapview-popup id='popup'><tr class='coord'><td><\/td><th><b>Feature ID&emsp;<\/b><\/th><td>1&emsp;<\/td><\/tr><tr><td>1<\/td><th>geometry&emsp;<\/th><td>sfc_POLYGON&emsp;<\/td><\/tr><\/table><\/div>",{"maxWidth":800,"minWidth":50,"autoPan":true,"keepInView":false,"closeButton":true,"closeOnClick":true,"className":""},"1",{"interactive":false,"permanent":false,"direction":"auto","opacity":1,"offset":[0,0],"textsize":"10px","textOnly":false,"className":"","sticky":true},{"stroke":true,"weight":1,"opacity":0.9,"fillOpacity":0.84,"bringToFront":false,"sendToBack":false}]},{"method":"addScaleBar","args":[{"maxWidth":100,"metric":true,"imperial":true,"updateWhenIdle":true,"position":"bottomleft"}]},{"method":"addHomeButton","args":[-73.5734107552259,45.5548173679962,-73.554463335684,45.5691220871125,true,"only.park","Zoom to only.park","<strong> only.park <\/strong>","bottomright"]},{"method":"addLegend","args":[{"colors":["#00FF00"],"labels":["only.park"],"na_color":null,"na_label":"NA","opacity":1,"position":"topright","type":"factor","title":"","extra":null,"layerId":null,"className":"info legend","group":"only.park"}]},{"method":"createMapPane","args":["point",440]},{"method":"addCircleMarkers","args":[[45.560516,45.56502],[-73.56619,-73.56407],6,null,"c(selected.point, selected.point.no.tree)",{"crs":{"crsClass":"L.CRS.EPSG3857","code":null,"proj4def":null,"projectedBounds":null,"options":{}},"pane":"point","stroke":true,"color":"#333333","weight":1,"opacity":0.9,"fill":true,"fillColor":"red","fillOpacity":0.6},null,null,null,{"maxWidth":800,"minWidth":50,"autoPan":true,"keepInView":false,"closeButton":true,"closeOnClick":true,"className":""},["1","2"],{"interactive":false,"permanent":false,"direction":"auto","opacity":1,"offset":[0,0],"textsize":"10px","textOnly":false,"className":"","sticky":true},null]},{"method":"addHomeButton","args":[-73.56619,45.560516,-73.56407,45.56502,true,"c(selected.point, selected.point.no.tree)","Zoom to c(selected.point, selected.point.no.tree)","<strong> c(selected.point, selected.point.no.tree) <\/strong>","bottomright"]},{"method":"addPolygons","args":[[[[{"lng":[-73.5649090414077,-73.5649107426892,-73.5649159503175,-73.5649246500102,-73.5649368179134,-73.5649524206677,-73.5649714154991,-73.5649937503367,-73.5650193639554,-73.5650481861439,-73.565080137897,-73.5651151316323,-73.5651530714304,-73.5651938532979,-73.5652373654521,-73.5652834886279,-73.5653320964043,-73.5653830555514,-73.5654362263949,-73.5654914631995,-73.5655486145681,-73.5656075238567,-73.5656680296039,-73.5657299659731,-73.5657931632073,-73.565857448094,-73.5659226444402,-73.5659885735551,-73.5660550547397,-73.5661219057822,-73.5661889434572,-73.5662559840279,-73.5663228437496,-73.5663893393731,-73.5664552886469,-73.566520510817,-73.5665848271218,-73.566648061282,-73.5667100399841,-73.5667705933551,-73.5668295554278,-73.5668867645963,-73.566942064058,-73.5669953022443,-73.5670463332351,-73.5670950171596,-73.567141220579,-73.5671848168524,-73.5672256864842,-73.5672637174513,-73.5672988055102,-73.5673308544832,-73.5673597765214,-73.5673854923461,-73.5674079314656,-73.5674270323693,-73.5674427426953,-73.567455019375,-73.5674638287508,-73.5674691466681,-73.5674709585423,-73.5674692593984,-73.5674640538849,-73.5674553562609,-73.5674431903576,-73.5674275895127,-73.5674085964792,-73.5673862633082,-73.5673606512065,-73.567331830369,-73.567299879786,-73.5672648870271,-73.5672269480008,-73.5671861666924,-73.5671426548781,-73.5670965318196,-73.5670479239363,-73.5669969644597,-73.5669437930673,-73.5668885555006,-73.566831403165,-73.566772492715,-73.566711985625,-73.566650047746,-73.5665868488519,-73.566522562173,-73.5664573639222,-73.5663914328111,-73.5663249495605,-73.5662580964049,-73.5661910565928,-73.5661240138844,-73.566057152048,-73.5659906543559,-73.565924703082,-73.5658594790025,-73.5657951609,-73.5657319250732,-73.5656699448539,-73.565609390132,-73.5655504268891,-73.5654932167444,-73.5654379165108,-73.5653846777655,-73.5653336464347,-73.5652849623929,-73.5652387590803,-73.5651951631366,-73.5651542940537,-73.5651162638487,-73.5650811767567,-73.565049128945,-73.5650202082497,-73.5649944939347,-73.5649720564752,-73.5649529573637,-73.5649372489423,-73.5649249742588,-73.564916166949,-73.5649108511448,-73.5649090414077],"lat":[45.560516735024,45.5604696403872,45.5604226728381,45.5603759611109,45.5603296332384,45.560283816201,45.5602386355787,45.5601942152067,45.5601506768363,45.560108139801,45.5600667206894,45.560026533026,45.5599876869595,45.5599502889615,45.5599144415344,45.5598802429304,45.5598477868825,45.5598171623474,45.5597884532619,45.5597617383125,45.5597370907202,45.5597145780396,45.5596942619739,45.5596761982057,45.5596604362441,45.5596470192899,45.5596359841161,45.5596273609682,45.5596211734804,45.5596174386114,45.5596161665977,45.5596173609257,45.5596210183219,45.5596271287622,45.559635675499,45.5596466351074,45.5596599775493,45.5596756662556,45.5596936582265,45.5597139041496,45.5597363485344,45.5597609298651,45.5597875807686,45.5598162281996,45.5598467936403,45.559879193316,45.5599133384246,45.5599491353796,45.5599864860672,45.5600252881148,45.5600654351713,45.5601068171992,45.560149320776,45.5601928294045,45.5602372238329,45.560282382381,45.5603281812741,45.5603744949819,45.560421196563,45.5604681580123,45.5605152506123,45.5605623452856,45.5606093129488,45.5606560248663,45.5607023530032,45.5607481703763,45.560793351402,45.5608377722405,45.5608813111355,45.5609238487477,45.5609652684821,45.5610054568077,45.5610443035681,45.5610817022844,45.5611175504463,45.5611517497934,45.5611842065847,45.5612148318553,45.5612435416604,45.5612702573055,45.5612949055621,45.5613174188681,45.5613377355138,45.5613557998101,45.5613715622419,45.5613849796035,45.5613960151173,45.5614046385342,45.561410826217,45.5614145612049,45.5614158332599,45.5614146388954,45.5614109813851,45.5614048707545,45.5613963237533,45.5613853638091,45.5613720209639,45.5613563317911,45.5613383392956,45.5613180927957,45.5612956477879,45.5612710657952,45.5612444141977,45.5612157660484,45.561185199873,45.5611527994541,45.5611186536022,45.5610828559116,45.5610455045044,45.5610067017612,45.5609665540404,45.5609251713869,45.5608826672302,45.5608391580737,45.560794763175,45.5607496042195,45.5607038049864,45.5606574910095,45.5606107892335,45.5605638276653,45.560516735024]}]],[[{"lng":[-73.5627889389951,-73.5627906421799,-73.562795851987,-73.5628045541279,-73.5628167247422,-73.562832330463,-73.562851328508,-73.5628736667977,-73.5628992840975,-73.5629281101859,-73.562960066047,-73.5629950640873,-73.5630330083755,-73.5630737949059,-73.5631173118831,-73.5631634400286,-73.563212052908,-73.5632630172772,-73.5633161934477,-73.5633714356696,-73.563428592531,-73.5634875073728,-73.5635480187183,-73.5636099607157,-73.5636731635924,-73.5637374541206,-73.5638026560918,-73.5638685907997,-73.56393507753,-73.5640019340558,-73.5640689771366,-73.5641360230209,-73.5642028879492,-73.5642693886581,-73.5643353428824,-73.5644005698543,-73.564464890799,-73.5645281294247,-73.5645901124055,-73.5646506698567,-73.5647096358,-73.5647668486187,-73.5648221515004,-73.564875392867,-73.5649264267898,-73.5649751133899,-73.5650213192211,-73.565064917636,-73.5651057891331,-73.565143821684,-73.5651789110412,-73.5652109610232,-73.5652398837784,-73.5652656000261,-73.5652880392735,-73.5653071400096,-73.565322849873,-73.5653351257963,-73.565343934124,-73.5653492507044,-73.5653510609565,-73.5653493599098,-73.5653441522179,-73.5653354521462,-73.5653232835322,-73.5653076797213,-73.5652886834744,-73.5652663468516,-73.5652407310692,-73.565211906332,-73.565179951641,-73.5651449545771,-73.5651070110608,-73.5650662250894,-73.5650227084521,-73.5649765804237,-73.5649279674373,-73.5648770027382,-73.5648238260186,-73.5647685830342,-73.5647114252054,-73.5646525092019,-73.5645919965131,-73.5645300530055,-73.5644668484683,-73.5644025561475,-73.5643373522712,-73.5642714155665,-73.5642049267696,-73.5641380681302,-73.5640710229118,-73.5640039748894,-73.5639371078459,-73.5638706050678,-73.5638046488431,-73.5637394199615,-73.5636750972186,-73.5636118569261,-73.5635498724279,-73.5634893136255,-73.563430346512,-73.5633731327169,-73.5633178290633,-73.5632645871377,-73.563213552875,-73.5631648661578,-73.5631186604336,-73.5630750623485,-73.5630341914007,-73.5629961596122,-73.5629610712223,-73.5629290224021,-73.5629001009901,-73.5628743862527,-73.5628519486657,-73.5628328497224,-73.5628171417641,-73.5628048678375,-73.5627960615764,-73.5627907471096,-73.5627889389951],"lat":[45.5650207113055,45.5649736167367,45.5649266493202,45.5648799377901,45.5648336101787,45.5647877934656,45.5647426132299,45.5646981933057,45.5646546554432,45.5646121189744,45.5645707004864,45.564530513502,45.564491668168,45.564454270954,45.5644184243602,45.5643842266366,45.5643517715138,45.5643211479459,45.564292439867,45.5642657259609,45.5642410794458,45.5642185678734,45.5641982529437,45.5641801903361,45.5641644295568,45.5641510138032,45.5641399798451,45.5641313579244,45.5641251716723,45.5641214380439,45.5641201672724,45.5641213628407,45.5641250214719,45.5641311331386,45.5641396810899,45.5641506418973,45.5641639855197,45.5641796753846,45.5641976684891,45.5642179155176,45.5642403609767,45.5642649433476,45.5642915952544,45.5643202436488,45.5643508100108,45.5643832105628,45.5644173565004,45.564453154235,45.5644905056504,45.5645293083721,45.5645694560472,45.5646108386366,45.564653342716,45.5646968517871,45.5647412465966,45.5647864054636,45.5648322046122,45.5648785185116,45.5649252202197,45.5649721817313,45.5650192743285,45.565066368934,45.5651133364646,45.565160048185,45.565206376061,45.56525219311,45.5652973737492,45.56534179414,45.5653853325272,45.565427869573,45.565469288684,45.5655094763307,45.5655483223589,45.5655857202913,45.5656215676199,45.5656557660868,45.5656882219531,45.5657188462566,45.5657475550552,45.565774269657,45.5657989168364,45.5658214290345,45.5658417445442,45.5658598076798,45.5658755689293,45.5658889850904,45.5659000193886,45.5659086415781,45.5659148280252,45.5659185617724,45.5659198325852,45.5659186369803,45.5659149782349,45.5659088663778,45.565900318162,45.5658893570187,45.5658760129929,45.5658603226613,45.5658423290321,45.5658220814265,45.5657996353444,45.5657750523113,45.5657483997104,45.5657197505975,45.5656891835008,45.5656567822054,45.5656226355243,45.5655868370542,45.5655494849191,45.5655106815016,45.5654705331621,45.5654291499471,45.5653866452877,45.5653431356886,45.5652987404088,45.5652535811345,45.5652077816458,45.5651614674775,45.5651147655744,45.5650678039441,45.5650207113055]}]]],null,"buffer",{"crs":{"crsClass":"L.CRS.EPSG3857","code":null,"proj4def":null,"projectedBounds":null,"options":{}},"pane":"polygon","stroke":true,"color":"#333333","weight":0.5,"opacity":0.9,"fill":true,"fillColor":"red","fillOpacity":0.6,"smoothFactor":1,"noClip":false},null,{"maxWidth":800,"minWidth":50,"autoPan":true,"keepInView":false,"closeButton":true,"closeOnClick":true,"className":""},["1","2"],{"interactive":false,"permanent":false,"direction":"auto","opacity":1,"offset":[0,0],"textsize":"10px","textOnly":false,"className":"","sticky":true},{"stroke":true,"weight":1,"opacity":0.9,"fillOpacity":0.84,"bringToFront":false,"sendToBack":false}]},{"method":"addHomeButton","args":[-73.5674709585423,45.5596161665977,-73.5627889389951,45.5659198325852,true,"buffer","Zoom to buffer","<strong> buffer <\/strong>","bottomright"]},{"method":"addLayersControl","args":[["CartoDB.Positron","CartoDB.DarkMatter","OpenStreetMap","Esri.WorldImagery","OpenTopoMap"],["only.park","c(selected.point, selected.point.no.tree)","buffer"],{"collapsed":true,"autoZIndex":true,"position":"topleft"}]},{"method":"addHomeButton","args":[-73.5734107552259,45.5548173679962,-73.554463335684,45.5691220871125,true,null,"Zoom to full extent","<strong>Zoom full<\/strong>","bottomleft"]}],"limits":{"lat":[45.5548173679962,45.5691220871125],"lng":[-73.5734107552259,-73.554463335684]},"fitBounds":[45.5548173679962,-73.5734107552259,45.5691220871125,-73.554463335684,[]]},"evals":[],"jsHooks":{"render":[{"code":"function(el, x, data) {\n  return (\n      function(el, x, data) {\n      // get the leaflet map\n      var map = this; //HTMLWidgets.find('#' + el.id);\n      // we need a new div element because we have to handle\n      // the mouseover output separately\n      // debugger;\n      function addElement () {\n      // generate new div Element\n      var newDiv = $(document.createElement('div'));\n      // append at end of leaflet htmlwidget container\n      $(el).append(newDiv);\n      //provide ID and style\n      newDiv.addClass('lnlt');\n      newDiv.css({\n      'position': 'relative',\n      'bottomleft':  '0px',\n      'background-color': 'rgba(255, 255, 255, 0.7)',\n      'box-shadow': '0 0 2px #bbb',\n      'background-clip': 'padding-box',\n      'margin': '0',\n      'padding-left': '5px',\n      'color': '#333',\n      'font': '9px/1.5 \"Helvetica Neue\", Arial, Helvetica, sans-serif',\n      'z-index': '700',\n      });\n      return newDiv;\n      }\n\n\n      // check for already existing lnlt class to not duplicate\n      var lnlt = $(el).find('.lnlt');\n\n      if(!lnlt.length) {\n      lnlt = addElement();\n\n      // grab the special div we generated in the beginning\n      // and put the mousmove output there\n\n      map.on('mousemove', function (e) {\n      if (e.originalEvent.ctrlKey) {\n      if (document.querySelector('.lnlt') === null) lnlt = addElement();\n      lnlt.text(\n                           ' lon: ' + (e.latlng.lng).toFixed(5) +\n                           ' | lat: ' + (e.latlng.lat).toFixed(5) +\n                           ' | zoom: ' + map.getZoom() +\n                           ' | x: ' + L.CRS.EPSG3857.project(e.latlng).x.toFixed(0) +\n                           ' | y: ' + L.CRS.EPSG3857.project(e.latlng).y.toFixed(0) +\n                           ' | epsg: 3857 ' +\n                           ' | proj4: +proj=merc +a=6378137 +b=6378137 +lat_ts=0.0 +lon_0=0.0 +x_0=0.0 +y_0=0 +k=1.0 +units=m +nadgrids=@null +no_defs ');\n      } else {\n      if (document.querySelector('.lnlt') === null) lnlt = addElement();\n      lnlt.text(\n                      ' lon: ' + (e.latlng.lng).toFixed(5) +\n                      ' | lat: ' + (e.latlng.lat).toFixed(5) +\n                      ' | zoom: ' + map.getZoom() + ' ');\n      }\n      });\n\n      // remove the lnlt div when mouse leaves map\n      map.on('mouseout', function (e) {\n      var strip = document.querySelector('.lnlt');\n      if( strip !==null) strip.remove();\n      });\n\n      };\n\n      //$(el).keypress(67, function(e) {\n      map.on('preclick', function(e) {\n      if (e.originalEvent.ctrlKey) {\n      if (document.querySelector('.lnlt') === null) lnlt = addElement();\n      lnlt.text(\n                      ' lon: ' + (e.latlng.lng).toFixed(5) +\n                      ' | lat: ' + (e.latlng.lat).toFixed(5) +\n                      ' | zoom: ' + map.getZoom() + ' ');\n      var txt = document.querySelector('.lnlt').textContent;\n      console.log(txt);\n      //txt.innerText.focus();\n      //txt.select();\n      setClipboardText('\"' + txt + '\"');\n      }\n      });\n\n      }\n      ).call(this.getMap(), el, x, data);\n}","data":null},{"code":"function(el, x, data) {\n  return (function(el,x,data){\n           var map = this;\n\n           map.on('keypress', function(e) {\n               console.log(e.originalEvent.code);\n               var key = e.originalEvent.code;\n               if (key === 'KeyE') {\n                   var bb = this.getBounds();\n                   var txt = JSON.stringify(bb);\n                   console.log(txt);\n\n                   setClipboardText('\\'' + txt + '\\'');\n               }\n           })\n        }).call(this.getMap(), el, x, data);\n}","data":null}]}}</script>




---
## Spatial simulation

- 5 random points were generated and added to the polygon
- 2 points were added to that polygon with known location.
- Uniformly distributed in the 100m buffer 

<div id="htmlwidget-1b091e404977ff18aa13" style="width:432px;height:432px;" class="leaflet html-widget"></div>
<script type="application/json" data-for="htmlwidget-1b091e404977ff18aa13">{"x":{"options":{"minZoom":1,"maxZoom":52,"crs":{"crsClass":"L.CRS.EPSG3857","code":null,"proj4def":null,"projectedBounds":null,"options":{}},"preferCanvas":false,"bounceAtZoomLimits":false,"maxBounds":[[[-90,-370]],[[90,370]]]},"calls":[{"method":"addProviderTiles","args":["CartoDB.Positron","CartoDB.Positron","CartoDB.Positron",{"errorTileUrl":"","noWrap":false,"detectRetina":false,"pane":"tilePane"}]},{"method":"addProviderTiles","args":["CartoDB.DarkMatter","CartoDB.DarkMatter","CartoDB.DarkMatter",{"errorTileUrl":"","noWrap":false,"detectRetina":false,"pane":"tilePane"}]},{"method":"addProviderTiles","args":["OpenStreetMap","OpenStreetMap","OpenStreetMap",{"errorTileUrl":"","noWrap":false,"detectRetina":false,"pane":"tilePane"}]},{"method":"addProviderTiles","args":["Esri.WorldImagery","Esri.WorldImagery","Esri.WorldImagery",{"errorTileUrl":"","noWrap":false,"detectRetina":false,"pane":"tilePane"}]},{"method":"addProviderTiles","args":["OpenTopoMap","OpenTopoMap","OpenTopoMap",{"errorTileUrl":"","noWrap":false,"detectRetina":false,"pane":"tilePane"}]},{"method":"createMapPane","args":["polygon",420]},{"method":"addPolygons","args":[[[[{"lng":[-73.5548356516709,-73.5547468608664,-73.5546059390972,-73.5544983138797,-73.554463335684,-73.5545090764015,-73.5545817234233,-73.5546059390972,-73.5546436079234,-73.5547095283691,-73.5547579597169,-73.55477679413,-73.5547579597169,-73.5547041471082,-73.554647643869,-73.5546691689125,-73.5547041471082,-73.5550485478042,-73.5554306173264,-73.5556539396527,-73.5565727899471,-73.5579382848942,-73.5594114050588,-73.5620266978442,-73.5643729275859,-73.5661635421421,-73.5662254266422,-73.5662657860987,-73.5673447289042,-73.5689106758189,-73.5704833493097,-73.5721824824311,-73.5734107552259,-73.5733650145084,-73.5731793610083,-73.5731094046169,-73.5703461271574,-73.5672344130563,-73.5639922033789,-73.5605387792122,-73.5563521582512,-73.5548356516709],"lat":[45.5548173679962,45.5548588170997,45.5554475794742,45.5564423383664,45.557307557042,45.5581120053505,45.5592955784404,45.56061476443,45.5612882390868,45.5627919893757,45.5640583485491,45.5643927115851,45.5647638051317,45.5650463620259,45.565221075661,45.5652606333898,45.5653077258877,45.5654150966353,45.5655526062404,45.5656712786427,45.5659721968247,45.5664143882237,45.566896132719,45.5677503596227,45.5685254627602,45.5691145528272,45.5691220871125,45.5690994842533,45.5678699691887,45.5661177106725,45.564365868429,45.5624241794468,45.5610560560784,45.5609204193132,45.560713195845,45.5606585640761,45.5597684355302,45.5587577222118,45.5577229703115,45.5566161367635,45.5552836686487,45.5548173679962]},{"lng":[-73.5716860611154,-73.5718528802025,-73.5726519974425,-73.5724824877249,-73.5716860611154],"lat":[45.5616301538998,45.561375837295,45.5616358053669,45.5618957722358,45.5616301538998]},{"lng":[-73.5727192632035,-73.572347956203,-73.5725605160076,-73.5726143286164,-73.5726224005077,-73.5725093940293,-73.5723022154856,-73.5719793398331,-73.5724179125944,-73.5727488601382,-73.572544372225,-73.5726493068121,-73.5727031194208,-73.5728161258992,-73.5727461695078,-73.5728484134644,-73.572802672747,-73.5729049167036,-73.5727192632035],"lat":[45.5615623362511,45.5614474228261,45.5611422417696,45.5611591963162,45.5611422417696,45.561106448821,45.5614304683664,45.5613212061704,45.5606449061256,45.560737214968,45.5610574689597,45.5610819588957,45.5609934183074,45.5610216759571,45.5611309387357,45.5611554286396,45.5612270144516,45.5612590396537,45.5615623362511]},{"lng":[-73.5728511040949,-73.5728968448123,-73.5728591759862,-73.5729237511167,-73.5731255483995,-73.5730394482255,-73.5730098512907,-73.5729614199428,-73.5728511040949],"lat":[45.561217595271,45.5611177518599,45.5611121003407,45.5610028375255,45.5610650043258,45.5611874538826,45.5611818023704,45.5612533881487,45.561217595271]},{"lng":[-73.5551628995978,-73.554961102315,-73.5551978777935,-73.5551790433805,-73.5552301653588,-73.5552463091414,-73.555391603185,-73.555588019207,-73.5551628995978],"lat":[45.5616645336488,45.561472383549,45.5613537022767,45.5613329801241,45.5613009549641,45.5613216771286,45.5612519752724,45.5614384746396,45.5616645336488]},{"lng":[-73.5692712202976,-73.5690667323843,-73.5689026039276,-73.5688824241993,-73.5687640364601,-73.5687411661014,-73.56867120971,-73.5686389221448,-73.568597217373,-73.5685689657534,-73.5685797282751,-73.5686254689926,-73.5686913894383,-73.5687505833079,-73.5688541725798,-73.5689012586124,-73.5689281649168,-73.5689577618516,-73.5691891560693,-73.5691703216562,-73.5692712202976],"lat":[45.5643818801377,45.5645052643284,45.5643771708121,45.5643903569228,45.5642933447498,45.5643074727466,45.5642547282072,45.5642782748827,45.5642236465804,45.5641680763563,45.5641068548594,45.5640588194843,45.5641125060773,45.564083308112,45.5641501808488,45.5641266341196,45.5641360528124,45.5641181572946,45.564297112216,45.5643074727466,45.5643818801377]},{"lng":[-73.5579813349813,-73.5580136225465,-73.557776847068,-73.5576530780679,-73.5576772937418,-73.5573517274589,-73.5573167492632,-73.5569562047845,-73.5569669673063,-73.5568889390236,-73.5570073267629,-73.5572306490892,-73.5572656272849,-73.5570396143281,-73.5570638300021,-73.5573517274589,-73.5572817710675,-73.5576880562636,-73.5577068906766,-73.557819897155,-73.5578737097638,-73.5577741564376,-73.5578145158941,-73.5577311063506,-73.5576988187853,-73.5575562153721,-73.5576207905026,-73.5575346903286,-73.5574055400676,-73.5575938841982,-73.5575831216765,-73.5576584593287,-73.5574270651111,-73.5574566620459,-73.5573624899806,-73.557346346198,-73.5568781765018,-73.5564234599579,-73.5568970109149,-73.556872795241,-73.5566736885886,-73.556571444632,-73.5566709979581,-73.5566279478711,-73.556547228958,-73.5563346691534,-73.5563857911318,-73.5562297345664,-73.5562943096969,-73.5559956497183,-73.5559310745878,-73.5560252466531,-73.5559902684574,-73.5558611181964,-73.555834211892,-73.5558691900877,-73.555858427566,-73.5559256933269,-73.5559310745878,-73.5560198653923,-73.5559821965661,-73.5562243533055,-73.556547228958,-73.5563104534795,-73.5561221093489,-73.5560252466531,-73.556178612588,-73.5560602248488,-73.5559472183704,-73.5558476650442,-73.5557938524355,-73.5555839832613,-73.5556647021745,-73.5562243533055,-73.5562943096969,-73.5565741352624,-73.5565364664363,-73.5564718913058,-73.5564073161753,-73.5563992442839,-73.5563481223056,-73.556331978523,-73.5562055188924,-73.5561247999793,-73.5563212160013,-73.5563158347403,-73.5560333185444,-73.5558288306312,-73.5557588742398,-73.556159778175,-73.5561490156532,-73.5561113468271,-73.5561005843054,-73.5559956497183,-73.555963362153,-73.5563239066317,-73.5563857911318,-73.5562835471751,-73.5562943096969,-73.5566790698495,-73.5567678606539,-73.5567382637191,-73.5568216732626,-73.5568620327192,-73.5569292984802,-73.556657544806,-73.5566952136321,-73.5567759325452,-73.5568512701975,-73.5567920763278,-73.55680552948,-73.5567678606539,-73.556848579567,-73.5577418688723,-73.5581104852423,-73.5581266290249,-73.5581616072206,-73.5585356048514,-73.5584172171122,-73.5579813349813],"lat":[45.5569717388549,45.5569208714089,45.556845512145,45.5568982636403,45.5569378272293,45.5570602857807,45.557030142162,45.5571714402346,45.5572223074538,45.5572694066898,45.5574144720885,45.5573259257208,45.5573767928001,45.5574521513515,45.5574841787053,45.5573636050433,45.5572863624051,45.5571186889957,45.5571337607833,45.5571111531004,45.5571921639221,45.5572336112741,45.5572637547835,45.557292014309,45.5572712906584,45.5573315776208,45.5573994003762,45.5574295437967,45.5575105541594,45.5577253252532,45.5577347450192,45.5578364783917,45.5579269079015,45.5579400955261,45.5579834262712,45.5579683547115,45.5581624007335,45.5576386633317,45.5574464994637,45.5574163560523,45.5575124381199,45.5573843286598,45.5573466493513,45.5573108539848,45.557354185215,45.5574464994637,45.5574841787053,45.5575595371127,45.5576198237658,45.5577366289723,45.5576688066239,45.5576273595929,45.557593448363,45.557651851024,45.557834594442,45.5578327104923,45.5578854610609,45.5578929968523,45.5579306757945,45.5579269079015,45.5578703894749,45.5577460487364,45.5581284898265,45.5582339903589,45.558299928091,45.5581831240555,45.5581096504249,45.5579796583816,45.5579570510391,45.5579664707662,45.5582245706765,45.5582189188664,45.5575529432561,45.5572873043891,45.5573306356375,45.5572232494389,45.5571780341361,45.5570894873958,45.5571139790613,45.5571045591912,45.5571177470089,45.5571045591912,45.5571478905805,45.5570555758413,45.5569651449293,45.5569425371785,45.5570706476458,45.5568351502383,45.5567654428163,45.5565789279378,45.5565582040242,45.556569507978,45.5565600880167,45.5566015358348,45.5565355961097,45.5563961804362,45.5564489323533,45.5564941482429,45.5565224081554,45.5563622684633,45.5564583523333,45.5564847282689,45.5565826959213,45.556577043946,45.556691967332,45.5567974705615,45.5568238463379,45.5568012385303,45.5569010896119,45.5569387692192,45.556953841055,45.5569651449292,45.5570386200564,45.5566712434601,45.5567918186078,45.5567748627433,45.5567823986837,45.5569010896119,45.5570913713705,45.5569717388549]}]]],null,"only.park",{"crs":{"crsClass":"L.CRS.EPSG3857","code":null,"proj4def":null,"projectedBounds":null,"options":{}},"pane":"polygon","stroke":true,"color":"#333333","weight":0.5,"opacity":0.9,"fill":true,"fillColor":"#00FF00","fillOpacity":0.6,"smoothFactor":1,"noClip":false},"<div class='scrollableContainer'><table class=mapview-popup id='popup'><tr class='coord'><td><\/td><th><b>Feature ID&emsp;<\/b><\/th><td>1&emsp;<\/td><\/tr><tr><td>1<\/td><th>geometry&emsp;<\/th><td>sfc_POLYGON&emsp;<\/td><\/tr><\/table><\/div>",{"maxWidth":800,"minWidth":50,"autoPan":true,"keepInView":false,"closeButton":true,"closeOnClick":true,"className":""},"1",{"interactive":false,"permanent":false,"direction":"auto","opacity":1,"offset":[0,0],"textsize":"10px","textOnly":false,"className":"","sticky":true},{"stroke":true,"weight":1,"opacity":0.9,"fillOpacity":0.84,"bringToFront":false,"sendToBack":false}]},{"method":"addScaleBar","args":[{"maxWidth":100,"metric":true,"imperial":true,"updateWhenIdle":true,"position":"bottomleft"}]},{"method":"addHomeButton","args":[-73.5734107552259,45.5548173679962,-73.554463335684,45.5691220871125,true,"only.park","Zoom to only.park","<strong> only.park <\/strong>","bottomright"]},{"method":"addLegend","args":[{"colors":["#00FF00"],"labels":["only.park"],"na_color":null,"na_label":"NA","opacity":1,"position":"topright","type":"factor","title":"","extra":null,"layerId":null,"className":"info legend","group":"only.park"}]},{"method":"createMapPane","args":["point",440]},{"method":"addCircleMarkers","args":[[45.5603224593996,45.5601521623187,45.5656216862751,45.5665696841017,45.5641270724154],[-73.5694219489232,-73.5595248442306,-73.5584802415173,-73.5671287388955,-73.56800549546],6,null,"st_sample(only.park, 5)",{"crs":{"crsClass":"L.CRS.EPSG3857","code":null,"proj4def":null,"projectedBounds":null,"options":{}},"pane":"point","stroke":true,"color":"#333333","weight":1,"opacity":0.9,"fill":true,"fillColor":"#6666ff","fillOpacity":0.6},null,null,null,{"maxWidth":800,"minWidth":50,"autoPan":true,"keepInView":false,"closeButton":true,"closeOnClick":true,"className":""},["1","2","3","4","5"],{"interactive":false,"permanent":false,"direction":"auto","opacity":1,"offset":[0,0],"textsize":"10px","textOnly":false,"className":"","sticky":true},null]},{"method":"addHomeButton","args":[-73.5694219489232,45.5601521623187,-73.5584802415173,45.5665696841017,true,"st_sample(only.park, 5)","Zoom to st_sample(only.park, 5)","<strong> st_sample(only.park, 5) <\/strong>","bottomright"]},{"method":"addCircleMarkers","args":[[45.560516,45.56502],[-73.56619,-73.56407],6,null,"c(selected.point, selected.point.no.tree)",{"crs":{"crsClass":"L.CRS.EPSG3857","code":null,"proj4def":null,"projectedBounds":null,"options":{}},"pane":"point","stroke":true,"color":"#333333","weight":1,"opacity":0.9,"fill":true,"fillColor":"red","fillOpacity":0.6},null,null,null,{"maxWidth":800,"minWidth":50,"autoPan":true,"keepInView":false,"closeButton":true,"closeOnClick":true,"className":""},["1","2"],{"interactive":false,"permanent":false,"direction":"auto","opacity":1,"offset":[0,0],"textsize":"10px","textOnly":false,"className":"","sticky":true},null]},{"method":"addHomeButton","args":[-73.56619,45.560516,-73.56407,45.56502,true,"c(selected.point, selected.point.no.tree)","Zoom to c(selected.point, selected.point.no.tree)","<strong> c(selected.point, selected.point.no.tree) <\/strong>","bottomright"]},{"method":"addPolygons","args":[[[[{"lng":[-73.5649090414077,-73.5649107426892,-73.5649159503175,-73.5649246500102,-73.5649368179134,-73.5649524206677,-73.5649714154991,-73.5649937503367,-73.5650193639554,-73.5650481861439,-73.565080137897,-73.5651151316323,-73.5651530714304,-73.5651938532979,-73.5652373654521,-73.5652834886279,-73.5653320964043,-73.5653830555514,-73.5654362263949,-73.5654914631995,-73.5655486145681,-73.5656075238567,-73.5656680296039,-73.5657299659731,-73.5657931632073,-73.565857448094,-73.5659226444402,-73.5659885735551,-73.5660550547397,-73.5661219057822,-73.5661889434572,-73.5662559840279,-73.5663228437496,-73.5663893393731,-73.5664552886469,-73.566520510817,-73.5665848271218,-73.566648061282,-73.5667100399841,-73.5667705933551,-73.5668295554278,-73.5668867645963,-73.566942064058,-73.5669953022443,-73.5670463332351,-73.5670950171596,-73.567141220579,-73.5671848168524,-73.5672256864842,-73.5672637174513,-73.5672988055102,-73.5673308544832,-73.5673597765214,-73.5673854923461,-73.5674079314656,-73.5674270323693,-73.5674427426953,-73.567455019375,-73.5674638287508,-73.5674691466681,-73.5674709585423,-73.5674692593984,-73.5674640538849,-73.5674553562609,-73.5674431903576,-73.5674275895127,-73.5674085964792,-73.5673862633082,-73.5673606512065,-73.567331830369,-73.567299879786,-73.5672648870271,-73.5672269480008,-73.5671861666924,-73.5671426548781,-73.5670965318196,-73.5670479239363,-73.5669969644597,-73.5669437930673,-73.5668885555006,-73.566831403165,-73.566772492715,-73.566711985625,-73.566650047746,-73.5665868488519,-73.566522562173,-73.5664573639222,-73.5663914328111,-73.5663249495605,-73.5662580964049,-73.5661910565928,-73.5661240138844,-73.566057152048,-73.5659906543559,-73.565924703082,-73.5658594790025,-73.5657951609,-73.5657319250732,-73.5656699448539,-73.565609390132,-73.5655504268891,-73.5654932167444,-73.5654379165108,-73.5653846777655,-73.5653336464347,-73.5652849623929,-73.5652387590803,-73.5651951631366,-73.5651542940537,-73.5651162638487,-73.5650811767567,-73.565049128945,-73.5650202082497,-73.5649944939347,-73.5649720564752,-73.5649529573637,-73.5649372489423,-73.5649249742588,-73.564916166949,-73.5649108511448,-73.5649090414077],"lat":[45.560516735024,45.5604696403872,45.5604226728381,45.5603759611109,45.5603296332384,45.560283816201,45.5602386355787,45.5601942152067,45.5601506768363,45.560108139801,45.5600667206894,45.560026533026,45.5599876869595,45.5599502889615,45.5599144415344,45.5598802429304,45.5598477868825,45.5598171623474,45.5597884532619,45.5597617383125,45.5597370907202,45.5597145780396,45.5596942619739,45.5596761982057,45.5596604362441,45.5596470192899,45.5596359841161,45.5596273609682,45.5596211734804,45.5596174386114,45.5596161665977,45.5596173609257,45.5596210183219,45.5596271287622,45.559635675499,45.5596466351074,45.5596599775493,45.5596756662556,45.5596936582265,45.5597139041496,45.5597363485344,45.5597609298651,45.5597875807686,45.5598162281996,45.5598467936403,45.559879193316,45.5599133384246,45.5599491353796,45.5599864860672,45.5600252881148,45.5600654351713,45.5601068171992,45.560149320776,45.5601928294045,45.5602372238329,45.560282382381,45.5603281812741,45.5603744949819,45.560421196563,45.5604681580123,45.5605152506123,45.5605623452856,45.5606093129488,45.5606560248663,45.5607023530032,45.5607481703763,45.560793351402,45.5608377722405,45.5608813111355,45.5609238487477,45.5609652684821,45.5610054568077,45.5610443035681,45.5610817022844,45.5611175504463,45.5611517497934,45.5611842065847,45.5612148318553,45.5612435416604,45.5612702573055,45.5612949055621,45.5613174188681,45.5613377355138,45.5613557998101,45.5613715622419,45.5613849796035,45.5613960151173,45.5614046385342,45.561410826217,45.5614145612049,45.5614158332599,45.5614146388954,45.5614109813851,45.5614048707545,45.5613963237533,45.5613853638091,45.5613720209639,45.5613563317911,45.5613383392956,45.5613180927957,45.5612956477879,45.5612710657952,45.5612444141977,45.5612157660484,45.561185199873,45.5611527994541,45.5611186536022,45.5610828559116,45.5610455045044,45.5610067017612,45.5609665540404,45.5609251713869,45.5608826672302,45.5608391580737,45.560794763175,45.5607496042195,45.5607038049864,45.5606574910095,45.5606107892335,45.5605638276653,45.560516735024]}]],[[{"lng":[-73.5627889389951,-73.5627906421799,-73.562795851987,-73.5628045541279,-73.5628167247422,-73.562832330463,-73.562851328508,-73.5628736667977,-73.5628992840975,-73.5629281101859,-73.562960066047,-73.5629950640873,-73.5630330083755,-73.5630737949059,-73.5631173118831,-73.5631634400286,-73.563212052908,-73.5632630172772,-73.5633161934477,-73.5633714356696,-73.563428592531,-73.5634875073728,-73.5635480187183,-73.5636099607157,-73.5636731635924,-73.5637374541206,-73.5638026560918,-73.5638685907997,-73.56393507753,-73.5640019340558,-73.5640689771366,-73.5641360230209,-73.5642028879492,-73.5642693886581,-73.5643353428824,-73.5644005698543,-73.564464890799,-73.5645281294247,-73.5645901124055,-73.5646506698567,-73.5647096358,-73.5647668486187,-73.5648221515004,-73.564875392867,-73.5649264267898,-73.5649751133899,-73.5650213192211,-73.565064917636,-73.5651057891331,-73.565143821684,-73.5651789110412,-73.5652109610232,-73.5652398837784,-73.5652656000261,-73.5652880392735,-73.5653071400096,-73.565322849873,-73.5653351257963,-73.565343934124,-73.5653492507044,-73.5653510609565,-73.5653493599098,-73.5653441522179,-73.5653354521462,-73.5653232835322,-73.5653076797213,-73.5652886834744,-73.5652663468516,-73.5652407310692,-73.565211906332,-73.565179951641,-73.5651449545771,-73.5651070110608,-73.5650662250894,-73.5650227084521,-73.5649765804237,-73.5649279674373,-73.5648770027382,-73.5648238260186,-73.5647685830342,-73.5647114252054,-73.5646525092019,-73.5645919965131,-73.5645300530055,-73.5644668484683,-73.5644025561475,-73.5643373522712,-73.5642714155665,-73.5642049267696,-73.5641380681302,-73.5640710229118,-73.5640039748894,-73.5639371078459,-73.5638706050678,-73.5638046488431,-73.5637394199615,-73.5636750972186,-73.5636118569261,-73.5635498724279,-73.5634893136255,-73.563430346512,-73.5633731327169,-73.5633178290633,-73.5632645871377,-73.563213552875,-73.5631648661578,-73.5631186604336,-73.5630750623485,-73.5630341914007,-73.5629961596122,-73.5629610712223,-73.5629290224021,-73.5629001009901,-73.5628743862527,-73.5628519486657,-73.5628328497224,-73.5628171417641,-73.5628048678375,-73.5627960615764,-73.5627907471096,-73.5627889389951],"lat":[45.5650207113055,45.5649736167367,45.5649266493202,45.5648799377901,45.5648336101787,45.5647877934656,45.5647426132299,45.5646981933057,45.5646546554432,45.5646121189744,45.5645707004864,45.564530513502,45.564491668168,45.564454270954,45.5644184243602,45.5643842266366,45.5643517715138,45.5643211479459,45.564292439867,45.5642657259609,45.5642410794458,45.5642185678734,45.5641982529437,45.5641801903361,45.5641644295568,45.5641510138032,45.5641399798451,45.5641313579244,45.5641251716723,45.5641214380439,45.5641201672724,45.5641213628407,45.5641250214719,45.5641311331386,45.5641396810899,45.5641506418973,45.5641639855197,45.5641796753846,45.5641976684891,45.5642179155176,45.5642403609767,45.5642649433476,45.5642915952544,45.5643202436488,45.5643508100108,45.5643832105628,45.5644173565004,45.564453154235,45.5644905056504,45.5645293083721,45.5645694560472,45.5646108386366,45.564653342716,45.5646968517871,45.5647412465966,45.5647864054636,45.5648322046122,45.5648785185116,45.5649252202197,45.5649721817313,45.5650192743285,45.565066368934,45.5651133364646,45.565160048185,45.565206376061,45.56525219311,45.5652973737492,45.56534179414,45.5653853325272,45.565427869573,45.565469288684,45.5655094763307,45.5655483223589,45.5655857202913,45.5656215676199,45.5656557660868,45.5656882219531,45.5657188462566,45.5657475550552,45.565774269657,45.5657989168364,45.5658214290345,45.5658417445442,45.5658598076798,45.5658755689293,45.5658889850904,45.5659000193886,45.5659086415781,45.5659148280252,45.5659185617724,45.5659198325852,45.5659186369803,45.5659149782349,45.5659088663778,45.565900318162,45.5658893570187,45.5658760129929,45.5658603226613,45.5658423290321,45.5658220814265,45.5657996353444,45.5657750523113,45.5657483997104,45.5657197505975,45.5656891835008,45.5656567822054,45.5656226355243,45.5655868370542,45.5655494849191,45.5655106815016,45.5654705331621,45.5654291499471,45.5653866452877,45.5653431356886,45.5652987404088,45.5652535811345,45.5652077816458,45.5651614674775,45.5651147655744,45.5650678039441,45.5650207113055]}]]],null,"buffer",{"crs":{"crsClass":"L.CRS.EPSG3857","code":null,"proj4def":null,"projectedBounds":null,"options":{}},"pane":"polygon","stroke":true,"color":"#333333","weight":0.5,"opacity":0.9,"fill":true,"fillColor":"red","fillOpacity":0.6,"smoothFactor":1,"noClip":false},null,{"maxWidth":800,"minWidth":50,"autoPan":true,"keepInView":false,"closeButton":true,"closeOnClick":true,"className":""},["1","2"],{"interactive":false,"permanent":false,"direction":"auto","opacity":1,"offset":[0,0],"textsize":"10px","textOnly":false,"className":"","sticky":true},{"stroke":true,"weight":1,"opacity":0.9,"fillOpacity":0.84,"bringToFront":false,"sendToBack":false}]},{"method":"addHomeButton","args":[-73.5674709585423,45.5596161665977,-73.5627889389951,45.5659198325852,true,"buffer","Zoom to buffer","<strong> buffer <\/strong>","bottomright"]},{"method":"addCircleMarkers","args":[[45.559870081328,45.5645624317597,45.5605160030243,45.5644463331871,45.5601638616041,45.5657007306182,45.5603014674917,45.5644866036306,45.5597551845272,45.5654906115718,45.5605617598929,45.5648439633749,45.5604935007425,45.5653395004477,45.5604946889424,45.5648968341685,45.5609701411934,45.5650502177768,45.5599899692864,45.56495174516],[-73.5664369930336,-73.5638212471159,-73.5661883679677,-73.5649999102995,-73.5662982706654,-73.5646219463405,-73.5665445901895,-73.5630810314204,-73.566132549555,-73.564507651364,-73.5659609348478,-73.5645919721458,-73.5663789841888,-73.5648476047703,-73.566620629527,-73.564080705822,-73.565647116796,-73.5642036990651,-73.5664434604454,-73.5634601991225],6,null,"rmd.ptdf.sf",{"crs":{"crsClass":"L.CRS.EPSG3857","code":null,"proj4def":null,"projectedBounds":null,"options":{}},"pane":"point","stroke":true,"color":"#333333","weight":1,"opacity":[0.9,0.9,0.9,0.9,0.9,0.9,0.9,0.9,0.9,0.9,0.9,0.9,0.9,0.9,0.9,0.9,0.9,0.9,0.9,0.9],"fill":true,"fillColor":"#FFC0CB","fillOpacity":[0.6,0.6,0.6,0.6,0.6,0.6,0.6,0.6,0.6,0.6,0.6,0.6,0.6,0.6,0.6,0.6,0.6,0.6,0.6,0.6]},null,null,["<div class='scrollableContainer'><table class=mapview-popup id='popup'><tr class='coord'><td><\/td><th><b>Feature ID&emsp;<\/b><\/th><td>1&emsp;<\/td><\/tr><tr><td>1<\/td><th>length.rdm.ppt_x.&emsp;<\/th><td>20&emsp;<\/td><\/tr><tr><td>2<\/td><th>geometry&emsp;<\/th><td>sfc_POINT&emsp;<\/td><\/tr><\/table><\/div>","<div class='scrollableContainer'><table class=mapview-popup id='popup'><tr class='coord'><td><\/td><th><b>Feature ID&emsp;<\/b><\/th><td>2&emsp;<\/td><\/tr><tr><td>1<\/td><th>length.rdm.ppt_x.&emsp;<\/th><td>20&emsp;<\/td><\/tr><tr><td>2<\/td><th>geometry&emsp;<\/th><td>sfc_POINT&emsp;<\/td><\/tr><\/table><\/div>","<div class='scrollableContainer'><table class=mapview-popup id='popup'><tr class='coord'><td><\/td><th><b>Feature ID&emsp;<\/b><\/th><td>3&emsp;<\/td><\/tr><tr><td>1<\/td><th>length.rdm.ppt_x.&emsp;<\/th><td>20&emsp;<\/td><\/tr><tr><td>2<\/td><th>geometry&emsp;<\/th><td>sfc_POINT&emsp;<\/td><\/tr><\/table><\/div>","<div class='scrollableContainer'><table class=mapview-popup id='popup'><tr class='coord'><td><\/td><th><b>Feature ID&emsp;<\/b><\/th><td>4&emsp;<\/td><\/tr><tr><td>1<\/td><th>length.rdm.ppt_x.&emsp;<\/th><td>20&emsp;<\/td><\/tr><tr><td>2<\/td><th>geometry&emsp;<\/th><td>sfc_POINT&emsp;<\/td><\/tr><\/table><\/div>","<div class='scrollableContainer'><table class=mapview-popup id='popup'><tr class='coord'><td><\/td><th><b>Feature ID&emsp;<\/b><\/th><td>5&emsp;<\/td><\/tr><tr><td>1<\/td><th>length.rdm.ppt_x.&emsp;<\/th><td>20&emsp;<\/td><\/tr><tr><td>2<\/td><th>geometry&emsp;<\/th><td>sfc_POINT&emsp;<\/td><\/tr><\/table><\/div>","<div class='scrollableContainer'><table class=mapview-popup id='popup'><tr class='coord'><td><\/td><th><b>Feature ID&emsp;<\/b><\/th><td>6&emsp;<\/td><\/tr><tr><td>1<\/td><th>length.rdm.ppt_x.&emsp;<\/th><td>20&emsp;<\/td><\/tr><tr><td>2<\/td><th>geometry&emsp;<\/th><td>sfc_POINT&emsp;<\/td><\/tr><\/table><\/div>","<div class='scrollableContainer'><table class=mapview-popup id='popup'><tr class='coord'><td><\/td><th><b>Feature ID&emsp;<\/b><\/th><td>7&emsp;<\/td><\/tr><tr><td>1<\/td><th>length.rdm.ppt_x.&emsp;<\/th><td>20&emsp;<\/td><\/tr><tr><td>2<\/td><th>geometry&emsp;<\/th><td>sfc_POINT&emsp;<\/td><\/tr><\/table><\/div>","<div class='scrollableContainer'><table class=mapview-popup id='popup'><tr class='coord'><td><\/td><th><b>Feature ID&emsp;<\/b><\/th><td>8&emsp;<\/td><\/tr><tr><td>1<\/td><th>length.rdm.ppt_x.&emsp;<\/th><td>20&emsp;<\/td><\/tr><tr><td>2<\/td><th>geometry&emsp;<\/th><td>sfc_POINT&emsp;<\/td><\/tr><\/table><\/div>","<div class='scrollableContainer'><table class=mapview-popup id='popup'><tr class='coord'><td><\/td><th><b>Feature ID&emsp;<\/b><\/th><td>9&emsp;<\/td><\/tr><tr><td>1<\/td><th>length.rdm.ppt_x.&emsp;<\/th><td>20&emsp;<\/td><\/tr><tr><td>2<\/td><th>geometry&emsp;<\/th><td>sfc_POINT&emsp;<\/td><\/tr><\/table><\/div>","<div class='scrollableContainer'><table class=mapview-popup id='popup'><tr class='coord'><td><\/td><th><b>Feature ID&emsp;<\/b><\/th><td>10&emsp;<\/td><\/tr><tr><td>1<\/td><th>length.rdm.ppt_x.&emsp;<\/th><td>20&emsp;<\/td><\/tr><tr><td>2<\/td><th>geometry&emsp;<\/th><td>sfc_POINT&emsp;<\/td><\/tr><\/table><\/div>","<div class='scrollableContainer'><table class=mapview-popup id='popup'><tr class='coord'><td><\/td><th><b>Feature ID&emsp;<\/b><\/th><td>11&emsp;<\/td><\/tr><tr><td>1<\/td><th>length.rdm.ppt_x.&emsp;<\/th><td>20&emsp;<\/td><\/tr><tr><td>2<\/td><th>geometry&emsp;<\/th><td>sfc_POINT&emsp;<\/td><\/tr><\/table><\/div>","<div class='scrollableContainer'><table class=mapview-popup id='popup'><tr class='coord'><td><\/td><th><b>Feature ID&emsp;<\/b><\/th><td>12&emsp;<\/td><\/tr><tr><td>1<\/td><th>length.rdm.ppt_x.&emsp;<\/th><td>20&emsp;<\/td><\/tr><tr><td>2<\/td><th>geometry&emsp;<\/th><td>sfc_POINT&emsp;<\/td><\/tr><\/table><\/div>","<div class='scrollableContainer'><table class=mapview-popup id='popup'><tr class='coord'><td><\/td><th><b>Feature ID&emsp;<\/b><\/th><td>13&emsp;<\/td><\/tr><tr><td>1<\/td><th>length.rdm.ppt_x.&emsp;<\/th><td>20&emsp;<\/td><\/tr><tr><td>2<\/td><th>geometry&emsp;<\/th><td>sfc_POINT&emsp;<\/td><\/tr><\/table><\/div>","<div class='scrollableContainer'><table class=mapview-popup id='popup'><tr class='coord'><td><\/td><th><b>Feature ID&emsp;<\/b><\/th><td>14&emsp;<\/td><\/tr><tr><td>1<\/td><th>length.rdm.ppt_x.&emsp;<\/th><td>20&emsp;<\/td><\/tr><tr><td>2<\/td><th>geometry&emsp;<\/th><td>sfc_POINT&emsp;<\/td><\/tr><\/table><\/div>","<div class='scrollableContainer'><table class=mapview-popup id='popup'><tr class='coord'><td><\/td><th><b>Feature ID&emsp;<\/b><\/th><td>15&emsp;<\/td><\/tr><tr><td>1<\/td><th>length.rdm.ppt_x.&emsp;<\/th><td>20&emsp;<\/td><\/tr><tr><td>2<\/td><th>geometry&emsp;<\/th><td>sfc_POINT&emsp;<\/td><\/tr><\/table><\/div>","<div class='scrollableContainer'><table class=mapview-popup id='popup'><tr class='coord'><td><\/td><th><b>Feature ID&emsp;<\/b><\/th><td>16&emsp;<\/td><\/tr><tr><td>1<\/td><th>length.rdm.ppt_x.&emsp;<\/th><td>20&emsp;<\/td><\/tr><tr><td>2<\/td><th>geometry&emsp;<\/th><td>sfc_POINT&emsp;<\/td><\/tr><\/table><\/div>","<div class='scrollableContainer'><table class=mapview-popup id='popup'><tr class='coord'><td><\/td><th><b>Feature ID&emsp;<\/b><\/th><td>17&emsp;<\/td><\/tr><tr><td>1<\/td><th>length.rdm.ppt_x.&emsp;<\/th><td>20&emsp;<\/td><\/tr><tr><td>2<\/td><th>geometry&emsp;<\/th><td>sfc_POINT&emsp;<\/td><\/tr><\/table><\/div>","<div class='scrollableContainer'><table class=mapview-popup id='popup'><tr class='coord'><td><\/td><th><b>Feature ID&emsp;<\/b><\/th><td>18&emsp;<\/td><\/tr><tr><td>1<\/td><th>length.rdm.ppt_x.&emsp;<\/th><td>20&emsp;<\/td><\/tr><tr><td>2<\/td><th>geometry&emsp;<\/th><td>sfc_POINT&emsp;<\/td><\/tr><\/table><\/div>","<div class='scrollableContainer'><table class=mapview-popup id='popup'><tr class='coord'><td><\/td><th><b>Feature ID&emsp;<\/b><\/th><td>19&emsp;<\/td><\/tr><tr><td>1<\/td><th>length.rdm.ppt_x.&emsp;<\/th><td>20&emsp;<\/td><\/tr><tr><td>2<\/td><th>geometry&emsp;<\/th><td>sfc_POINT&emsp;<\/td><\/tr><\/table><\/div>","<div class='scrollableContainer'><table class=mapview-popup id='popup'><tr class='coord'><td><\/td><th><b>Feature ID&emsp;<\/b><\/th><td>20&emsp;<\/td><\/tr><tr><td>1<\/td><th>length.rdm.ppt_x.&emsp;<\/th><td>20&emsp;<\/td><\/tr><tr><td>2<\/td><th>geometry&emsp;<\/th><td>sfc_POINT&emsp;<\/td><\/tr><\/table><\/div>"],{"maxWidth":800,"minWidth":50,"autoPan":true,"keepInView":false,"closeButton":true,"closeOnClick":true,"className":""},["20","20","20","20","20","20","20","20","20","20","20","20","20","20","20","20","20","20","20","20"],{"interactive":false,"permanent":false,"direction":"auto","opacity":1,"offset":[0,0],"textsize":"10px","textOnly":false,"className":"","sticky":true},null]},{"method":"addHomeButton","args":[-73.566620629527,45.5597551845272,-73.5630810314204,45.5657007306182,true,"rmd.ptdf.sf","Zoom to rmd.ptdf.sf","<strong> rmd.ptdf.sf <\/strong>","bottomright"]},{"method":"addLayersControl","args":[["CartoDB.Positron","CartoDB.DarkMatter","OpenStreetMap","Esri.WorldImagery","OpenTopoMap"],["only.park","st_sample(only.park, 5)","c(selected.point, selected.point.no.tree)","buffer","rmd.ptdf.sf"],{"collapsed":true,"autoZIndex":true,"position":"topleft"}]},{"method":"addLegend","args":[{"colors":["#FFC0CB"],"labels":["20"],"na_color":null,"na_label":"NA","opacity":1,"position":"topright","type":"factor","title":"","extra":null,"layerId":null,"className":"info legend","group":"rmd.ptdf.sf"}]},{"method":"addHomeButton","args":[-73.5734107552259,45.5548173679962,-73.554463335684,45.5691220871125,true,null,"Zoom to full extent","<strong>Zoom full<\/strong>","bottomleft"]}],"limits":{"lat":[45.5548173679962,45.5691220871125],"lng":[-73.5734107552259,-73.554463335684]},"fitBounds":[45.5548173679962,-73.5734107552259,45.5691220871125,-73.554463335684,[]]},"evals":[],"jsHooks":{"render":[{"code":"function(el, x, data) {\n  return (\n      function(el, x, data) {\n      // get the leaflet map\n      var map = this; //HTMLWidgets.find('#' + el.id);\n      // we need a new div element because we have to handle\n      // the mouseover output separately\n      // debugger;\n      function addElement () {\n      // generate new div Element\n      var newDiv = $(document.createElement('div'));\n      // append at end of leaflet htmlwidget container\n      $(el).append(newDiv);\n      //provide ID and style\n      newDiv.addClass('lnlt');\n      newDiv.css({\n      'position': 'relative',\n      'bottomleft':  '0px',\n      'background-color': 'rgba(255, 255, 255, 0.7)',\n      'box-shadow': '0 0 2px #bbb',\n      'background-clip': 'padding-box',\n      'margin': '0',\n      'padding-left': '5px',\n      'color': '#333',\n      'font': '9px/1.5 \"Helvetica Neue\", Arial, Helvetica, sans-serif',\n      'z-index': '700',\n      });\n      return newDiv;\n      }\n\n\n      // check for already existing lnlt class to not duplicate\n      var lnlt = $(el).find('.lnlt');\n\n      if(!lnlt.length) {\n      lnlt = addElement();\n\n      // grab the special div we generated in the beginning\n      // and put the mousmove output there\n\n      map.on('mousemove', function (e) {\n      if (e.originalEvent.ctrlKey) {\n      if (document.querySelector('.lnlt') === null) lnlt = addElement();\n      lnlt.text(\n                           ' lon: ' + (e.latlng.lng).toFixed(5) +\n                           ' | lat: ' + (e.latlng.lat).toFixed(5) +\n                           ' | zoom: ' + map.getZoom() +\n                           ' | x: ' + L.CRS.EPSG3857.project(e.latlng).x.toFixed(0) +\n                           ' | y: ' + L.CRS.EPSG3857.project(e.latlng).y.toFixed(0) +\n                           ' | epsg: 3857 ' +\n                           ' | proj4: +proj=merc +a=6378137 +b=6378137 +lat_ts=0.0 +lon_0=0.0 +x_0=0.0 +y_0=0 +k=1.0 +units=m +nadgrids=@null +no_defs ');\n      } else {\n      if (document.querySelector('.lnlt') === null) lnlt = addElement();\n      lnlt.text(\n                      ' lon: ' + (e.latlng.lng).toFixed(5) +\n                      ' | lat: ' + (e.latlng.lat).toFixed(5) +\n                      ' | zoom: ' + map.getZoom() + ' ');\n      }\n      });\n\n      // remove the lnlt div when mouse leaves map\n      map.on('mouseout', function (e) {\n      var strip = document.querySelector('.lnlt');\n      if( strip !==null) strip.remove();\n      });\n\n      };\n\n      //$(el).keypress(67, function(e) {\n      map.on('preclick', function(e) {\n      if (e.originalEvent.ctrlKey) {\n      if (document.querySelector('.lnlt') === null) lnlt = addElement();\n      lnlt.text(\n                      ' lon: ' + (e.latlng.lng).toFixed(5) +\n                      ' | lat: ' + (e.latlng.lat).toFixed(5) +\n                      ' | zoom: ' + map.getZoom() + ' ');\n      var txt = document.querySelector('.lnlt').textContent;\n      console.log(txt);\n      //txt.innerText.focus();\n      //txt.select();\n      setClipboardText('\"' + txt + '\"');\n      }\n      });\n\n      }\n      ).call(this.getMap(), el, x, data);\n}","data":null},{"code":"function(el, x, data) {\n  return (function(el,x,data){\n           var map = this;\n\n           map.on('keypress', function(e) {\n               console.log(e.originalEvent.code);\n               var key = e.originalEvent.code;\n               if (key === 'KeyE') {\n                   var bb = this.getBounds();\n                   var txt = JSON.stringify(bb);\n                   console.log(txt);\n\n                   setClipboardText('\\'' + txt + '\\'');\n               }\n           })\n        }).call(this.getMap(), el, x, data);\n}","data":null}]}}</script>

---
## Spatial simulation

- What if we want to add a grid to the polygon:

<div id="htmlwidget-1d4a57a1256a5de20eef" style="width:432px;height:432px;" class="leaflet html-widget"></div>
<script type="application/json" data-for="htmlwidget-1d4a57a1256a5de20eef">{"x":{"options":{"minZoom":1,"maxZoom":52,"crs":{"crsClass":"L.CRS.EPSG3857","code":null,"proj4def":null,"projectedBounds":null,"options":{}},"preferCanvas":false,"bounceAtZoomLimits":false,"maxBounds":[[[-90,-370]],[[90,370]]]},"calls":[{"method":"addProviderTiles","args":["CartoDB.Positron","CartoDB.Positron","CartoDB.Positron",{"errorTileUrl":"","noWrap":false,"detectRetina":false,"pane":"tilePane"}]},{"method":"addProviderTiles","args":["CartoDB.DarkMatter","CartoDB.DarkMatter","CartoDB.DarkMatter",{"errorTileUrl":"","noWrap":false,"detectRetina":false,"pane":"tilePane"}]},{"method":"addProviderTiles","args":["OpenStreetMap","OpenStreetMap","OpenStreetMap",{"errorTileUrl":"","noWrap":false,"detectRetina":false,"pane":"tilePane"}]},{"method":"addProviderTiles","args":["Esri.WorldImagery","Esri.WorldImagery","Esri.WorldImagery",{"errorTileUrl":"","noWrap":false,"detectRetina":false,"pane":"tilePane"}]},{"method":"addProviderTiles","args":["OpenTopoMap","OpenTopoMap","OpenTopoMap",{"errorTileUrl":"","noWrap":false,"detectRetina":false,"pane":"tilePane"}]},{"method":"createMapPane","args":["polygon",420]},{"method":"addPolygons","args":[[[[{"lng":[-73.5548356516709,-73.5547468608664,-73.5546059390972,-73.5544983138797,-73.554463335684,-73.5545090764015,-73.5545817234233,-73.5546059390972,-73.5546436079234,-73.5547095283691,-73.5547579597169,-73.55477679413,-73.5547579597169,-73.5547041471082,-73.554647643869,-73.5546691689125,-73.5547041471082,-73.5550485478042,-73.5554306173264,-73.5556539396527,-73.5565727899471,-73.5579382848942,-73.5594114050588,-73.5620266978442,-73.5643729275859,-73.5661635421421,-73.5662254266422,-73.5662657860987,-73.5673447289042,-73.5689106758189,-73.5704833493097,-73.5721824824311,-73.5734107552259,-73.5733650145084,-73.5731793610083,-73.5731094046169,-73.5703461271574,-73.5672344130563,-73.5639922033789,-73.5605387792122,-73.5563521582512,-73.5548356516709],"lat":[45.5548173679962,45.5548588170997,45.5554475794742,45.5564423383664,45.557307557042,45.5581120053505,45.5592955784404,45.56061476443,45.5612882390868,45.5627919893757,45.5640583485491,45.5643927115851,45.5647638051317,45.5650463620259,45.565221075661,45.5652606333898,45.5653077258877,45.5654150966353,45.5655526062404,45.5656712786427,45.5659721968247,45.5664143882237,45.566896132719,45.5677503596227,45.5685254627602,45.5691145528272,45.5691220871125,45.5690994842533,45.5678699691887,45.5661177106725,45.564365868429,45.5624241794468,45.5610560560784,45.5609204193132,45.560713195845,45.5606585640761,45.5597684355302,45.5587577222118,45.5577229703115,45.5566161367635,45.5552836686487,45.5548173679962]},{"lng":[-73.5716860611154,-73.5718528802025,-73.5726519974425,-73.5724824877249,-73.5716860611154],"lat":[45.5616301538998,45.561375837295,45.5616358053669,45.5618957722358,45.5616301538998]},{"lng":[-73.5727192632035,-73.572347956203,-73.5725605160076,-73.5726143286164,-73.5726224005077,-73.5725093940293,-73.5723022154856,-73.5719793398331,-73.5724179125944,-73.5727488601382,-73.572544372225,-73.5726493068121,-73.5727031194208,-73.5728161258992,-73.5727461695078,-73.5728484134644,-73.572802672747,-73.5729049167036,-73.5727192632035],"lat":[45.5615623362511,45.5614474228261,45.5611422417696,45.5611591963162,45.5611422417696,45.561106448821,45.5614304683664,45.5613212061704,45.5606449061256,45.560737214968,45.5610574689597,45.5610819588957,45.5609934183074,45.5610216759571,45.5611309387357,45.5611554286396,45.5612270144516,45.5612590396537,45.5615623362511]},{"lng":[-73.5728511040949,-73.5728968448123,-73.5728591759862,-73.5729237511167,-73.5731255483995,-73.5730394482255,-73.5730098512907,-73.5729614199428,-73.5728511040949],"lat":[45.561217595271,45.5611177518599,45.5611121003407,45.5610028375255,45.5610650043258,45.5611874538826,45.5611818023704,45.5612533881487,45.561217595271]},{"lng":[-73.5551628995978,-73.554961102315,-73.5551978777935,-73.5551790433805,-73.5552301653588,-73.5552463091414,-73.555391603185,-73.555588019207,-73.5551628995978],"lat":[45.5616645336488,45.561472383549,45.5613537022767,45.5613329801241,45.5613009549641,45.5613216771286,45.5612519752724,45.5614384746396,45.5616645336488]},{"lng":[-73.5692712202976,-73.5690667323843,-73.5689026039276,-73.5688824241993,-73.5687640364601,-73.5687411661014,-73.56867120971,-73.5686389221448,-73.568597217373,-73.5685689657534,-73.5685797282751,-73.5686254689926,-73.5686913894383,-73.5687505833079,-73.5688541725798,-73.5689012586124,-73.5689281649168,-73.5689577618516,-73.5691891560693,-73.5691703216562,-73.5692712202976],"lat":[45.5643818801377,45.5645052643284,45.5643771708121,45.5643903569228,45.5642933447498,45.5643074727466,45.5642547282072,45.5642782748827,45.5642236465804,45.5641680763563,45.5641068548594,45.5640588194843,45.5641125060773,45.564083308112,45.5641501808488,45.5641266341196,45.5641360528124,45.5641181572946,45.564297112216,45.5643074727466,45.5643818801377]},{"lng":[-73.5579813349813,-73.5580136225465,-73.557776847068,-73.5576530780679,-73.5576772937418,-73.5573517274589,-73.5573167492632,-73.5569562047845,-73.5569669673063,-73.5568889390236,-73.5570073267629,-73.5572306490892,-73.5572656272849,-73.5570396143281,-73.5570638300021,-73.5573517274589,-73.5572817710675,-73.5576880562636,-73.5577068906766,-73.557819897155,-73.5578737097638,-73.5577741564376,-73.5578145158941,-73.5577311063506,-73.5576988187853,-73.5575562153721,-73.5576207905026,-73.5575346903286,-73.5574055400676,-73.5575938841982,-73.5575831216765,-73.5576584593287,-73.5574270651111,-73.5574566620459,-73.5573624899806,-73.557346346198,-73.5568781765018,-73.5564234599579,-73.5568970109149,-73.556872795241,-73.5566736885886,-73.556571444632,-73.5566709979581,-73.5566279478711,-73.556547228958,-73.5563346691534,-73.5563857911318,-73.5562297345664,-73.5562943096969,-73.5559956497183,-73.5559310745878,-73.5560252466531,-73.5559902684574,-73.5558611181964,-73.555834211892,-73.5558691900877,-73.555858427566,-73.5559256933269,-73.5559310745878,-73.5560198653923,-73.5559821965661,-73.5562243533055,-73.556547228958,-73.5563104534795,-73.5561221093489,-73.5560252466531,-73.556178612588,-73.5560602248488,-73.5559472183704,-73.5558476650442,-73.5557938524355,-73.5555839832613,-73.5556647021745,-73.5562243533055,-73.5562943096969,-73.5565741352624,-73.5565364664363,-73.5564718913058,-73.5564073161753,-73.5563992442839,-73.5563481223056,-73.556331978523,-73.5562055188924,-73.5561247999793,-73.5563212160013,-73.5563158347403,-73.5560333185444,-73.5558288306312,-73.5557588742398,-73.556159778175,-73.5561490156532,-73.5561113468271,-73.5561005843054,-73.5559956497183,-73.555963362153,-73.5563239066317,-73.5563857911318,-73.5562835471751,-73.5562943096969,-73.5566790698495,-73.5567678606539,-73.5567382637191,-73.5568216732626,-73.5568620327192,-73.5569292984802,-73.556657544806,-73.5566952136321,-73.5567759325452,-73.5568512701975,-73.5567920763278,-73.55680552948,-73.5567678606539,-73.556848579567,-73.5577418688723,-73.5581104852423,-73.5581266290249,-73.5581616072206,-73.5585356048514,-73.5584172171122,-73.5579813349813],"lat":[45.5569717388549,45.5569208714089,45.556845512145,45.5568982636403,45.5569378272293,45.5570602857807,45.557030142162,45.5571714402346,45.5572223074538,45.5572694066898,45.5574144720885,45.5573259257208,45.5573767928001,45.5574521513515,45.5574841787053,45.5573636050433,45.5572863624051,45.5571186889957,45.5571337607833,45.5571111531004,45.5571921639221,45.5572336112741,45.5572637547835,45.557292014309,45.5572712906584,45.5573315776208,45.5573994003762,45.5574295437967,45.5575105541594,45.5577253252532,45.5577347450192,45.5578364783917,45.5579269079015,45.5579400955261,45.5579834262712,45.5579683547115,45.5581624007335,45.5576386633317,45.5574464994637,45.5574163560523,45.5575124381199,45.5573843286598,45.5573466493513,45.5573108539848,45.557354185215,45.5574464994637,45.5574841787053,45.5575595371127,45.5576198237658,45.5577366289723,45.5576688066239,45.5576273595929,45.557593448363,45.557651851024,45.557834594442,45.5578327104923,45.5578854610609,45.5578929968523,45.5579306757945,45.5579269079015,45.5578703894749,45.5577460487364,45.5581284898265,45.5582339903589,45.558299928091,45.5581831240555,45.5581096504249,45.5579796583816,45.5579570510391,45.5579664707662,45.5582245706765,45.5582189188664,45.5575529432561,45.5572873043891,45.5573306356375,45.5572232494389,45.5571780341361,45.5570894873958,45.5571139790613,45.5571045591912,45.5571177470089,45.5571045591912,45.5571478905805,45.5570555758413,45.5569651449293,45.5569425371785,45.5570706476458,45.5568351502383,45.5567654428163,45.5565789279378,45.5565582040242,45.556569507978,45.5565600880167,45.5566015358348,45.5565355961097,45.5563961804362,45.5564489323533,45.5564941482429,45.5565224081554,45.5563622684633,45.5564583523333,45.5564847282689,45.5565826959213,45.556577043946,45.556691967332,45.5567974705615,45.5568238463379,45.5568012385303,45.5569010896119,45.5569387692192,45.556953841055,45.5569651449292,45.5570386200564,45.5566712434601,45.5567918186078,45.5567748627433,45.5567823986837,45.5569010896119,45.5570913713705,45.5569717388549]}]]],null,"only.park",{"crs":{"crsClass":"L.CRS.EPSG3857","code":null,"proj4def":null,"projectedBounds":null,"options":{}},"pane":"polygon","stroke":true,"color":"#333333","weight":0.5,"opacity":0.9,"fill":true,"fillColor":"#00FF00","fillOpacity":0.6,"smoothFactor":1,"noClip":false},"<div class='scrollableContainer'><table class=mapview-popup id='popup'><tr class='coord'><td><\/td><th><b>Feature ID&emsp;<\/b><\/th><td>1&emsp;<\/td><\/tr><tr><td>1<\/td><th>geometry&emsp;<\/th><td>sfc_POLYGON&emsp;<\/td><\/tr><\/table><\/div>",{"maxWidth":800,"minWidth":50,"autoPan":true,"keepInView":false,"closeButton":true,"closeOnClick":true,"className":""},"1",{"interactive":false,"permanent":false,"direction":"auto","opacity":1,"offset":[0,0],"textsize":"10px","textOnly":false,"className":"","sticky":true},{"stroke":true,"weight":1,"opacity":0.9,"fillOpacity":0.84,"bringToFront":false,"sendToBack":false}]},{"method":"addScaleBar","args":[{"maxWidth":100,"metric":true,"imperial":true,"updateWhenIdle":true,"position":"bottomleft"}]},{"method":"addHomeButton","args":[-73.5734107552259,45.5548173679962,-73.554463335684,45.5691220871125,true,"only.park","Zoom to only.park","<strong> only.park <\/strong>","bottomright"]},{"method":"addLegend","args":[{"colors":["#00FF00"],"labels":["only.park"],"na_color":null,"na_label":"NA","opacity":1,"position":"topright","type":"factor","title":"","extra":null,"layerId":null,"className":"info legend","group":"only.park"}]},{"method":"createMapPane","args":["point",440]},{"method":"addCircleMarkers","args":[[45.5558589183692,45.5558597166328,45.5568417791309,45.5568426347806,45.5568450640686,45.5578254092168,45.5578262993121,45.5578271664631,45.5578280106698,45.5578296302505,45.558808987415,45.5588099119583,45.5588108135565,45.5588116922097,45.5588125479178,45.5588133806809,45.5588141904988,45.5588149773717,45.5597915317809,45.5597925137203,45.559793472714,45.5597944087618,45.5597953218637,45.5597962120198,45.5597970792301,45.5597979234945,45.5597987448131,45.5597995431858,45.5607759881275,45.5607769815738,45.5607779520735,45.5607788996266,45.560779824233,45.5607807258929,45.5607816046061,45.5607824603727,45.5607832931926,45.5607841030659,45.5607848899926,45.5617604385325,45.5617614434865,45.561762425493,45.5617633845522,45.5617643206639,45.5617652338283,45.5617661240452,45.5617669913147,45.5617678356368,45.5617686570115,45.5617694554387,45.5627458994578,45.562746892972,45.562747863538,45.5627488111558,45.5627497358254,45.5627506375468,45.5627515163201,45.5627523721451,45.5627532050219,45.5627540149506,45.562754801931,45.5637313545097,45.5637323365833,45.563733295708,45.5637342318837,45.5637351451104,45.5637360353882,45.5637369027169,45.5637377470967,45.5637385685275,45.5637393670093,45.5647168036877,45.56471777432,45.5647187220026,45.5647196467354,45.5647205485184,45.5647214273517,45.5647222832352,45.5647231161689,45.5647239261529,45.5657022469913,45.5657032061815,45.5657041424211,45.5657050557102,45.5657059460488,45.5657068134368,45.5657076578743,45.5657084793612,45.5666876844197,45.566688632167,45.566689556963,45.5666904588076,45.5666913377009,45.5666921936429,45.5676731159725,45.5676740522761,45.5676749656276,45.567675856027,45.5686585416491],[-73.5571374763921,-73.555518563795,-73.5611858264953,-73.5595668857914,-73.5547100632561,-73.5636153488103,-73.5619963799336,-73.560377410982,-73.5587584419575,-73.5555205036979,-73.5660449558653,-73.5644259588187,-73.5628069616943,-73.5611879644941,-73.55956896722,-73.5579499698741,-73.5563309724583,-73.5547119749745,-73.5700936727912,-73.5684746476603,-73.5668556224467,-73.5652365971524,-73.5636175717794,-73.5619985463296,-73.560379520805,-73.5587604952075,-73.5571414695391,-73.5555224438018,-73.5709044241951,-73.5692853708175,-73.5676663173563,-73.5660472638134,-73.5644282101909,-73.5628091564905,-73.5611901027144,-73.5595710488644,-73.5579519949425,-73.5563329409507,-73.5547138868909,-73.5717152039312,-73.570096122306,-73.5684770405962,-73.5668579588038,-73.5652388769306,-73.5636197949788,-73.5620007129501,-73.5603816308466,-73.5587625486703,-73.557143466423,-73.5555243841067,-73.5709069021277,-73.5692877921683,-73.5676686821253,-73.5660495720007,-73.5644304617963,-73.5628113515142,-73.5611922411562,-73.5595731307244,-73.5579540202207,-73.556334909647,-73.5547157990054,-73.5700985720746,-73.5684794337801,-73.566860295403,-73.5652411569451,-73.5636220184085,-73.5620028797951,-73.5603837411069,-73.5587646023458,-73.5571454635137,-73.5555263246127,-73.56929021377,-73.5676710471394,-73.5660518804271,-73.5644327136351,-73.5628135467653,-73.5611943798196,-73.5595752128001,-73.5579560457087,-73.5563368785474,-73.568481827212,-73.5668626322443,-73.5652434371958,-73.5636242420686,-73.5620050468646,-73.5603858515858,-73.5587666562341,-73.5571474608114,-73.5676734123985,-73.5660541890927,-73.5644349657072,-73.5628157422438,-73.5611965187046,-73.5595772950916,-73.5668649693277,-73.5652457176828,-73.5636264659592,-73.5620072141587,-73.5660564979975],6,null,"rdm.pt",{"crs":{"crsClass":"L.CRS.EPSG3857","code":null,"proj4def":null,"projectedBounds":null,"options":{}},"pane":"point","stroke":true,"color":"#333333","weight":1,"opacity":0.9,"fill":true,"fillColor":"#6666ff","fillOpacity":0.6},null,null,null,{"maxWidth":800,"minWidth":50,"autoPan":true,"keepInView":false,"closeButton":true,"closeOnClick":true,"className":""},["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67","68","69","70","71","72","73","74","75","76","77","78","79","80","81","82","83","84","85","86","87","88","89","90","91","92","93","94","95","96","97","98","99"],{"interactive":false,"permanent":false,"direction":"auto","opacity":1,"offset":[0,0],"textsize":"10px","textOnly":false,"className":"","sticky":true},null]},{"method":"addHomeButton","args":[-73.5717152039312,45.5558589183692,-73.5547100632561,45.5686585416491,true,"rdm.pt","Zoom to rdm.pt","<strong> rdm.pt <\/strong>","bottomright"]},{"method":"addLayersControl","args":[["CartoDB.Positron","CartoDB.DarkMatter","OpenStreetMap","Esri.WorldImagery","OpenTopoMap"],["only.park","rdm.pt"],{"collapsed":true,"autoZIndex":true,"position":"topleft"}]},{"method":"addHomeButton","args":[-73.5734107552259,45.5548173679962,-73.554463335684,45.5691220871125,true,null,"Zoom to full extent","<strong>Zoom full<\/strong>","bottomleft"]}],"limits":{"lat":[45.5548173679962,45.5691220871125],"lng":[-73.5734107552259,-73.554463335684]},"fitBounds":[45.5548173679962,-73.5734107552259,45.5691220871125,-73.554463335684,[]]},"evals":[],"jsHooks":{"render":[{"code":"function(el, x, data) {\n  return (\n      function(el, x, data) {\n      // get the leaflet map\n      var map = this; //HTMLWidgets.find('#' + el.id);\n      // we need a new div element because we have to handle\n      // the mouseover output separately\n      // debugger;\n      function addElement () {\n      // generate new div Element\n      var newDiv = $(document.createElement('div'));\n      // append at end of leaflet htmlwidget container\n      $(el).append(newDiv);\n      //provide ID and style\n      newDiv.addClass('lnlt');\n      newDiv.css({\n      'position': 'relative',\n      'bottomleft':  '0px',\n      'background-color': 'rgba(255, 255, 255, 0.7)',\n      'box-shadow': '0 0 2px #bbb',\n      'background-clip': 'padding-box',\n      'margin': '0',\n      'padding-left': '5px',\n      'color': '#333',\n      'font': '9px/1.5 \"Helvetica Neue\", Arial, Helvetica, sans-serif',\n      'z-index': '700',\n      });\n      return newDiv;\n      }\n\n\n      // check for already existing lnlt class to not duplicate\n      var lnlt = $(el).find('.lnlt');\n\n      if(!lnlt.length) {\n      lnlt = addElement();\n\n      // grab the special div we generated in the beginning\n      // and put the mousmove output there\n\n      map.on('mousemove', function (e) {\n      if (e.originalEvent.ctrlKey) {\n      if (document.querySelector('.lnlt') === null) lnlt = addElement();\n      lnlt.text(\n                           ' lon: ' + (e.latlng.lng).toFixed(5) +\n                           ' | lat: ' + (e.latlng.lat).toFixed(5) +\n                           ' | zoom: ' + map.getZoom() +\n                           ' | x: ' + L.CRS.EPSG3857.project(e.latlng).x.toFixed(0) +\n                           ' | y: ' + L.CRS.EPSG3857.project(e.latlng).y.toFixed(0) +\n                           ' | epsg: 3857 ' +\n                           ' | proj4: +proj=merc +a=6378137 +b=6378137 +lat_ts=0.0 +lon_0=0.0 +x_0=0.0 +y_0=0 +k=1.0 +units=m +nadgrids=@null +no_defs ');\n      } else {\n      if (document.querySelector('.lnlt') === null) lnlt = addElement();\n      lnlt.text(\n                      ' lon: ' + (e.latlng.lng).toFixed(5) +\n                      ' | lat: ' + (e.latlng.lat).toFixed(5) +\n                      ' | zoom: ' + map.getZoom() + ' ');\n      }\n      });\n\n      // remove the lnlt div when mouse leaves map\n      map.on('mouseout', function (e) {\n      var strip = document.querySelector('.lnlt');\n      if( strip !==null) strip.remove();\n      });\n\n      };\n\n      //$(el).keypress(67, function(e) {\n      map.on('preclick', function(e) {\n      if (e.originalEvent.ctrlKey) {\n      if (document.querySelector('.lnlt') === null) lnlt = addElement();\n      lnlt.text(\n                      ' lon: ' + (e.latlng.lng).toFixed(5) +\n                      ' | lat: ' + (e.latlng.lat).toFixed(5) +\n                      ' | zoom: ' + map.getZoom() + ' ');\n      var txt = document.querySelector('.lnlt').textContent;\n      console.log(txt);\n      //txt.innerText.focus();\n      //txt.select();\n      setClipboardText('\"' + txt + '\"');\n      }\n      });\n\n      }\n      ).call(this.getMap(), el, x, data);\n}","data":null},{"code":"function(el, x, data) {\n  return (function(el,x,data){\n           var map = this;\n\n           map.on('keypress', function(e) {\n               console.log(e.originalEvent.code);\n               var key = e.originalEvent.code;\n               if (key === 'KeyE') {\n                   var bb = this.getBounds();\n                   var txt = JSON.stringify(bb);\n                   console.log(txt);\n\n                   setClipboardText('\\'' + txt + '\\'');\n               }\n           })\n        }).call(this.getMap(), el, x, data);\n}","data":null}]}}</script>





???

#################################
---
class: inverse, center, middle

# Advanced simulations


---
## Advanced simulations

- Now that we have the building blocks to make simulations, we are going to add another layer:
- We want to simulate a process and then see how our simulation responds to a fluctuation in a parameter
- For this, we need a certain workflow that resembles this: 
1. create a blank vector in which a value of interest will be stored in 
2. Within a loop or many `for(){}` loops, we are going to simulate data, get a desired result and store the result for future analysis 
3. Summarize our findings (in a plot, with calculations, etc.) 

.alert[ATTENTION]. This is an iterative process. Usually, we make small increments writing up the code and then put it all together in a cohesive function that will make exactly what we are looking for. 


---
## Advanced simulations
.tiny[

```r
# Defining the population 
n = 600 # Number of elements to be generated 
set.seed(13) # Set RNG 
x = rnorm(n) # Generate numbers from normal distribution 
reedddd = scales::alpha("blue",.4) # Make colour transparent 

# Definte the function 
sample.mean.pop.est &lt;- function(x,n, sample.size, ylim = NULL) {
  # x: is the actual values of the trait measured 
  # n: size of the population (number of individuals or items)
  # sample.size: how big is the sample size from which the MEAN will be calculated from 
  # ylim: add a maximum if needed 
  # histogram of the population 
  
  # Just get the stats from the histogram 
  pop.hist = hist(x, plot = F) # Make base histogram 
  
  # Make empty vector
  tmp.v = c(NA) 
  
  # For loop to calculate the mean based on a sample from the population 
  for (i in 1:n) {
    tmp = sample(x = x, size = sample.size, replace = FALSE)
    # Record that information (mean of the sample)
    tmp.v[i] = mean(tmp)
  } # End i
  
  # Sample histogram 
  sample.hist = hist(tmp.v, plot = F)
  # Population histogram 
  hist(x, ylim = range(c(0,c(sample.hist$counts, pop.hist$counts), ylim)), 
       main = paste("Sample n =", sample.size))
  # Add the sample estimate 
  sample.hist = hist(tmp.v, col = reedddd, add=T)
} # End sample.mean.pop.est
```
]

---
## Advanced simulations

.small[

```r
par(mfrow=c(2,2), lwd = .3)
sample.mean.pop.est(x = x, n = n, sample.size = 1, ylim = 300)
sample.mean.pop.est(x = x, n = n, sample.size = 10, ylim = 300)
sample.mean.pop.est(x = x, n = n, sample.size = 50, ylim = 300)
sample.mean.pop.est(x = x, n = n, sample.size = 500, ylim = 300)
```

&lt;img src="workshopXX-en_files/figure-html/simulate_sampling_plots-1.png" width="432" style="display: block; margin: auto;" /&gt;
]

---
## Advanced simulations (natural selection)

- The sentence "METHINKS IT IS LIKE A WEASEL" is from Shakespeare's *Hamlet*
- Richard Dawkins, in his 1986 book *The Blind Watchmaker*, used it to simulate natural selection on random genetic mutation. 


```
#   0 0.00 TJWPPBT IYDHKFIEOSQQEZJCSMCQ
#  20 0.43 MJTPVNT IID KS EOSE E WESMCQ 
#  40 0.64 MJTPVNQ  IT IS LOLE A WESSCQ 
#  60 0.79 MJTPVNKS IT IS LIKE A WESSCQ 
#  80 0.93 MJTPINKS IT IS LIKE A WEASEL 
# 100 0.93 MJTPINKS IT IS LIKE A WEASEL
# 119 1.00 METHINKS IT IS LIKE A WEASEL
```

- After the same number of generations, but if a monkey would randomly be typing on a keyboard! 

```
# 119 0.04 NSPZXZKXBVKFGJQHGFALTHZLCINM
```

- Source: [Evolutionary algorithm from Rosettacode.org](https://rosettacode.org/wiki/Evolutionary_algorithm).

???

- The full tradgedy's name is *The Tragedy of Hamlet, Prince of Denmark*
- Get more information here https://en.wikipedia.org/wiki/Weasel_program 


---
## References
.tiny[
Book
- Otto, S.P. &amp; Day, T., 2007. A Biologist's Guide to Mathematical Modeling in Ecology and Evolution. pp.1–973.

&lt;!-- ![:scale 5%](images/Otto_Day.2007. Biologist Guide to Mathematical Modeling in Ecology and Evolution.jpg) --&gt;

Links 
- [Getting started simulating data in R: some helpful functions and how to use them (Ariel Muldoon)](https://aosmith.rbind.io/2018/08/29/getting-started-simulating-data/)
- [Simulate Linear model](https://aosmith.rbind.io/2018/01/09/simulate-simulate-part1/) and [Simulate Linear mixed model](https://aosmith.rbind.io/2018/04/23/simulate-simulate-part-2/) by Ariel Muldoon
- [Simulate Poisson model (Ariel Muldoon)](https://aosmith.rbind.io/2018/07/18/simulate-poisson-edition/)
- [Simulate binomial generalized linear mixed model (Ariel Muldoon)](https://aosmith.rbind.io/2020/08/20/simulate-binomial-glmm/)
- [Statistical Simulation in R with Code — Part 1](https://towardsdatascience.com/statistical-simulation-in-r-part-1-d9cb4dc393c9)
- [Power Analysis by Data Simulation in R – Part 2](https://www.r-bloggers.com/2020/05/power-analysis-by-data-simulation-in-r-part-ii/)
- [bios221; Lab 3: Simulations in R](https://web.stanford.edu/class/bios221/labs/simulation/Lab_3_simulation.html)
- [R Programming for Data Science; See 20 Simulation](https://bookdown.org/rdpeng/rprogdatascience/simulation.html)
- [Simulation in R (YouTube)](https://www.youtube.com/watch?v=tvv4IA8PEzw)
- [Introduction to Simulations in R](http://www.columbia.edu/~cjd11/charles_dimaggio/DIRE/resources/R/simRreg.pdf)
- [Crump Lab; see Chapter 5 Simulating and Analyzing Data in R](https://www.crumplab.com/programmingforpsych/simulating-and-analyzing-data-in-r.html)
- [Intro to R Lecture Notes; see Chapter 13 Simulations](https://users.phhp.ufl.edu/rlp176/Courses/PHC6089/R_notes/simulations.html)
- [Simulation Cheat Sheet, by Nick Huntington-Klein](https://nickch-k.github.io/introcausality/Cheat%20Sheets/Simulation_Cheat_Sheet.pdf)
- [Simulating mixed effets; see chapter 4](https://debruine.github.io/tutorials/sim-lmer.html)
- [Cours 'Econ 224'; Lab #9 - Logistic Regression Part I](https://ditraglia.com/econ224/lab09.pdf)
- [ECON 41 Labs; useful ressources about distributions](https://bookdown.org/gabriel_butler/ECON41Labs/)
- [Binomial regression in R](https://kkorthauer.org/fungeno2019/methylation/vignettes/1-binomial-regression.html#4_pitfalls_of_glm)
- [Simple Data Simulations in R, of course](https://it.unt.edu/simple-data-simulations)
- [Probability Cheat sheet](https://static1.squarespace.com/static/54bf3241e4b0f0d81bf7ff36/t/55e9494fe4b011aed10e48e5/1441352015658/probability_cheatsheet.pdf)
]

---
class: inverse, center, bottom

# Thank you for attending this workshop!

![:scale 50%](images/qcbs_logo.png)


    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script src="qcbsR-macros.js"></script>
<script>var slideshow = remark.create({
"highlightLines": true
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
